// fumadocs-ui/dist/layouts/home/slots/header.js
"use client";
import { cn } from "../../../utils/cn.js";
import { buttonVariants } from "../../../components/ui/button.js";
import { mergeRefs } from "../../../utils/merge-refs.js";
import { Collapsible, CollapsibleContent, CollapsibleTrigger } from "../../../components/ui/collapsible.js";
import { LinkItem } from "../../shared/client.js";
import "../../shared/index.js";
import { useIsScrollTop } from "../../../utils/use-is-scroll-top.js";
import { useHomeLayout } from "../index.js";
import Link from "fumadocs-core/link";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { ChevronDown, Languages } from "lucide-react";
import { cva } from "class-variance-authority";
import { Fragment as Fragment$1, createContext, use, useEffect, useEffectEvent, useMemo, useRef, useState } from "react";
import { useTranslations } from "@fuma-translate/react";
import { NavigationMenu } from "@base-ui/react/navigation-menu";
//#region src/layouts/home/slots/header.tsx
const navItemVariants = cva("[&_svg]:size-4", {
	variants: { variant: {
		main: "inline-flex items-center gap-1 p-2 text-fd-muted-foreground transition-colors hover:text-fd-accent-foreground data-[active=true]:text-fd-primary",
		button: buttonVariants({
			color: "secondary",
			className: "gap-1.5"
		}),
		icon: buttonVariants({
			color: "ghost",
			size: "icon"
		})
	} },
	defaultVariants: { variant: "main" }
});
const MobileNavigationMenuContext = createContext(null);
function Header(props) {
	const { navItems, menuItems, slots, props: { nav } } = useHomeLayout();
	const headerRef = useRef(null);
	const listRef = useRef(null);
	const [open, setOpen] = useState(false);
	const t = useTranslations({ note: "home layout header" });
	const transparentMode = nav?.transparentMode ?? "none";
	const isTop = useIsScrollTop({ enabled: transparentMode === "top" }) ?? true;
	const isNavTransparent = transparentMode === "top" ? isTop : transparentMode === "always";
	const onClick = useEffectEvent((e) => {
		const element = headerRef.current;
		if (!open || !element) return;
		if (element !== e.target && !element.contains(e.target)) setOpen(false);
	});
	useEffect(() => {
		window.addEventListener("click", onClick);
		return () => {
			window.removeEventListener("click", onClick);
		};
	}, []);
	const list = /* @__PURE__ */ jsxs(NavigationMenu.List, {
		ref: listRef,
		className: "flex h-14 w-full mx-auto max-w-(--fd-layout-width) items-center px-4",
		children: [
			slots.navTitle && /* @__PURE__ */ jsx(slots.navTitle, { className: "inline-flex items-center gap-2.5 font-semibold" }),
			nav?.children,
			/* @__PURE__ */ jsx("ul", {
				className: "flex flex-row items-center gap-2 px-6 max-sm:hidden",
				children: navItems.filter((item) => !isSecondary(item)).map((item, i) => /* @__PURE__ */ jsx(NavigationMenuLinkItem, {
					item,
					className: "text-sm"
				}, i))
			}),
			/* @__PURE__ */ jsxs("div", {
				className: "flex flex-row items-center justify-end gap-1.5 flex-1 max-lg:hidden",
				children: [
					slots.searchTrigger && /* @__PURE__ */ jsx(slots.searchTrigger.full, {
						hideIfDisabled: true,
						className: "w-full rounded-full ps-2.5 max-w-[240px]"
					}),
					slots.themeSwitch && /* @__PURE__ */ jsx(slots.themeSwitch, {}),
					slots.languageSelect && /* @__PURE__ */ jsx(slots.languageSelect.root, { children: /* @__PURE__ */ jsx(Languages, { className: "size-5" }) }),
					/* @__PURE__ */ jsx("ul", {
						className: "flex flex-row gap-2 items-center empty:hidden",
						children: navItems.filter(isSecondary).map((item, i) => /* @__PURE__ */ jsx(NavigationMenuLinkItem, {
							className: cn(item.type === "icon" && "-mx-1 first:ms-0 last:me-0"),
							item
						}, i))
					})
				]
			}),
			/* @__PURE__ */ jsxs("div", {
				className: "flex flex-row items-center ms-auto -me-1.5 lg:hidden",
				children: [slots.searchTrigger && /* @__PURE__ */ jsx(slots.searchTrigger.sm, {
					hideIfDisabled: true,
					className: "p-2"
				}), /* @__PURE__ */ jsx(CollapsibleTrigger, {
					"aria-label": t("Toggle Menu", { note: "aria-label" }),
					className: cn(buttonVariants({
						size: "icon",
						color: "ghost"
					})),
					onPointerEnter: nav?.enableHoverToOpen ? () => {
						setOpen(true);
					} : void 0,
					children: /* @__PURE__ */ jsx(ChevronDown, { className: cn("transition-transform", open && "rotate-180") })
				})]
			})
		]
	});
	return /* @__PURE__ */ jsx(Collapsible, {
		open,
		onOpenChange: setOpen,
		render: /* @__PURE__ */ jsx("header", {
			id: "nd-nav",
			...props,
			ref: mergeRefs(headerRef, props.ref),
			className: cn("sticky h-14 top-0 z-40", props.className),
			children: /* @__PURE__ */ jsxs(NavigationMenu.Root, {
				className: (s) => cn("backdrop-blur-lg border-b transition-[box-shadow,background-color,border-radius]", open && "max-lg:shadow-lg max-lg:rounded-b-2xl", (open || !isNavTransparent || s.open) && "bg-fd-background/80"),
				children: [
					list,
					/* @__PURE__ */ jsx(CollapsibleContent, {
						className: "mx-auto max-w-(--fd-layout-width) lg:hidden",
						children: /* @__PURE__ */ jsx("div", {
							className: "flex flex-col pt-2 p-4 sm:flex-row sm:items-center sm:justify-end",
							children: /* @__PURE__ */ jsxs(MobileNavigationMenuContext, {
								value: useMemo(() => ({ setOpen }), []),
								children: [menuItems.filter((item) => !isSecondary(item)).map((item, i) => /* @__PURE__ */ jsx(MobileNavigationMenuLinkItem, {
									item,
									className: "sm:hidden"
								}, i)), /* @__PURE__ */ jsxs("div", {
									className: "-ms-1.5 flex flex-row items-center gap-2 max-sm:mt-2",
									children: [
										menuItems.filter(isSecondary).map((item, i) => /* @__PURE__ */ jsx(MobileNavigationMenuLinkItem, {
											item,
											className: cn(item.type === "icon" && "-mx-1 first:ms-0")
										}, i)),
										/* @__PURE__ */ jsx("div", {
											role: "separator",
											className: "flex-1"
										}),
										slots.languageSelect && /* @__PURE__ */ jsxs(slots.languageSelect.root, { children: [
											/* @__PURE__ */ jsx(Languages, { className: "size-5" }),
											slots.languageSelect.text && /* @__PURE__ */ jsx(slots.languageSelect.text, {}),
											/* @__PURE__ */ jsx(ChevronDown, { className: "size-3 text-fd-muted-foreground" })
										] }),
										slots.themeSwitch && /* @__PURE__ */ jsx(slots.themeSwitch, {})
									]
								})]
							})
						})
					}),
					/* @__PURE__ */ jsx(NavigationMenu.Portal, { children: /* @__PURE__ */ jsx(NavigationMenu.Positioner, {
						side: "bottom",
						anchor: listRef,
						collisionPadding: {
							top: 5,
							bottom: 5
						},
						className: "z-40 box-border h-(--positioner-height) w-(--anchor-width) max-w-(--available-width) duration-(--duration) ease-(--easing) before:absolute before:content-[''] data-instant:transition-none data-[side=bottom]:before:top-[-10px] data-[side=bottom]:before:right-0 data-[side=bottom]:before:left-0 data-[side=bottom]:before:h-2.5 data-[side=left]:before:top-0 data-[side=left]:before:right-[-10px] data-[side=left]:before:bottom-0 data-[side=left]:before:w-2.5 data-[side=right]:before:top-0 data-[side=right]:before:bottom-0 data-[side=right]:before:left-[-10px] data-[side=right]:before:w-2.5 data-[side=top]:before:right-0 data-[side=top]:before:bottom-[-10px] data-[side=top]:before:left-0 data-[side=top]:before:h-2.5",
						style: {
							["--duration"]: "0.35s",
							["--easing"]: "cubic-bezier(0.22, 1, 0.36, 1)"
						},
						children: /* @__PURE__ */ jsx(NavigationMenu.Popup, {
							className: "relative border h-(--popup-height) w-full rounded-xl bg-fd-popover/80 text-fd-popover-foreground backdrop-blur-md shadow-lg transition-[opacity,width,height] duration-(--duration) ease-(--easing) data-ending-style:opacity-0 data-ending-style:duration-150 data-starting-style:opacity-0",
							children: /* @__PURE__ */ jsx(NavigationMenu.Viewport, { className: "relative size-full overflow-hidden" })
						})
					}) })
				]
			})
		})
	});
}
function isSecondary(item) {
	if ("secondary" in item && item.secondary != null) return item.secondary;
	return item.type === "icon";
}
function NavigationMenuLinkItem({ item, className, ...props }) {
	if (item.type === "custom") return item.children;
	if (item.type === "menu") {
		const children = item.items.map((child, j) => {
			if (child.type === "custom") return /* @__PURE__ */ jsx(Fragment$1, { children: child.children }, j);
			const { banner = child.icon ? /* @__PURE__ */ jsx("div", {
				className: "w-fit rounded-md border bg-fd-muted p-1 [&_svg]:size-4",
				children: child.icon
			}) : null, ...rest } = child.menu ?? {};
			return /* @__PURE__ */ jsx(NavigationMenu.Link, { render: /* @__PURE__ */ jsx(Link, {
				href: child.url,
				external: child.external,
				...rest,
				className: cn("flex flex-col gap-2 rounded-lg border bg-fd-card p-3 transition-colors hover:bg-fd-accent/80 hover:text-fd-accent-foreground", rest.className),
				children: rest.children ?? /* @__PURE__ */ jsxs(Fragment, { children: [
					banner,
					/* @__PURE__ */ jsx("p", {
						className: "text-base font-medium",
						children: child.text
					}),
					/* @__PURE__ */ jsx("p", {
						className: "text-sm text-fd-muted-foreground empty:hidden",
						children: child.description
					})
				] })
			}) }, `${j}-${child.url}`);
		});
		return /* @__PURE__ */ jsxs(NavigationMenu.Item, {
			className: cn("list-none", className),
			...props,
			children: [/* @__PURE__ */ jsx(NavigationMenu.Trigger, {
				className: cn(navItemVariants(), "rounded-md"),
				children: item.url ? /* @__PURE__ */ jsx(Link, {
					href: item.url,
					external: item.external,
					children: item.text
				}) : item.text
			}), /* @__PURE__ */ jsx(NavigationMenu.Content, {
				className: cn("h-full w-(--anchor-width) max-w-(--available-width) p-3", "transition-[opacity,transform,translate] duration-(--duration) ease-(--easing)", "data-starting-style:opacity-0 data-ending-style:opacity-0", "data-starting-style:data-[activation-direction=left]:-translate-x-1/2", "data-starting-style:data-[activation-direction=right]:translate-x-1/2", "data-ending-style:data-[activation-direction=left]:translate-x-1/2", "data-ending-style:data-[activation-direction=right]:-translate-x-1/2", "grid grid-cols-1 gap-2 md:grid-cols-2 lg:grid-cols-3"),
				children
			})]
		});
	}
	return /* @__PURE__ */ jsx(NavigationMenu.Item, {
		className: cn("list-none", className),
		...props,
		children: /* @__PURE__ */ jsx(NavigationMenu.Link, { render: /* @__PURE__ */ jsx(LinkItem, {
			item,
			"aria-label": item.type === "icon" ? item.label : void 0,
			className: cn(navItemVariants({ variant: item.type })),
			children: item.type === "icon" ? item.icon : item.text
		}) })
	});
}
function MobileNavigationMenuLinkItem({ item, ...props }) {
	const { setOpen } = use(MobileNavigationMenuContext);
	if (item.type === "custom") return /* @__PURE__ */ jsx("div", {
		className: cn("grid", props.className),
		children: item.children
	});
	if (item.type === "menu") return /* @__PURE__ */ jsxs("div", {
		className: cn("mb-4 flex flex-col", props.className),
		children: [/* @__PURE__ */ jsx("p", {
			className: "mb-1 text-sm text-fd-muted-foreground",
			children: item.url ? /* @__PURE__ */ jsxs(Link, {
				href: item.url,
				external: item.external,
				onClick: () => setOpen(false),
				children: [item.icon, item.text]
			}) : /* @__PURE__ */ jsxs(Fragment, { children: [item.icon, item.text] })
		}), item.items.map((child, i) => /* @__PURE__ */ jsx(MobileNavigationMenuLinkItem, { item: child }, i))]
	});
	return /* @__PURE__ */ jsxs(LinkItem, {
		item,
		className: cn({
			main: "inline-flex items-center gap-2 py-1.5 transition-colors hover:text-fd-popover-foreground/50 data-[active=true]:font-medium data-[active=true]:text-fd-primary [&_svg]:size-4",
			icon: buttonVariants({
				size: "icon",
				color: "ghost"
			}),
			button: buttonVariants({
				color: "secondary",
				className: "gap-1.5 [&_svg]:size-4"
			})
		}[item.type ?? "main"], props.className),
		"aria-label": item.type === "icon" ? item.label : void 0,
		onClick: () => setOpen(false),
		children: [item.icon, item.type === "icon" ? void 0 : item.text]
	});
}
//#endregion
export { Header, navItemVariants };
