// fumadocs-ui/dist/components/ui/accordion.js
"use client";
import { cn } from "../../utils/cn.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { ChevronRight } from "lucide-react";
import { Accordion as Accordion$1 } from "@base-ui/react/accordion";
//#region src/components/ui/accordion.tsx
function Accordion({ className, ...props }) {
	return /* @__PURE__ */ jsx(Accordion$1.Root, {
		className: (s) => cn("divide-y divide-fd-border overflow-hidden rounded-lg border bg-fd-card", typeof className === "function" ? className(s) : className),
		...props
	});
}
function AccordionItem({ children, ...props }) {
	return /* @__PURE__ */ jsx(Accordion$1.Item, {
		...props,
		children
	});
}
function AccordionHeader({ className, children, ...props }) {
	return /* @__PURE__ */ jsx(Accordion$1.Header, {
		className: (s) => cn("scroll-m-24 not-prose flex flex-row items-center text-fd-card-foreground font-medium has-focus-visible:bg-fd-accent", typeof className === "function" ? className(s) : className),
		...props,
		children
	});
}
function AccordionTrigger({ className, children, ...props }) {
	return /* @__PURE__ */ jsxs(Accordion$1.Trigger, {
		className: (s) => cn("group flex flex-1 items-center gap-2 px-3 py-2.5 text-start focus-visible:outline-none", typeof className === "function" ? className(s) : className),
		...props,
		children: [/* @__PURE__ */ jsx(ChevronRight, { className: "size-4 shrink-0 text-fd-muted-foreground transition-transform duration-200 group-data-[panel-open]:rotate-90" }), children]
	});
}
function AccordionContent({ className, children, ...props }) {
	return /* @__PURE__ */ jsx(Accordion$1.Panel, {
		className: (s) => cn("h-(--accordion-panel-height) overflow-hidden transition-[height] ease-out data-[ending-style]:h-0 data-[starting-style]:h-0", typeof className === "function" ? className(s) : className),
		...props,
		children
	});
}
//#endregion
export { Accordion, AccordionContent, AccordionHeader, AccordionItem, AccordionTrigger };
