// fumadocs-ui/dist/components/ui/popover.js
"use client";
import { cn } from "../../utils/cn.js";
import { jsx } from "react/jsx-runtime";
import "react";
import { Popover as Popover$1 } from "@base-ui/react/popover";
//#region src/components/ui/popover.tsx
const Popover = Popover$1.Root;
const PopoverTrigger = Popover$1.Trigger;
function PopoverContent({ className, align = "center", sideOffset = 4, ...props }) {
	return /* @__PURE__ */ jsx(Popover$1.Portal, { children: /* @__PURE__ */ jsx(Popover$1.Positioner, {
		align,
		side: "bottom",
		sideOffset,
		className: "z-50",
		children: /* @__PURE__ */ jsx(Popover$1.Popup, {
			className: (s) => cn("z-50 origin-(--transform-origin) overflow-y-auto max-h-(--available-height) min-w-[240px] max-w-[98vw] rounded-xl border bg-fd-popover/60 backdrop-blur-lg p-2 text-sm text-fd-popover-foreground shadow-lg focus-visible:outline-none data-[closed]:animate-fd-popover-out data-[open]:animate-fd-popover-in", typeof className === "function" ? className(s) : className),
			...props
		})
	}) });
}
const PopoverClose = Popover$1.Close;
//#endregion
export { Popover, PopoverClose, PopoverContent, PopoverTrigger };
