// fumadocs-mdx/dist/plugins/last-modified.js
import { n as ident } from "../codegen-ESdi9CDJ.js";
import path from "node:path";
import { x } from "tinyexec";
//#region src/plugins/last-modified.ts
const cache = /* @__PURE__ */ new Map();
const ExtendTypes = `{
  /**
   * Last modified date of document file, obtained from version control.
   *
   */
  lastModified?: Date;
}`;
/**
* Injects `lastModified` property to page exports.
*/
function lastModified(options = {}) {
	const { versionControl = "git", filter = () => true } = options;
	let fn;
	return {
		name: "last-modified",
		"index-file": {
			generateTypeConfig() {
				const lines = [];
				lines.push("{");
				lines.push("  DocData: {");
				for (const collection of this.core.getCollections()) if (filter(collection.name)) lines.push(ident(`${collection.name}: ${ExtendTypes},`, 2));
				lines.push("  }");
				lines.push("}");
				return lines.join("\n");
			},
			serverOptions(options) {
				options.doc ??= {};
				options.doc.passthroughs ??= [];
				options.doc.passthroughs.push("lastModified");
			}
		},
		config() {
			const { workspace } = this.core.getOptions();
			const cwd = workspace ? path.resolve(workspace.dir) : process.cwd();
			switch (versionControl) {
				case "git":
					fn = (v) => getGitTimestamp(v, cwd);
					break;
				default: fn = versionControl;
			}
		},
		doc: { async vfile(file) {
			if (!filter(this.collection.name)) return;
			const timestamp = await fn(this.filePath);
			if (timestamp) {
				file.data["mdx-export"] ??= [];
				file.data["mdx-export"].push({
					name: "lastModified",
					value: timestamp
				});
			}
		} }
	};
}
async function getGitTimestamp(file, cwd) {
	return (await getGitTimestamps(cwd)).get(path.resolve(cwd, file)) ?? null;
}
function getGitTimestamps(cwd) {
	const cached = cache.get(cwd);
	if (cached) return cached;
	const promise = (async () => {
		const timestamps = /* @__PURE__ */ new Map();
		const root = await x("git", ["rev-parse", "--show-toplevel"], { nodeOptions: { cwd } });
		if (root.exitCode !== 0) return timestamps;
		const out = await x("git", [
			"-c",
			"core.quotepath=off",
			"log",
			"--format=commit:%aI",
			"--name-only"
		], { nodeOptions: { cwd } });
		if (out.exitCode !== 0) return timestamps;
		let date;
		for (const line of out.stdout.split("\n")) if (line.startsWith("commit:")) {
			const parsed = new Date(line.slice(7));
			date = isNaN(parsed.getTime()) ? void 0 : parsed;
		} else if (line.length > 0 && date) {
			const file = path.join(root.stdout.trim(), line);
			if (!timestamps.has(file)) timestamps.set(file, date);
		}
		return timestamps;
	})();
	cache.set(cwd, promise);
	return promise;
}
//#endregion
export { lastModified as default };
