// fumapress/dist/lib/shared.js
import { __toESM } from "../_virtual/_rolldown/runtime.js";
import { getGitRootDir } from "./fs.js";
import { require_deepmerge } from "../node_modules/.pnpm/@fastify_deepmerge@3.2.1/node_modules/@fastify/deepmerge/index.js";
import { applyDefaultsPlugin } from "../plugins/internal/defaults.js";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Fragment as Fragment$1, isValidElement } from "react";
import path from "node:path";
import { loader } from "fumadocs-core/source";
import { AsyncLocalStorage } from "node:async_hooks";
import { dynamicLoader } from "fumadocs-core/source/dynamic";
//#region src/lib/shared.tsx
var import_deepmerge = /* @__PURE__ */ __toESM(require_deepmerge(), 1);
const appContext = new AsyncLocalStorage({ name: "fumapress:core" });
function getPressContext() {
	const store = appContext.getStore();
	if (!store) throw new Error("[Fumapress] Missing server context for Fumapress, make sure to use the middlewares from createRouter()");
	return store;
}
const PLUGIN_ORDER = {
	pre: -1,
	post: 1,
	_: 0
};
function flattenPlugins(plugins) {
	const out = [];
	for (const plugin of plugins) {
		if (!plugin) continue;
		if (Array.isArray(plugin)) out.push(...flattenPlugins(plugin));
		else out.push(plugin);
	}
	return out;
}
function resolvePlugins(plugins) {
	return flattenPlugins(plugins).sort((a, b) => PLUGIN_ORDER[a.enforce ?? "_"] - PLUGIN_ORDER[b.enforce ?? "_"]);
}
async function initApp(builder) {
	const config = builder.get();
	const { translations, site, mode = "default", layouts } = config;
	const plugins = resolvePlugins([...config.plugins, ...applyDefaultsPlugin()]);
	const ctx = {
		$context: void 0,
		getLoader() {
			throw new Error("[Fumapress] Content loader is not initialized yet, please access it after init()");
		},
		revalidateLoader: () => Promise.resolve(void 0),
		invalidateLoader: () => void 0,
		i18nConfig: translations && "config" in translations ? translations.config : void 0,
		layouts: {
			...layouts,
			root: layouts.root ?? (await import("../layouts/root.js")).createRootLayout(),
			page: layouts.page ?? (await import("../layouts/docs.js")).createDocsLayoutPage(),
			notFound: layouts.notFound ?? (await import("fumadocs-ui/layouts/home/not-found")).DefaultNotFound
		},
		plugins,
		adapters: config.adapters,
		data: {},
		translationsConfig: translations,
		mode,
		metaConfig: config.meta,
		siteConfig: {
			name: site?.name ?? "Fumapress",
			baseUrl: site?.baseUrl ?? getDefaultBaseUrl(),
			git: site?.git ? {
				...site.git,
				rootDir: site.git.rootDir ?? getGitRootDir() ?? process.cwd()
			} : void 0
		}
	};
	if ("loader" in config && config.loader) {
		ctx.i18nConfig ??= config.loader._i18n;
		ctx.getLoader = () => config.loader;
	} else if ("content" in config) ctx.i18nConfig ??= config.i18n;
	else {
		console.warn("[Fumapress] loader is not specified in your config, is it a mistake?");
		const emptyLoader = loader({}, { baseUrl: "/" });
		ctx.getLoader = () => emptyLoader;
	}
	for (const plugin of plugins) await plugin.init?.call(ctx);
	if ("content" in config) {
		let loaderOptions = {
			baseUrl: "/",
			i18n: ctx.i18nConfig,
			...config.loaderOptions
		};
		for (const plugin of plugins) {
			if (!plugin.configureLoader) continue;
			loaderOptions = await plugin.configureLoader.call(ctx, loaderOptions);
		}
		const source = dynamicLoader(config.content, loaderOptions);
		ctx.revalidateLoader = source.revalidate.bind(source);
		ctx.invalidateLoader = source.invalidate.bind(source);
		ctx.getLoader = () => {
			if (config.loaderOptions?.alwaysRevalidate) source.invalidate();
			return source.get();
		};
	}
	return ctx;
}
function getDefaultBaseUrl() {
	console.warn("[Fumapress] It is recommended to specify \"site.baseUrl\" in your config for better SEO.");
	if (import.meta.env.DEV) return "http://localhost:3000";
}
function renderRootMeta(context) {
	return context.metaConfig?.root?.call(context);
}
function renderPageMeta(page, context) {
	return /* @__PURE__ */ jsxs(Fragment, { children: [
		/* @__PURE__ */ jsx("title", { children: page.data.title }),
		/* @__PURE__ */ jsx("meta", {
			property: "og:title",
			content: page.data.title
		}),
		page.data.description && /* @__PURE__ */ jsx("meta", {
			property: "og:description",
			content: page.data.description
		}),
		context.metaConfig?.page?.call(context, page),
		context.data["core:page-meta"]?.map((hook, i) => /* @__PURE__ */ jsx(Fragment$1, { children: hook(page) }, i))
	] });
}
function getGitHubFileUrl(ctx, absolutePath) {
	const { git } = ctx.siteConfig;
	if (!git) return;
	const p = path.relative(git.rootDir, absolutePath).replaceAll(path.sep, "/");
	if (p.startsWith("../")) return;
	return `https://github.com/${git.user}/${git.repo}/blob/${git.branch}/${p}`;
}
function baseLayoutProps(ctx) {
	const { name, git } = ctx.siteConfig;
	return {
		githubUrl: git ? `https://github.com/${git.user}/${git.repo}` : void 0,
		nav: { title: name }
	};
}
function createTransformChildren(Component) {
	return function({ props, children }) {
		if (props.children) for (const transformer of props.children) children = transformer(children);
		return /* @__PURE__ */ jsx(Component, {
			...props,
			children
		});
	};
}
async function renderBody(ctx, page, errorMessage) {
	for (const adapter of ctx.adapters) {
		const body = await adapter["core:render-body"]?.call(ctx, page);
		if (body !== void 0) return body;
	}
	throw new Error(errorMessage);
}
async function renderToc(ctx, page) {
	for (const adapter of ctx.adapters) {
		const toc = await adapter["core:render-toc"]?.call(ctx, page);
		if (toc !== void 0) return toc;
	}
}
async function getCreationDate(ctx, page) {
	for (const adapter of ctx.adapters) {
		const date = await adapter["core:get-creation-date"]?.call(ctx, page);
		if (date !== void 0) return date;
	}
}
async function getLastModifiedDate(ctx, page) {
	for (const adapter of ctx.adapters) {
		const date = await adapter["core:get-modified-date"]?.call(ctx, page);
		if (date !== void 0) return date;
	}
}
const mergeLayoutConfigs = (0, import_deepmerge.default)({
	all: true,
	onlyDefinedProperties: true,
	isMergeableObject(value) {
		if (isValidElement(value)) return false;
		return import_deepmerge.default.isMergeableObject(value);
	}
});
//#endregion
export { appContext, baseLayoutProps, createTransformChildren, getCreationDate, getGitHubFileUrl, getLastModifiedDate, getPressContext, initApp, mergeLayoutConfigs, renderBody, renderPageMeta, renderRootMeta, renderToc };
