// fumapress/dist/lib/vitefu.js
import fsSync from "node:fs";
import path from "node:path";
import fs from "node:fs/promises";
import { createRequire } from "node:module";
//#region src/lib/vitefu.ts
let pnp;
let pnpWorkspaceLocators = [];
if (process.versions.pnp) try {
	pnp = createRequire(import.meta.url)("pnpapi");
	pnpWorkspaceLocators = pnp.getDependencyTreeRoots();
} catch {}
async function crawlFrameworkPkgs(options) {
	const pkgJsonPath = await findClosestPkgJsonPath(options.root);
	if (!pkgJsonPath) return {
		optimizeDeps: {
			include: [],
			exclude: []
		},
		ssr: {
			noExternal: [],
			external: []
		}
	};
	const pkgJson = await readJson(pkgJsonPath).catch((e) => {
		throw new Error(`Unable to read ${pkgJsonPath}`, { cause: e });
	});
	const optimizeDepsIncludeByPkgJsonPath = /* @__PURE__ */ new Map();
	const optimizeDepsSubpathsByPkgJsonPath = /* @__PURE__ */ new Map();
	let optimizeDepsInclude = [];
	let optimizeDepsExclude = [];
	let ssrNoExternal = [];
	let ssrExternal = [];
	await crawl(pkgJsonPath, pkgJson);
	optimizeDepsInclude = [...optimizeDepsIncludeByPkgJsonPath.entries()].flatMap(([depPkgJsonPath, chain]) => (optimizeDepsSubpathsByPkgJsonPath.get(depPkgJsonPath) ?? ["."]).map((subpath) => subpath === "." ? chain : chain + subpath.slice(1)));
	if (options.viteUserConfig) {
		const userOptimizeDepsExclude = options.viteUserConfig.optimizeDeps?.exclude;
		if (userOptimizeDepsExclude) optimizeDepsInclude = optimizeDepsInclude.filter((dep) => !isDepExcluded(dep, userOptimizeDepsExclude));
		const userOptimizeDepsInclude = options.viteUserConfig.optimizeDeps?.include;
		if (userOptimizeDepsInclude) optimizeDepsExclude = optimizeDepsExclude.filter((dep) => !isDepIncluded(dep, userOptimizeDepsInclude));
		const userSsrExternal = options.viteUserConfig.ssr?.external;
		if (userSsrExternal) ssrNoExternal = ssrNoExternal.filter((dep) => !isDepExternaled(dep, userSsrExternal));
		const userSsrNoExternal = options.viteUserConfig.ssr?.noExternal;
		if (userSsrNoExternal) ssrExternal = ssrExternal.filter((dep) => !isDepNoExternaled(dep, userSsrNoExternal));
	}
	optimizeDepsInclude.sort();
	optimizeDepsExclude.sort();
	ssrExternal.sort();
	ssrNoExternal.sort();
	return {
		optimizeDeps: {
			include: optimizeDepsInclude,
			exclude: optimizeDepsExclude
		},
		ssr: {
			noExternal: ssrNoExternal,
			external: ssrExternal
		}
	};
	async function crawl(currentPkgJsonPath, currentPkgJson, parentDepNames = [], parentIsFrameworkPkg = false, hasFrameworkAncestor = false) {
		const crawlDevDependencies = parentDepNames.length === 0 || isPrivateWorkspacePackage(currentPkgJsonPath, currentPkgJson, options.workspaceRoot);
		const deps = [...Object.keys(currentPkgJson.dependencies ?? {}), ...crawlDevDependencies ? Object.keys(currentPkgJson.devDependencies ?? {}) : []].filter((dep) => !parentDepNames.includes(dep));
		await Promise.all(deps.map(async (dep) => {
			const frameworkByName = options.isFrameworkPkgByName?.(dep);
			const semiFrameworkByName = options.isSemiFrameworkPkgByName?.(dep);
			if (frameworkByName === false || semiFrameworkByName === false) return;
			const depPkgJsonPath = await findDepPkgJsonPath(dep, currentPkgJsonPath, !!options.workspaceRoot);
			if (!depPkgJsonPath) return;
			const depPkgJson = await readJson(depPkgJsonPath).catch(() => {});
			if (!depPkgJson) return;
			const isFrameworkPkg = frameworkByName === true || options.isFrameworkPkgByJson?.(depPkgJson) === true;
			const isSemiFrameworkPkg = semiFrameworkByName === true || options.isSemiFrameworkPkgByJson?.(depPkgJson) === true;
			const depChain = parentDepNames.concat(dep);
			if (isFrameworkPkg || isSemiFrameworkPkg) {
				if (isFrameworkPkg) {
					pushUnique(optimizeDepsExclude, dep);
					pushUnique(ssrNoExternal, dep);
				} else pushUnique(ssrNoExternal, dep);
				await crawl(depPkgJsonPath, depPkgJson, depChain, true, true);
				return;
			}
			if (!hasFrameworkAncestor) return;
			if (await pkgNeedsOptimization(depPkgJson, depPkgJsonPath)) addOptimizedDep(depPkgJsonPath, depChain, depPkgJson);
			else await crawl(depPkgJsonPath, depPkgJson, depChain, false, true);
			if (!options.isBuild && parentIsFrameworkPkg) pushUnique(ssrExternal, dep);
		}));
	}
	function addOptimizedDep(depPkgJsonPath, depChain, depPkgJson) {
		const includePath = depChain.join(" > ");
		const current = optimizeDepsIncludeByPkgJsonPath.get(depPkgJsonPath);
		if (!current || compareDepChains(includePath, current) < 0) optimizeDepsIncludeByPkgJsonPath.set(depPkgJsonPath, includePath);
		if (!optimizeDepsSubpathsByPkgJsonPath.has(depPkgJsonPath)) optimizeDepsSubpathsByPkgJsonPath.set(depPkgJsonPath, getExportsSubpaths(depPkgJson.exports));
	}
}
async function findClosestPkgJsonPath(dir, predicate) {
	let currentDir = dir.endsWith("package.json") ? path.dirname(dir) : dir;
	while (currentDir) {
		const pkg = path.join(currentDir, "package.json");
		try {
			if ((await fs.stat(pkg)).isFile() && (!predicate || await predicate(pkg))) return pkg;
		} catch {}
		const nextDir = path.dirname(currentDir);
		if (nextDir === currentDir) break;
		currentDir = nextDir;
	}
}
async function pkgNeedsOptimization(pkgJson, pkgJsonPath) {
	if (pkgJson.module || pkgJson.type === "module") return false;
	if (pkgJson.exports) return !exportsHasEsmEntry(pkgJson.exports);
	if (pkgJson.main) {
		const entryExt = path.extname(pkgJson.main);
		return !entryExt || entryExt === ".js" || entryExt === ".cjs";
	}
	try {
		await fs.access(path.join(path.dirname(pkgJsonPath), "index.js"));
		return true;
	} catch {
		return false;
	}
}
function exportsHasEsmEntry(exportsField) {
	if (typeof exportsField === "string") return exportsField.endsWith(".mjs");
	if (Array.isArray(exportsField)) return exportsField.some(exportsHasEsmEntry);
	if (exportsField && typeof exportsField === "object") return Object.entries(exportsField).some(([key, value]) => key === "import" || key === "module" || exportsHasEsmEntry(value));
	return false;
}
function getExportsSubpaths(exportsField) {
	if (!exportsField || typeof exportsField !== "object" || Array.isArray(exportsField)) return ["."];
	const subpathKeys = Object.keys(exportsField).filter((key) => key.startsWith("."));
	if (subpathKeys.length === 0) return ["."];
	const subpaths = subpathKeys.filter((key) => key === "." || /^\.\/[^*.]+$/.test(key));
	return subpaths.length > 0 ? subpaths : ["."];
}
async function findDepPkgJsonPath(dep, parent, usePnpWorkspaceLocators) {
	if (pnp) {
		if (usePnpWorkspaceLocators) try {
			const locator = pnpWorkspaceLocators.find((root) => root.name === dep);
			if (locator) {
				const pkgPath = pnp.getPackageInformation(locator).packageLocation;
				return path.resolve(pkgPath, "package.json");
			}
		} catch {}
		try {
			const depRoot = pnp.resolveToUnqualified(dep, parent);
			if (!depRoot) return;
			return path.join(depRoot, "package.json");
		} catch {
			return;
		}
	}
	let root = parent;
	while (root) {
		const pkg = path.join(root, "node_modules", dep, "package.json");
		try {
			await fs.access(pkg);
			return fsSync.realpathSync(pkg);
		} catch {}
		const nextRoot = path.dirname(root);
		if (nextRoot === root) break;
		root = nextRoot;
	}
}
async function readJson(pkgJsonPath) {
	return JSON.parse(await fs.readFile(pkgJsonPath, "utf8"));
}
function isPrivateWorkspacePackage(pkgJsonPath, pkgJson, workspaceRoot) {
	return !!(workspaceRoot && pkgJson.private && !pkgJsonPath.match(/[/\\]node_modules[/\\]/) && !path.relative(workspaceRoot, pkgJsonPath).startsWith(".."));
}
function pushUnique(array, value) {
	if (!array.includes(value)) array.push(value);
}
function compareDepChains(left, right) {
	const leftDepth = left.split(" > ").length;
	const rightDepth = right.split(" > ").length;
	if (leftDepth !== rightDepth) return leftDepth - rightDepth;
	return left.localeCompare(right);
}
function isDepIncluded(dep, optimizeDepsInclude) {
	return optimizeDepsInclude.some((include) => dep === include);
}
function isDepExcluded(dep, optimizeDepsExclude) {
	return optimizeDepsExclude.some((exclude) => dep === exclude || dep.startsWith(`${exclude} > `));
}
function isDepNoExternaled(dep, ssrNoExternal) {
	if (typeof ssrNoExternal === "boolean") return ssrNoExternal;
	return (Array.isArray(ssrNoExternal) ? ssrNoExternal : [ssrNoExternal]).some((noExternal) => {
		if (typeof noExternal === "string") return dep === noExternal;
		return noExternal.test(dep);
	});
}
function isDepExternaled(dep, ssrExternal) {
	if (typeof ssrExternal === "boolean") return ssrExternal;
	return ssrExternal.some((external) => dep === external);
}
//#endregion
export { crawlFrameworkPkgs, findClosestPkgJsonPath, pkgNeedsOptimization };
