// waku/dist/lib/utils/react-debug-channel.js
// This file should not include Node specific code.
import { base64ToBytes, bytesToBase64 } from './base64-web.js';
export const DEBUG_ID_HEADER = 'X-Waku-Debug-Id';
export const DEBUG_CMD_EVENT = 'waku:debug-cmd';
export const DEBUG_DATA_EVENT = 'waku:debug-data';
export function assertIsDebugEventPayload(payload) {
    if (!payload || typeof payload !== 'object' || typeof payload.i !== 'string' || 'b' in payload && typeof payload.b !== 'string' || 'd' in payload && payload.d !== true) {
        throw new Error('Invalid debug event payload');
    }
}
const createWsDebugChannel = (debugId)=>{
    const hot = import.meta.hot;
    let closed = false;
    let onDataEvent;
    const cleanup = (notify)=>{
        if (closed) {
            return;
        }
        closed = true;
        if (onDataEvent) {
            hot.off(DEBUG_DATA_EVENT, onDataEvent);
        }
        if (notify) {
            hot.send(DEBUG_CMD_EVENT, {
                i: debugId,
                d: true
            });
        }
    };
    const readable = new ReadableStream({
        start (controller) {
            onDataEvent = (payload)=>{
                assertIsDebugEventPayload(payload);
                if (closed || payload.i !== debugId) {
                    return;
                }
                if ('b' in payload) {
                    // chunk
                    controller.enqueue(base64ToBytes(payload.b));
                }
                if ('d' in payload) {
                    // done
                    cleanup();
                    controller.close();
                }
            };
            hot.on(DEBUG_DATA_EVENT, onDataEvent);
            hot.send(DEBUG_CMD_EVENT, {
                i: debugId
            });
        },
        cancel () {
            cleanup(true);
        }
    });
    const writable = new WritableStream({
        write (chunk) {
            if (closed) {
                throw new Error('Channel is closed');
            }
            hot.send(DEBUG_CMD_EVENT, {
                i: debugId,
                b: bytesToBase64(chunk)
            });
        },
        close () {
            cleanup(true);
        },
        abort () {
            cleanup(true);
        }
    });
    return {
        readable,
        writable
    };
};
export const setupDebugChannel = (baseFetchFn, prefetched, debugId)=>{
    if (prefetched) {
        if (debugId) {
            const debugChannel = createWsDebugChannel(debugId);
            return {
                debugChannel
            };
        }
        return {};
    }
    const newDebugId = crypto.randomUUID();
    const debugChannel = createWsDebugChannel(newDebugId);
    const fetchFn = (input, init)=>{
        const headers = new Headers(init?.headers);
        headers.set(DEBUG_ID_HEADER, newDebugId);
        return baseFetchFn(input, {
            ...init,
            headers
        });
    };
    return {
        fetchFn,
        debugChannel
    };
};
