// waku/dist/lib/utils/render.js
import { sanitizeLog } from './log.js';
const validateRscElementIds = (elements)=>{
    for (const id of Object.keys(elements)){
        if (id.startsWith('_')) {
            throw new Error(`RSC element IDs starting with "_" are reserved for Waku internals: ${id}`);
        }
    }
};
export function createRenderUtils(temporaryReferences, renderToReadableStream, createFromReadableStream, loadSsrEntryModule, buildId, debugChannel, debugId) {
    const onError = (e)=>{
        if (e && typeof e === 'object' && 'digest' in e && typeof e.digest === 'string') {
            return e.digest;
        }
        console.error('Error during rendering:', sanitizeLog(e));
    };
    return {
        async renderRsc (elements, options) {
            validateRscElementIds(elements);
            const data = buildId ? {
                ...elements,
                _buildId: buildId
            } : {
                ...elements
            };
            if (options && 'value' in options) {
                data._value = options.value;
            }
            return renderToReadableStream(data, {
                temporaryReferences,
                onError,
                debugChannel
            }, {
                onClientReference (metadata) {
                    options?.unstable_clientModuleCallback?.(metadata.deps.js);
                }
            });
        },
        async parseRsc (stream) {
            return createFromReadableStream(stream, {});
        },
        async renderHtml (elementsStream, html, options) {
            const { INTERNAL_renderHtmlStream: renderHtmlStream } = await loadSsrEntryModule();
            const rscHtmlStream = renderToReadableStream(html, {
                onError
            });
            const htmlResult = await renderHtmlStream(elementsStream, rscHtmlStream, {
                rscPath: options.rscPath,
                formState: options.formState,
                nonce: options.nonce,
                extraScriptContent: options.unstable_extraScriptContent,
                debugId
            });
            return new Response(htmlResult.stream, {
                status: htmlResult.status || options.status || 200,
                headers: {
                    'content-type': 'text/html; charset=utf-8'
                }
            });
        }
    };
}
