// waku/dist/lib/utils/request.js
import { decodeFuncId, decodeRscPath } from '../utils/rsc-path.js';
import { createCustomError } from './custom-errors.js';
import { removeBase } from './path.js';
export async function getInput(req, config, temporaryReferences, decodeReply, decodeAction, decodeFormState, loadServerAction) {
    const url = new URL(req.url);
    const pathname = removeBase(url.pathname, config.basePath);
    const rscPathPrefix = '/' + config.rscBase + '/';
    let rscPath;
    let input;
    if (pathname.startsWith(rscPathPrefix)) {
        rscPath = decodeRscPath(pathname.slice(rscPathPrefix.length));
        // server action: js
        const actionId = decodeFuncId(rscPath);
        if (actionId) {
            validateServerActionRequest(req);
            const body = await getActionBody(req);
            const args = await decodeReply(body, {
                temporaryReferences
            });
            const action = await loadServerAction(actionId);
            input = {
                type: 'function',
                fn: action,
                args,
                pathname,
                req
            };
        } else {
            // client RSC request
            let rscParams = url.searchParams;
            if (req.body) {
                validateServerActionRequest(req);
                const body = await getActionBody(req);
                rscParams = await decodeReply(body, {
                    temporaryReferences
                });
            }
            input = {
                type: 'component',
                rscPath,
                rscParams,
                pathname,
                req
            };
        }
    } else if (req.method === 'POST') {
        const contentType = req.headers.get('content-type');
        if (typeof contentType === 'string' && contentType.startsWith('multipart/form-data')) {
            // server action: no js (progressive enhancement)
            validateServerActionRequest(req);
            input = {
                type: 'action',
                fn: async ()=>{
                    const formData = await getActionBody(req);
                    const decodedAction = await decodeAction(formData);
                    const result = await decodedAction();
                    return await decodeFormState(result, formData);
                },
                pathname,
                req
            };
        } else {
            // POST API request
            input = {
                type: 'custom',
                pathname,
                req
            };
        }
    } else {
        // SSR
        input = {
            type: 'custom',
            pathname,
            req
        };
    }
    return input;
}
function validateServerActionRequest(req) {
    if (req.method !== 'POST') {
        throw createCustomError('Method Not Allowed', {
            status: 405
        });
    }
    const origin = req.headers.get('origin');
    if (origin) {
        if (origin === 'null') {
            throw createCustomError('Forbidden', {
                status: 403
            });
        }
        const requestOrigin = new URL(req.url).origin;
        let originUrl;
        try {
            originUrl = new URL(origin);
        } catch  {
            throw createCustomError('Forbidden', {
                status: 403
            });
        }
        if (originUrl.origin !== requestOrigin) {
            throw createCustomError('Forbidden', {
                status: 403
            });
        }
    } else if (req.headers.get('sec-fetch-site') === 'cross-site') {
        throw createCustomError('Forbidden', {
            status: 403
        });
    }
}
async function getActionBody(req) {
    if (!req.body) {
        throw new Error('missing request body for server function');
    }
    const contentType = req.headers.get('content-type');
    if (contentType?.startsWith('multipart/form-data')) {
        return req.formData();
    } else {
        return req.text();
    }
}
