// fumapress/dist/plugins/image/cloudflare.client.js
"use client";
import { ImageProvider } from "../../components/image.js";
import { joinPathname } from "../../lib/pathname.js";
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
//#region src/plugins/image/cloudflare.client.tsx
function createProvider(config) {
	return {
		name: "cloudflare",
		defaultQuality: config.defaultQuality,
		deviceSizes: config.sizes,
		sizes: config.sizes,
		buildImageUrl({ src, width, quality }) {
			const parts = [
				`width=${width}`,
				`quality=${quality}`,
				`format=${config.format}`
			];
			if (config.fit) parts.push(`fit=${config.fit}`);
			return joinPathname(config.path, parts.join(","), src);
		},
		canOptimize(src) {
			if (import.meta.env.DEV) return false;
			if (!config.dangerouslyAllowSVG && src.split("?", 1)[0].endsWith(".svg")) return false;
			return true;
		}
	};
}
function CloudflareImageProvider({ config, children }) {
	return /* @__PURE__ */ jsx(ImageProvider, {
		provider: useMemo(() => createProvider(config), [config]),
		children
	});
}
//#endregion
export { CloudflareImageProvider, createProvider };
