// fumapress/dist/plugins/image/self-hosted.client.js
"use client";
import { ImageProvider } from "../../components/image.js";
import { resolveBaseUrl } from "../../lib/pathname.js";
import { validateImageSrc } from "./self-hosted.utils.js";
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
//#region src/plugins/image/self-hosted.client.tsx
function createProvider(config) {
	return {
		name: "self-hosted",
		defaultQuality: config.quality,
		deviceSizes: config.deviceSizes,
		sizes: [...config.imageSizes, ...config.deviceSizes].sort((a, b) => a - b),
		buildImageUrl({ src, width, quality }) {
			const params = new URLSearchParams({
				src,
				width: String(width),
				quality: String(quality)
			});
			return `${resolveBaseUrl(import.meta.env.BASE_URL, config.path)}?${params.toString()}`;
		},
		validate(src) {
			if (import.meta.env.DEV) {
				const validation = validateImageSrc(config, src);
				if (!validation.allowed) throw new Error(`[Fumapress] Image src "${src}" is not allowed: ${validation.reason}`);
			}
		},
		canOptimize(src) {
			if (src.split("?", 1)[0].endsWith(".svg")) return false;
			return true;
		}
	};
}
function SelfHostedImageProvider({ config, children }) {
	return /* @__PURE__ */ jsx(ImageProvider, {
		provider: useMemo(() => createProvider(config), [config]),
		children
	});
}
//#endregion
export { SelfHostedImageProvider, createProvider };
