// fumapress/dist/plugins/image/self-hosted.js
import { ImageOptimizationCache, createImageOptimizer, resolveImageConfig } from "./self-hosted.utils.js";
import { SelfHostedImageProvider } from "./self-hosted.client.js";
import { jsx } from "react/jsx-runtime";
//#region src/plugins/image/self-hosted.tsx
function imagePlugin(_options) {
	const config = resolveImageConfig(_options);
	return {
		name: "image:self-hosted",
		init() {
			(this.data["core:provider"] ??= []).push((props) => {
				props.children = /* @__PURE__ */ jsx(SelfHostedImageProvider, {
					config,
					children: props.children
				});
				return props;
			});
		},
		createPages({ createApi }) {
			if (this.mode === "static") throw new Error("[Fumapress] Image Optimization is not compatible with static mode, please disable it");
			const optimizer = createImageOptimizer(config, import.meta.env.DEV ? void 0 : new ImageOptimizationCache());
			createApi({
				render: "dynamic",
				path: config.path,
				handlers: { GET: optimizer }
			});
		}
	};
}
//#endregion
export { imagePlugin };
