// fumadocs-mdx/dist/config/index.js
import { r as applyMdxPreset, t as remarkInclude } from "../remark-include-B6rmf4sh.js";
import { metaSchema as metaSchema$1, pageSchema } from "fumadocs-core/source/schema";
//#region src/config/define.ts
function defineCollections(options) {
	return options;
}
function defineDocs(options) {
	const dir = options.dir ?? "content/docs";
	return {
		type: "docs",
		dir,
		docs: defineCollections({
			type: "doc",
			dir,
			schema: pageSchema,
			...options?.docs
		}),
		meta: defineCollections({
			type: "meta",
			dir,
			schema: metaSchema$1,
			...options?.meta
		})
	};
}
function defineConfig(config = {}) {
	return config;
}
function defineSatteriConfig(config = {}) {
	return config;
}
function defineSatteriCollections(options) {
	return options;
}
//#endregion
//#region src/config/index.ts
/** @deprecated import `pageSchema` from `fumadocs-core/source/schema` instead (since 16.2.3) */
const frontmatterSchema = pageSchema;
/** @deprecated import from `fumadocs-core/source/schema` instead (since 16.2.3) */
const metaSchema = metaSchema$1;
//#endregion
export { applyMdxPreset, defineCollections, defineConfig, defineDocs, defineSatteriCollections, defineSatteriConfig, frontmatterSchema, metaSchema, remarkInclude };
