// fumapress/dist/plugins/asyncapi.js
import { jsx } from "react/jsx-runtime";
import { asyncapiTranslations } from "@fumadocs/asyncapi/i18n";
//#region src/plugins/asyncapi.tsx
/**
* this will register the OpenAPI adapter & required layout configs.
*/
function asyncapiPlugin(options) {
	const { server, disableLoaderPlugin = false } = options;
	function initRenderers(data) {
		(data.renderers ??= []).push(function(data) {
			if (isAsyncAPI(this.page.data)) data.pageProps.full ??= true;
			return data;
		});
	}
	return {
		name: "core:openapi",
		init() {
			this.adapters.push(adapter(options));
			initRenderers(this.data["core:docs-layout"] ??= {});
			initRenderers(this.data["core:notebook-layout"] ??= {});
			if (this.translationsConfig) this.translationsConfig.extend(asyncapiTranslations());
		},
		configureLoader(options) {
			if (!disableLoaderPlugin) {
				options.plugins ??= [];
				options.plugins.push(server.loaderPlugin());
			}
			return options;
		}
	};
}
function adapter(options) {
	return {
		async "core:render-body"(page) {
			if (isAsyncAPI(page.data)) return /* @__PURE__ */ jsx(options.ClientAPIPage ?? (await import("../components/asyncapi.js")).default, { ...page.data.getAsyncAPIPageProps() });
		},
		"core:render-toc"(page) {
			if (isAsyncAPI(page.data)) return page.data.toc;
		},
		"core:get-structured-data"(page) {
			if (isAsyncAPI(page.data)) return page.data.structuredData;
		}
	};
}
function isAsyncAPI(data) {
	return "getAsyncAPIPageProps" in data && typeof data.getAsyncAPIPageProps === "function";
}
//#endregion
export { asyncapiPlugin };
