// fumapress/dist/plugins/blog.js
import { groupTags, groupTagsI18n } from "../lib/shared/blog.js";
import { joinPathname } from "../lib/pathname.js";
import { createBlogLayout, createBlogLayoutPage } from "../layouts/blog.js";
import { createBlogTagPage, createBlogTagsPage } from "../layouts/blog.tags.js";
import { createBlogIndexPage } from "../layouts/blog.index.js";
import { jsx } from "react/jsx-runtime";
import { AsyncLocalStorage } from "node:async_hooks";
//#region src/plugins/blog.tsx
const blogContext = new AsyncLocalStorage({ name: "fumapress:blog" });
function getBlogContext() {
	const store = blogContext.getStore();
	if (!store) throw new Error("[Fumapress] Missing blog context for Fumapress, make sure the blog plugin is configured");
	return store;
}
function blogPlugin({ paths = {}, isBlog = (page) => page.type === "blog", layouts = {} } = {}) {
	const blogCtx = {
		indexPath: paths.index ?? "/blog",
		tagsPath: paths.tags ?? "/blog/tags",
		isBlog
	};
	const Layout = layouts.layout ?? createBlogLayout();
	const Page = layouts.page ?? createBlogLayoutPage();
	return {
		name: "core:blog",
		renderPage({ page, lang, slugs }) {
			if (!isBlog.call(this, page)) return;
			return /* @__PURE__ */ jsx(Layout, {
				lang,
				children: /* @__PURE__ */ jsx(Page, {
					lang,
					slugs,
					page
				})
			});
		},
		async createPages({ createPage, createLayout, createInterceptor }) {
			const renderMode = this.mode === "default" ? "static" : this.mode;
			const blogPages = (await this.getLoader()).getPages().filter(isBlog.bind(this));
			createInterceptor((next) => blogContext.run(blogCtx, next));
			createLayout({
				render: renderMode,
				path: this.i18nConfig ? "/[lang]/(blog)" : "/(blog)",
				component: Layout
			});
			if (blogCtx.indexPath !== false) {
				const IndexPage = layouts.index ?? createBlogIndexPage();
				createPage({
					render: renderMode,
					path: this.i18nConfig ? joinPathname("/[lang]/(blog)", blogCtx.indexPath) : joinPathname("/(blog)", blogCtx.indexPath),
					staticPaths: this.i18nConfig ? this.i18nConfig.languages : [],
					component: IndexPage
				});
			}
			if (this.i18nConfig && blogCtx.tagsPath !== false) {
				const TagsPage = layouts.tags ?? createBlogTagsPage();
				const TagPage = layouts.tag ?? createBlogTagPage();
				createPage({
					path: joinPathname("/[lang]/(blog)", blogCtx.tagsPath),
					render: renderMode,
					staticPaths: this.i18nConfig.languages,
					component: TagsPage
				});
				const groupedTags = await groupTagsI18n(this, blogPages);
				const staticPaths = [];
				for (const [locale, tags] of groupedTags) for (const tag of tags.keys()) staticPaths.push([locale, tag]);
				createPage({
					path: joinPathname("/[lang]/(blog)", blogCtx.tagsPath, "[tag]"),
					render: renderMode,
					staticPaths,
					component: TagPage
				});
			} else if (blogCtx.tagsPath !== false) {
				const TagsPage = layouts.tags ?? createBlogTagsPage();
				const TagPage = layouts.tag ?? createBlogTagPage();
				createPage({
					path: joinPathname("/(blog)", blogCtx.tagsPath),
					render: renderMode,
					staticPaths: [],
					component: TagsPage
				});
				const grouped = await groupTags(this, blogPages);
				createPage({
					path: joinPathname("/(blog)", blogCtx.tagsPath, "[tag]"),
					render: renderMode,
					staticPaths: Array.from(grouped.keys()),
					component: TagPage
				});
			}
		}
	};
}
//#endregion
export { blogPlugin, getBlogContext };
