// fumapress/dist/plugins/flexsearch.js
//#region src/plugins/flexsearch.ts
function flexsearchPlugin({ buildIndex = async function buildIndexDefault(page) {
	for (const adapter of this.adapters) {
		const structuredData = await adapter["core:get-structured-data"]?.call(this, page);
		if (structuredData !== void 0) return {
			id: page.url,
			title: page.data.title ?? page.path,
			description: page.data.description,
			url: page.url,
			structuredData
		};
	}
	throw new Error("[Fumapress] Please specify the `buildIndex` option to flexsearchPlugin()");
} } = {}) {
	return {
		name: "core:flexsearch",
		init() {
			(this.data["core:provider"] ??= []).push(async (props) => {
				props.search ??= { enabled: true };
				if (this.mode === "static") props.search.SearchDialog ??= (await import("../components/flexsearch-static.js")).default;
				return props;
			});
		},
		async createPages({ createApiIsomorphic }) {
			const { flexsearchFromSource } = await import("fumadocs-core/search/flexsearch");
			const render = this.mode === "default" ? "dynamic" : this.mode;
			const server = flexsearchFromSource(this.getLoader, { buildIndex: buildIndex.bind(this) });
			createApiIsomorphic({
				render,
				path: "/api/search",
				handler: render === "static" ? server.staticGET : server.GET
			});
		}
	};
}
//#endregion
export { flexsearchPlugin };
