// fumapress/dist/plugins/llms.txt.js
import { joinPathname } from "../lib/pathname.js";
import { unstable_notFound } from "waku/router/server";
import { llms } from "fumadocs-core/source/llms";
import { isMarkdownPreferred } from "fumadocs-core/negotiation";
//#region src/plugins/llms.txt.ts
function llmsPlugin(options = {}) {
	let basePath = "/";
	const { autoRedirect = true, getLLMText: _getLLMText = async function getLLMTextDefault(page) {
		for (const adapter of this.adapters) {
			const txt = await adapter["core:get-text"]?.call(this, page);
			if (txt !== void 0) return `# ${page.data.title} (${page.url})\n\n${txt}`;
		}
	} } = options;
	function markdownPathToSlugs(segs) {
		if (segs.length === 0) return segs;
		const slugs = [...segs];
		slugs[slugs.length - 1] = slugs[slugs.length - 1].replace(/\.md$/, "");
		if (slugs.length === 1 && slugs[0] === "index") slugs.pop();
		return slugs;
	}
	function slugsToMarkdownPath(slugs, lang) {
		const segments = [...slugs];
		if (segments.length === 0) segments.push("index.md");
		else segments[segments.length - 1] += ".md";
		return {
			staticPath: lang ? [lang, ...segments] : segments,
			pathname: joinPathname(lang ?? "", basePath, ...segments)
		};
	}
	function initRenderers(data) {
		data.renderers ??= [];
		data.renderers.push(function(res) {
			res.markdownUrl ??= slugsToMarkdownPath(this.page.slugs, this.page.locale).pathname;
			return res;
		});
	}
	return {
		name: "core:llms.txt",
		init() {
			if (this.mode === "dynamic") basePath = "/_llms.txt";
			initRenderers(this.data["core:docs-layout"] ??= {});
			initRenderers(this.data["core:notebook-layout"] ??= {});
		},
		createMiddlewares({ app }) {
			if (this.mode === "static") return;
			const middlewares = [];
			const parsePathname = (pathname) => {
				const slugs = pathname.split("/").filter((v) => v.length > 0);
				if (this.i18nConfig) return slugs.length > 0 ? {
					lang: slugs.shift(),
					slugs
				} : void 0;
				return { slugs };
			};
			if (autoRedirect) middlewares.push(async ({ req }, next) => {
				if (req.path.endsWith(".md") || !isMarkdownPreferred(req.raw)) return next();
				const parsed = parsePathname(req.path);
				if (!parsed) return next();
				const { lang, slugs } = parsed;
				const url = new URL(slugsToMarkdownPath(slugs, lang).pathname, req.url);
				const res = await app.fetch(new Request(url));
				if (!res.ok) return next();
				return res;
			});
			if (this.mode === "dynamic") middlewares.push(async ({ req }, next) => {
				if (!req.path.endsWith(".md")) return next();
				const parsed = parsePathname(req.path);
				if (!parsed || parsed.slugs[0] === "_llms.txt") return next();
				const { lang, slugs } = parsed;
				slugs[slugs.length - 1] = slugs[slugs.length - 1].replace(/\.md$/, "");
				const url = new URL(slugsToMarkdownPath(slugs, lang).pathname, req.url);
				const res = await app.fetch(new Request(url));
				if (!res.ok) return next();
				return res;
			});
			return middlewares;
		},
		async createPages({ createApiIsomorphic }) {
			const renderMode = this.mode === "default" ? "static" : this.mode;
			const getLLMText = _getLLMText.bind(this);
			createApiIsomorphic({
				render: renderMode,
				path: "/llms.txt",
				handler: async () => {
					const source = await this.getLoader();
					return new Response(llms(source).index());
				}
			});
			createApiIsomorphic({
				render: renderMode,
				path: "/llms-full.txt",
				handler: async () => {
					const source = await this.getLoader();
					const scanned = await Promise.all(source.getPages().map(getLLMText));
					return new Response(scanned.filter((item) => item !== void 0).join("\n\n"));
				}
			});
			createApiIsomorphic({
				render: renderMode,
				path: joinPathname(this.i18nConfig ? "[lang]" : "", basePath, "[...slugs]"),
				staticPaths: (await this.getLoader()).getPages().map((page) => slugsToMarkdownPath(page.slugs, page.locale).staticPath),
				handler: async (_req, { params }) => {
					const page = (await this.getLoader()).getPage(markdownPathToSlugs(params.slugs), params.lang);
					if (!page) unstable_notFound();
					const txt = await getLLMText(page);
					return new Response(txt ?? "", { headers: { "Content-Type": "text/markdown" } });
				}
			});
		}
	};
}
//#endregion
export { llmsPlugin };
