// fumapress/dist/plugins/openapi.js
import { isFullPathname, resolveBaseUrl } from "../lib/pathname.js";
import { jsx } from "react/jsx-runtime";
import { openapiTranslations } from "fumadocs-openapi/i18n";
//#region src/plugins/openapi.tsx
/**
* this will register the OpenAPI adapter & required layout configs.
*/
function openapiPlugin(options) {
	const { server, disableLoaderPlugin = false } = options;
	function initRenderers(data) {
		(data.renderers ??= []).push(function(data) {
			if (isOpenAPI(this.page.data)) data.pageProps.full ??= true;
			return data;
		});
	}
	return {
		name: "core:openapi",
		init() {
			this.adapters.push(adapter(options));
			initRenderers(this.data["core:docs-layout"] ??= {});
			initRenderers(this.data["core:notebook-layout"] ??= {});
			if (this.translationsConfig) this.translationsConfig.extend(openapiTranslations());
		},
		configureLoader(options) {
			if (!disableLoaderPlugin) {
				options.plugins ??= [];
				options.plugins.push(server.loaderPlugin());
			}
			return options;
		},
		async createPages({ createApi }) {
			const proxyUrl = server.options.proxyUrl;
			const { createProxy = typeof proxyUrl === "string" && isFullPathname(proxyUrl) } = options;
			if (createProxy) {
				if (!proxyUrl) throw new Error(`[Fumapress] The "proxyUrl" option in createOpenAPI() is required to create proxy server`);
				if (this.mode === "static") throw new Error(`[Fumapress] static mode is not compatible with proxy server`);
				createApi({
					path: proxyUrl,
					render: "dynamic",
					handlers: { all: (typeof createProxy === "function" ? await createProxy() : server.createProxy()).handle }
				});
			}
		}
	};
}
function adapter(options) {
	return {
		async "core:render-body"(page) {
			if (isOpenAPI(page.data)) {
				const ClientAPIPage = options.ClientAPIPage ?? (await import("../components/openapi.js")).default;
				const { payload, ...props } = page.data.getOpenAPIPageProps();
				return /* @__PURE__ */ jsx(ClientAPIPage, {
					payload: {
						...payload,
						proxyUrl: payload.proxyUrl && isFullPathname(payload.proxyUrl) ? resolveBaseUrl(import.meta.env.BASE_URL, payload.proxyUrl) : payload.proxyUrl
					},
					...props
				});
			}
		},
		"core:render-toc"(page) {
			if (isOpenAPI(page.data)) return page.data.toc;
		},
		"core:get-structured-data"(page) {
			if (isOpenAPI(page.data)) return page.data.structuredData;
		}
	};
}
function isOpenAPI(data) {
	return "getOpenAPIPageProps" in data && typeof data.getOpenAPIPageProps === "function";
}
//#endregion
export { openapiPlugin };
