// fumapress/dist/plugins/orama-search.js
//#region src/plugins/orama-search.ts
function oramaSearchPlugin({ buildIndex = async function buildIndexDefault(page) {
	for (const adapter of this.adapters) {
		const structuredData = await adapter["core:get-structured-data"]?.call(this, page);
		if (structuredData !== void 0) return {
			id: page.url,
			title: page.data.title ?? page.path,
			description: page.data.description,
			url: page.url,
			structuredData
		};
	}
	throw new Error("[Fumapress] Please specify the `buildIndex` option to oramaSearchPlugin()");
} } = {}) {
	return {
		name: "core:orama-search",
		init() {
			(this.data["core:provider"] ??= []).push(async (props) => {
				props.search ??= { enabled: true };
				if (this.mode === "static") props.search.SearchDialog ??= (await import("../components/orama-search-static.js")).default;
				return props;
			});
		},
		async createPages({ createApiIsomorphic }) {
			const { createFromSource } = await import("fumadocs-core/search/server");
			const renderMode = this.mode === "default" ? "dynamic" : this.mode;
			const server = createFromSource(this.getLoader, { buildIndex: buildIndex.bind(this) });
			createApiIsomorphic({
				render: renderMode,
				path: "/api/search",
				handler: renderMode === "static" ? server.staticGET : server.GET
			});
		}
	};
}
//#endregion
export { oramaSearchPlugin };
