// fumapress/dist/plugins/sitemap.js
import { getLastModifiedDate } from "../lib/shared.js";
import { js2xml } from "xml-js";
//#region src/plugins/sitemap.ts
function formatLastmod(value) {
	return (value instanceof Date ? value : new Date(value)).toISOString();
}
function imageToElement(image) {
	const element = { "image:loc": { _text: image.loc } };
	if (image.caption) element["image:caption"] = { _text: image.caption };
	if (image.geo_location) element["image:geo_location"] = { _text: image.geo_location };
	if (image.title) element["image:title"] = { _text: image.title };
	if (image.license) element["image:license"] = { _text: image.license };
	return element;
}
function entryToUrlElement(entry) {
	const url = { loc: { _text: entry.loc } };
	if (entry.lastmod) url.lastmod = { _text: formatLastmod(entry.lastmod) };
	if (entry.changefreq) url.changefreq = { _text: entry.changefreq };
	if (entry.priority !== void 0) url.priority = { _text: String(entry.priority) };
	if (entry.alternates?.length) url["xhtml:link"] = entry.alternates.map((alternate) => ({ _attributes: {
		rel: alternate.rel,
		hreflang: alternate.hreflang,
		href: alternate.href
	} }));
	if (entry.images?.length) url["image:image"] = entry.images.map(imageToElement);
	return url;
}
function buildSitemap(entries) {
	return js2xml({
		_declaration: { _attributes: {
			version: "1.0",
			encoding: "UTF-8"
		} },
		urlset: {
			_attributes: {
				xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9",
				"xmlns:xhtml": "http://www.w3.org/1999/xhtml",
				"xmlns:image": "http://www.google.com/schemas/sitemap-image/1.1"
			},
			url: entries.map(entryToUrlElement)
		}
	}, {
		compact: true,
		spaces: 0
	});
}
function sitemapPlugin(options = {}) {
	const { path = "/sitemap.xml", getEntry: _getEntry = async function getEntryDefault(page) {
		return {
			loc: new URL(page.url, this.siteConfig.baseUrl).href,
			lastmod: await getLastModifiedDate(this, page),
			priority: .8
		};
	}, additionalEntries } = options;
	return {
		name: "core:sitemap",
		async createPages({ createApiIsomorphic, unstable_getCreated }) {
			const renderMode = this.mode === "default" ? "static" : this.mode;
			const getEntry = _getEntry.bind(this);
			createApiIsomorphic({
				render: renderMode,
				path,
				handler: async () => {
					const source = await this.getLoader();
					const entries = [];
					const pageLocs = /* @__PURE__ */ new Set();
					for (const entry of await Promise.all(source.getPages().map(getEntry))) {
						if (!entry) continue;
						pageLocs.add(entry.loc);
						entries.push(entry);
					}
					for (const route of await unstable_getCreated().unstable_getRouterConfigs()) if (route.isStatic && route.type === "route") {
						const segments = route.path.map((v) => v.name);
						if (segments.at(-1) === "404") continue;
						const loc = new URL("/" + segments.join("/"), this.siteConfig.baseUrl).href;
						if (pageLocs.has(loc)) continue;
						entries.push({
							loc,
							priority: 1
						});
					}
					if (additionalEntries) entries.push(...typeof additionalEntries === "function" ? await additionalEntries.call(this) : additionalEntries);
					return new Response(buildSitemap(entries), { headers: { "Content-Type": "application/xml" } });
				}
			});
		}
	};
}
//#endregion
export { sitemapPlugin };
