// fumapress/dist/plugins/takumi.js
import { joinPathname } from "../lib/pathname.js";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { unstable_notFound } from "waku/router/server";
import { ImageResponse } from "takumi-js/response";
//#region src/plugins/takumi.tsx
function takumiPlugin(options = {}) {
	const { width = 1200, height = 630, generate = function fn(page) {
		return { node: generateDefault({
			title: page.data.title,
			description: page.data.description,
			site: this.siteConfig.name
		}) };
	} } = options;
	let basePath;
	function slugsToImagePath(slugs, lang) {
		const segments = [...slugs];
		if (segments.length === 0) segments.push("index.webp");
		else segments[segments.length - 1] += ".webp";
		return {
			staticPath: lang ? [lang, ...segments] : segments,
			pathname: joinPathname(lang ?? "", basePath, ...segments)
		};
	}
	function imagePathToSlugs(segs) {
		if (segs.length === 0) return segs;
		const slugs = [...segs];
		slugs[slugs.length - 1] = slugs[slugs.length - 1].replace(/\.webp$/, "");
		if (slugs.length === 1 && slugs[0] === "index") slugs.pop();
		return slugs;
	}
	return {
		name: "core:takumi",
		init() {
			const renderMode = this.mode === "default" ? "static" : this.mode;
			basePath = options.basePath ?? (renderMode === "dynamic" ? "/_takumi" : "/");
			(this.data["core:page-meta"] ??= []).push((page) => {
				const pathname = slugsToImagePath(page.slugs, page.locale).pathname;
				return /* @__PURE__ */ jsxs(Fragment, { children: [
					/* @__PURE__ */ jsx("meta", {
						property: "og:image",
						content: this.siteConfig.baseUrl ? new URL(pathname, this.siteConfig.baseUrl).href : pathname
					}),
					/* @__PURE__ */ jsx("meta", {
						property: "og:image:width",
						content: `${width}`
					}),
					/* @__PURE__ */ jsx("meta", {
						property: "og:image:height",
						content: `${height}`
					}),
					/* @__PURE__ */ jsx("meta", {
						property: "twitter:card",
						content: "summary_large_image"
					})
				] });
			});
		},
		async createPages({ createApiIsomorphic }) {
			createApiIsomorphic({
				render: this.mode === "default" ? "static" : this.mode,
				path: joinPathname(this.i18nConfig ? "[lang]" : "", basePath, "[...slugs]"),
				staticPaths: (await this.getLoader()).getPages().map((page) => slugsToImagePath(page.slugs, page.locale).staticPath),
				handler: async (_, { params }) => {
					const page = (await this.getLoader()).getPage(imagePathToSlugs(params.slugs), params.lang);
					if (!page) unstable_notFound();
					const { node, options } = await generate.call(this, page);
					return new ImageResponse(node, {
						width,
						height,
						...options,
						format: "webp"
					});
				}
			});
		}
	};
}
function generateDefault({ site, title, description }) {
	const primaryColor = "rgba(255,150,255,0.3)";
	const primaryTextColor = "rgb(255,150,255)";
	return /* @__PURE__ */ jsxs("div", {
		style: {
			display: "flex",
			flexDirection: "column",
			width: "100%",
			height: "100%",
			color: "white",
			padding: "4rem",
			backgroundColor: "#0c0c0c",
			borderBottom: `18px solid ${primaryColor}`
		},
		children: [
			/* @__PURE__ */ jsx("p", {
				style: {
					fontWeight: 800,
					fontSize: "82px",
					margin: 0
				},
				children: title
			}),
			/* @__PURE__ */ jsx("p", {
				style: {
					fontSize: "52px",
					color: "rgba(240,240,240,0.8)",
					margin: 0,
					marginTop: "16px",
					paddingBottom: "28px",
					borderBottom: `10px dashed ${primaryColor}`
				},
				children: description
			}),
			/* @__PURE__ */ jsxs("div", {
				style: {
					display: "flex",
					flexDirection: "row",
					alignItems: "center",
					gap: "20px",
					marginTop: "auto",
					color: primaryTextColor
				},
				children: [/* @__PURE__ */ jsx("svg", {
					xmlns: "http://www.w3.org/2000/svg",
					width: "56",
					height: "56",
					viewBox: "0 0 24 24",
					fill: "none",
					stroke: "currentColor",
					strokeWidth: "2",
					strokeLinecap: "round",
					strokeLinejoin: "round",
					className: "lucide lucide-book-icon lucide-book",
					children: /* @__PURE__ */ jsx("circle", {
						cx: "12",
						cy: "12",
						r: "11",
						stroke: primaryTextColor,
						strokeWidth: "2"
					})
				}), site && /* @__PURE__ */ jsx("p", {
					style: {
						fontSize: "56px",
						fontWeight: 600,
						margin: 0
					},
					children: site
				})]
			})
		]
	});
}
//#endregion
export { takumiPlugin };
