// fumadocs-mdx/dist/rolldown/index.js
import { t as createMdxLoader } from "../mdx-wSuDwc6c.js";
import { i as buildConfig, n as createCore, t as _Defaults } from "../core-Cq6V99iz.js";
import { a as createIntegratedConfigLoader, r as toVite } from "../adapter-DPEpBHhE.js";
import { t as createMetaLoader } from "../meta-UmuO8lue.js";
import { r as metaLoaderGlob, t as mdxLoaderGlob } from "../loaders-CUuRGbiE.js";
import fs from "node:fs/promises";
//#region src/rolldown/index.ts
async function mdx(config, options) {
	const core = createCore({
		configPath: options?.configPath ?? _Defaults.configPath,
		outDir: options?.outDir ?? _Defaults.outDir,
		environment: "rolldown"
	});
	await core.init({ config: buildConfig(await config, process.cwd()) });
	const configLoader = createIntegratedConfigLoader(core);
	const mdxLoader = toVite(createMdxLoader(configLoader));
	const metaLoader = toVite(createMetaLoader(configLoader, { json: "json" }));
	return [
		{
			name: "fumadocs-mdx",
			load: {
				filter: { id: [metaLoaderGlob, mdxLoaderGlob] },
				async handler(id) {
					const idx = id.lastIndexOf("?");
					if (idx === -1) return null;
					return fs.readFile(id.slice(0, idx), "utf-8");
				}
			}
		},
		{
			name: "fumadocs-mdx:mdx",
			transform: {
				filter: { id: mdxLoaderGlob },
				handler: mdxLoader.transform
			}
		},
		{
			name: "fumadocs-mdx:meta",
			transform: {
				filter: { id: metaLoaderGlob },
				handler: metaLoader.transform
			}
		}
	];
}
//#endregion
export { mdx as default };
