// fumadocs-ui/dist/contexts/search.js
"use client";
import { jsx, jsxs } from "react/jsx-runtime";
import { Suspense, createContext, lazy, use, useEffect, useEffectEvent, useMemo, useState } from "react";
import { Dialog } from "@base-ui/react/dialog";
//#region src/contexts/search.tsx
/** built-in Base UI Dialog handle */
const dialogHandle = Dialog.createHandle();
const SearchContext = createContext({
	enabled: false,
	open: false,
	hotKey: [],
	setOpenSearch: () => void 0,
	dialogHandle
});
function useSearchContext() {
	return use(SearchContext);
}
function MetaOrControl() {
	const [key, setKey] = useState("⌘");
	useEffect(() => {
		if (/Windows|Linux/i.test(window.navigator.userAgent)) setKey("Ctrl");
	}, []);
	return key;
}
const DEFAULT_HOT_KEYS = [{
	key: (e) => e.metaKey || e.ctrlKey,
	display: /* @__PURE__ */ jsx(MetaOrControl, {})
}, {
	key: "k",
	display: "K"
}];
const DefaultSearchDialog = lazy(() => import("../components/dialog/search-default.js"));
function SearchProvider({ SearchDialog = DefaultSearchDialog, children, options, hotKey = DEFAULT_HOT_KEYS, links }) {
	const [isOpen, setIsOpen] = useState(false);
	const onKeyDown = useEffectEvent((e) => {
		if (hotKey.every((v) => typeof v.key === "string" ? e.key === v.key : v.key(e))) {
			setIsOpen((open) => !open);
			e.preventDefault();
		}
	});
	useEffect(() => {
		window.addEventListener("keydown", onKeyDown);
		return () => {
			window.removeEventListener("keydown", onKeyDown);
		};
	}, []);
	return /* @__PURE__ */ jsxs(SearchContext, {
		value: useMemo(() => ({
			enabled: true,
			open: isOpen,
			hotKey,
			dialogHandle,
			setOpenSearch: setIsOpen
		}), [isOpen, hotKey]),
		children: [/* @__PURE__ */ jsx(Suspense, {
			fallback: null,
			children: /* @__PURE__ */ jsx(SearchDialog, {
				open: isOpen,
				onOpenChange: setIsOpen,
				links,
				dialogHandle,
				...options
			})
		}), children]
	});
}
/**
* Show children only when search is enabled via React Context
*/
function SearchOnly({ children }) {
	if (useSearchContext().enabled) return children;
}
//#endregion
export { SearchOnly, SearchProvider, useSearchContext };
