// fumadocs-ui/dist/layouts/home/navbar.js
"use client";
import { cn } from "../../utils/cn.js";
import { navItemVariants } from "./slots/header.js";
import Link from "fumadocs-core/link";
import { jsx } from "react/jsx-runtime";
import { NavigationMenu } from "@base-ui/react/navigation-menu";
//#region src/layouts/home/navbar.tsx
function NavbarMenu({ className, children, ...props }) {
	return /* @__PURE__ */ jsx(NavigationMenu.Item, {
		className: (s) => cn("list-none", typeof className === "function" ? className(s) : className),
		...props,
		children
	});
}
function NavbarMenuContent({ className, ...props }) {
	return /* @__PURE__ */ jsx(NavigationMenu.Content, {
		...props,
		className: (s) => cn("h-full w-(--anchor-width) max-w-(--available-width) p-3", "transition-[opacity,transform,translate] duration-(--duration) ease-(--easing)", "data-starting-style:opacity-0 data-ending-style:opacity-0", "data-starting-style:data-[activation-direction=left]:-translate-x-1/2", "data-starting-style:data-[activation-direction=right]:translate-x-1/2", "data-ending-style:data-[activation-direction=left]:translate-x-1/2", "data-ending-style:data-[activation-direction=right]:-translate-x-1/2", "grid grid-cols-1 gap-2 md:grid-cols-2 lg:grid-cols-3", typeof className === "function" ? className(s) : className)
	});
}
function NavbarMenuTrigger({ className, children, ...props }) {
	return /* @__PURE__ */ jsx(NavigationMenu.Trigger, {
		...props,
		className: (s) => cn(navItemVariants(), "text-sm rounded-md", typeof className === "function" ? className(s) : className),
		children
	});
}
function NavbarMenuLink(props) {
	return /* @__PURE__ */ jsx(NavigationMenu.Link, { render: /* @__PURE__ */ jsx(Link, {
		...props,
		className: cn("flex flex-col gap-2 rounded-lg border bg-fd-card p-3 transition-colors hover:bg-fd-accent/80 hover:text-fd-accent-foreground", props.className),
		children: props.children
	}) });
}
//#endregion
export { NavbarMenu, NavbarMenuContent, NavbarMenuLink, NavbarMenuTrigger };
