// waku/dist/minimal/client.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { createContext, memo, startTransition, use, useCallback, useEffect, useState } from 'react';
import RSDWClient from 'react-server-dom-webpack/client';
import { createCustomError } from '../lib/utils/custom-errors.js';
import { consumeInitialRscEntry } from '../lib/utils/initial-rsc.js';
import { setupDebugChannel } from '../lib/utils/react-debug-channel.js';
import { encodeFuncId, encodeRscPath } from '../lib/utils/rsc-path.js';
import { CALL_SERVER_ELEMENTS_LISTENERS, ENTRY, FETCH_ENHANCERS, FETCH_RSC_INPUT_TRANSFORMERS, SET_ELEMENTS, fetchRscStore } from './client-utils/fetch-store.js';
import { addPrefetchEntry, consumePrefetchEntry, hasPrefetchEntry } from './client-utils/prefetch-cache.js';
const { createFromFetch, encodeReply, createTemporaryReferenceSet } = RSDWClient;
const DEFAULT_HTML_HEAD = [
    /*#__PURE__*/ _jsx("meta", {
        charSet: "utf-8"
    }, "charset"),
    /*#__PURE__*/ _jsx("meta", {
        name: "viewport",
        content: "width=device-width, initial-scale=1"
    }, "viewport"),
    /*#__PURE__*/ _jsx("meta", {
        name: "generator",
        content: "Waku"
    }, "generator")
];
const BASE_RSC_PATH = `${import.meta.env?.WAKU_CONFIG_BASE_PATH ?? '/'}${import.meta.env?.WAKU_CONFIG_RSC_BASE ?? 'RSC'}/`;
const checkStatus = async (responsePromise)=>{
    const response = await responsePromise;
    if (!response.ok) {
        const location = response.headers.get('location');
        const err = createCustomError(await response.text() || response.statusText, {
            status: response.status,
            ...location && {
                location
            }
        });
        throw err;
    }
    return response;
};
// TODO(daishi) do we still this?
const getCached = (c, m, k)=>(m.has(k) ? m : m.set(k, c())).get(k);
const cache1 = new WeakMap();
const mergeElementsPromise = (a, b, isEager)=>{
    const getResult = ()=>{
        if (!isEager) {
            return Promise.all([
                a,
                b
            ]).then(([a, b])=>{
                const nextElements = {
                    ...a,
                    ...b
                };
                delete nextElements._value;
                return nextElements;
            });
        }
        return Promise.resolve(a).then((aRes)=>{
            const bPromise = Promise.resolve(b);
            const base = {
                ...aRes
            };
            delete base._value;
            return new Proxy(base, {
                get (target, key) {
                    if (key !== '_value' && !isEager(key) && // a lazy key still equal to aRes's value hasn't been streamed yet
                    target[key] === aRes[key]) {
                        target[key] = bPromise.then((bRes)=>key in bRes ? bRes[key] : aRes[key]);
                    }
                    return target[key];
                }
            });
        });
    };
    const cache2 = getCached(()=>new WeakMap(), cache1, a);
    return getCached(getResult, cache2, b);
};
const getFetchFn = ()=>{
    let fetchFn = fetch;
    const enhancers = fetchRscStore[FETCH_ENHANCERS];
    if (enhancers) {
        for (const enhance of enhancers){
            fetchFn = enhance(fetchFn);
        }
    }
    return fetchFn;
};
const getSetElements = ()=>{
    const setElements = fetchRscStore[SET_ELEMENTS];
    if (!setElements) {
        throw new Error('Missing Root component');
    }
    return setElements;
};
const requestRsc = (fetchFn, rscPath, rscParams, temporaryReferences, signal)=>{
    const url = BASE_RSC_PATH + encodeRscPath(rscPath);
    const init = {};
    if (signal) {
        init.signal = signal;
    }
    if (rscParams === undefined) {
        return fetchFn(url, init);
    }
    if (rscParams instanceof URLSearchParams) {
        return fetchFn(url + '?' + rscParams, init);
    }
    return encodeReply(rscParams, {
        temporaryReferences
    }).then((body)=>fetchFn(url, {
            ...init,
            method: 'POST',
            body
        }));
};
const decodeRsc = (responsePromise, temporaryReferences, debugChannel)=>createFromFetch(checkStatus(responsePromise), {
        callServer: (funcId, args)=>unstable_callServerRsc(funcId, args),
        debugChannel,
        temporaryReferences
    });
const reloadOnBuildIdMismatch = (elements, onBuildIdMismatch)=>{
    if (!import.meta.env?.WAKU_BUILD_ID) {
        return;
    }
    Promise.resolve(elements).then((data)=>{
        if (data._buildId !== import.meta.env.WAKU_BUILD_ID) {
            (onBuildIdMismatch ?? (()=>window.location.reload()))();
        }
    }, ()=>{});
};
const prefetchRscInternal = (rscPath, rscParams)=>{
    const temporaryReferences = createTemporaryReferenceSet();
    const responsePromise = requestRsc(getFetchFn(), rscPath, rscParams, temporaryReferences, undefined);
    addPrefetchEntry(rscPath, rscParams, decodeRsc(responsePromise, temporaryReferences, undefined));
};
const fetchRscElements = (rscPath, rscParams, options)=>{
    const prefetched = consumePrefetchEntry(rscPath, rscParams);
    if (prefetched) {
        reloadOnBuildIdMismatch(prefetched, options?.onBuildIdMismatch);
        return prefetched;
    }
    const initial = consumeInitialRscEntry();
    const baseFetchFn = getFetchFn();
    const debug = import.meta.hot ? setupDebugChannel(baseFetchFn, !!initial, initial?.debugId) : undefined;
    const fetchFn = debug?.fetchFn || baseFetchFn;
    const temporaryReferences = createTemporaryReferenceSet();
    const responsePromise = initial ? initial.response : requestRsc(fetchFn, rscPath, rscParams, temporaryReferences, options?.signal);
    const elements = decodeRsc(responsePromise, temporaryReferences, debug?.debugChannel);
    reloadOnBuildIdMismatch(elements, options?.onBuildIdMismatch);
    return elements;
};
const applyInputTransformers = (rscPath, rscParams, prefetchOnly)=>{
    const fetchRscInputTransformers = fetchRscStore[FETCH_RSC_INPUT_TRANSFORMERS];
    if (fetchRscInputTransformers) {
        for (const transformFetchRscInput of fetchRscInputTransformers){
            [rscPath, rscParams, prefetchOnly] = transformFetchRscInput(rscPath, rscParams, prefetchOnly);
        }
    }
    return [
        rscPath,
        rscParams,
        prefetchOnly
    ];
};
const fetchRscInternal = (rscPath, rscParams, prefetchOnly, options)=>{
    [rscPath, rscParams, prefetchOnly] = applyInputTransformers(rscPath, rscParams, prefetchOnly);
    if (prefetchOnly) {
        if (!hasPrefetchEntry(rscPath, rscParams)) {
            prefetchRscInternal(rscPath, rscParams);
        }
        return undefined;
    }
    return fetchRscElements(rscPath, rscParams, options);
};
/**
 * callServer callback
 * This is not a public API.
 */ export const unstable_callServerRsc = async (funcId, args)=>{
    const rscPath = encodeFuncId(funcId);
    const rscParams = args.length === 1 && args[0] instanceof URLSearchParams ? args[0] : args;
    const { _value: value, ...data } = await fetchRscInternal(rscPath, rscParams, false, undefined);
    if (Object.keys(data).length) {
        const setElements = getSetElements();
        const callServerElementsListeners = fetchRscStore[CALL_SERVER_ELEMENTS_LISTENERS];
        startTransition(()=>{
            callServerElementsListeners?.forEach((listener)=>{
                listener(data);
            });
            setElements((prev)=>mergeElementsPromise(prev, data));
        });
    }
    return value;
};
/**
 * Register a listener that runs when a server action returns new elements.
 * Returns a function that unregisters the listener.
 */ export const unstable_registerCallServerElementsListener = (listener)=>{
    const callServerElementsListeners = fetchRscStore[CALL_SERVER_ELEMENTS_LISTENERS] ||= new Set();
    callServerElementsListeners.add(listener);
    return ()=>{
        callServerElementsListeners.delete(listener);
    };
};
/**
 * Register a fetch enhancer applied to every RSC request (e.g. to add headers).
 * Enhancers are composed in registration order. Returns a function that
 * unregisters the enhancer.
 */ export const unstable_registerFetchEnhancer = (enhance)=>{
    const fetchEnhancers = fetchRscStore[FETCH_ENHANCERS] ||= new Set();
    fetchEnhancers.add(enhance);
    return ()=>{
        fetchEnhancers.delete(enhance);
    };
};
/**
 * Register a transformer that rewrites the RSC fetch input
 * (`rscPath`, `rscParams`, `prefetchOnly`) before each request. Returns a
 * function that unregisters the transformer.
 */ export const unstable_registerFetchRscInputTransformer = (transformFetchRscInput)=>{
    const fetchRscInputTransformers = fetchRscStore[FETCH_RSC_INPUT_TRANSFORMERS] ||= new Set();
    fetchRscInputTransformers.add(transformFetchRscInput);
    return ()=>{
        fetchRscInputTransformers.delete(transformFetchRscInput);
    };
};
const registerHmrRefetch = (refetch)=>{
    globalThis.__WAKU_RSC_RELOAD_LISTENERS__ ||= [];
    const index = globalThis.__WAKU_RSC_RELOAD_LISTENERS__.indexOf(globalThis.__WAKU_REFETCH_RSC__);
    if (index !== -1) {
        globalThis.__WAKU_RSC_RELOAD_LISTENERS__.splice(index, 1, refetch);
    } else {
        globalThis.__WAKU_RSC_RELOAD_LISTENERS__.push(refetch);
    }
    globalThis.__WAKU_REFETCH_RSC__ = refetch;
};
/** Fetch elements for an RSC path, reusing a cached or prefetched result. */ export const unstable_fetchRsc = (rscPath, rscParams, options)=>{
    if (import.meta.hot) {
        registerHmrRefetch(()=>{
            delete fetchRscStore[ENTRY];
            const data = unstable_fetchRsc(rscPath, rscParams, options);
            getSetElements()(()=>data);
        });
    }
    const entry = fetchRscStore[ENTRY];
    if (entry && entry[0] === rscPath && entry[1] === rscParams) {
        return entry[2];
    }
    const data = fetchRscInternal(rscPath, rscParams, false, options);
    fetchRscStore[ENTRY] = [
        rscPath,
        rscParams,
        data
    ];
    return data;
};
/** Eagerly fetch and cache elements so a later fetch reuses them. */ export const unstable_prefetchRsc = (rscPath, rscParams)=>{
    fetchRscInternal(rscPath, rscParams, true, undefined);
};
const RefetchContext = /*#__PURE__*/ createContext(()=>{
    throw new Error('Missing Root component');
});
const ElementsContext = /*#__PURE__*/ createContext(null);
/**
 * Client root. Seeds the initial elements, bridges the store to React state,
 * and provides the elements to `Slot` descendants.
 *
 * This API is technically unstable and may change or be removed,
 * even though it does not carry the `unstable_` prefix.
 */ export const Root = ({ initialRscPath, initialRscParams, children })=>{
    const [elements, setElements] = useState(()=>unstable_fetchRsc(initialRscPath || '', initialRscParams));
    useEffect(()=>{
        fetchRscStore[SET_ELEMENTS] = setElements;
    }, []);
    const refetch = useCallback(async (rscPath, rscParams, options)=>{
        delete fetchRscStore[ENTRY];
        const data = unstable_fetchRsc(rscPath, rscParams, options);
        const dataWithoutErrors = Promise.resolve(data).catch(()=>({}));
        setElements((prev)=>mergeElementsPromise(prev, dataWithoutErrors, options?.unstable_isEager));
        return data;
    }, []);
    return /*#__PURE__*/ _jsx(RefetchContext, {
        value: refetch,
        children: /*#__PURE__*/ _jsxs(ElementsContext, {
            value: elements,
            children: [
                DEFAULT_HTML_HEAD,
                children
            ]
        })
    });
};
/**
 * This API is technically unstable and may change or be removed,
 * even though it does not carry the `unstable_` prefix.
 */ export const useRefetch = ()=>use(RefetchContext);
const ChildrenContext = /*#__PURE__*/ createContext(undefined);
const ChildrenContextProvider = /*#__PURE__*/ memo(ChildrenContext);
/**
 * This API is technically unstable and may change or be removed,
 * even though it does not carry the `unstable_` prefix.
 */ export const Children = ()=>use(ChildrenContext);
export const useElementsPromise_UNSTABLE = ()=>{
    const elementsPromise = use(ElementsContext);
    if (!elementsPromise) {
        throw new Error('Missing Root component');
    }
    return elementsPromise;
};
/**
 * Slot component
 * This is used under the Root component.
 * Slot id is the key of elements returned by the server.
 *
 * If the server returns this
 * ```
 *   { 'foo': <div>foo</div>, 'bar': <div>bar</div> }
 * ```
 * then you can use this component like this
 * ```
 *   <Root><Slot id="foo" /><Slot id="bar" /></Root>
 * ```
 *
 * This API is technically unstable and may change or be removed,
 * even though it does not carry the `unstable_` prefix.
 */ export const Slot = ({ id, children })=>{
    const elementsPromise = useElementsPromise_UNSTABLE();
    const elements = use(elementsPromise);
    if (id in elements && elements[id] === undefined) {
        throw new Error('Element cannot be undefined, use null instead: ' + id);
    }
    const element = elements[id];
    const isValidElement = element !== undefined;
    if (!isValidElement) {
        throw new Error('Invalid element: ' + id);
    }
    return /*#__PURE__*/ _jsx(ChildrenContextProvider, {
        value: children,
        children: element
    });
};
/**
 * ServerRoot for SSR
 * This is not a public API.
 */ export const INTERNAL_ServerRoot = ({ elementsPromise, children })=>/*#__PURE__*/ _jsx(RefetchContext, {
        value: async ()=>({}),
        children: /*#__PURE__*/ _jsxs(ElementsContext, {
            value: elementsPromise,
            children: [
                DEFAULT_HTML_HEAD,
                children
            ]
        })
    });
