// fumadocs-mdx/dist/vite/index.js
import { t as createMdxLoader } from "../mdx-wSuDwc6c.js";
import { i as buildConfig, n as createCore, t as _Defaults } from "../core-Cq6V99iz.js";
import { a as createIntegratedConfigLoader, r as toVite } from "../adapter-DPEpBHhE.js";
import { t as createMetaLoader } from "../meta-UmuO8lue.js";
import { r as metaLoaderGlob, t as mdxLoaderGlob } from "../loaders-CUuRGbiE.js";
import { t as indexFile } from "../index-file-C9Bw7He1.js";
import path from "node:path";
import { runnerImport } from "vite";
//#region src/vite/index.ts
function mdx(forcedConfig, pluginOptions = {}) {
	let root;
	let core;
	let options;
	const metaPlugin = { name: "fumadocs-mdx:meta" };
	const mdxPlugin = { name: "fumadocs-mdx:mdx" };
	return [
		{
			name: "fumadocs-mdx",
			async config(config) {
				root = config.root ?? process.cwd();
				options = applyDefaults(root, pluginOptions);
				core = createViteCore(options);
				await core.init({ config: buildConfig(forcedConfig ? await forcedConfig : (await runnerImport(options.configPath)).module, root) });
				const configLoader = createIntegratedConfigLoader(core);
				const mdxLoader = toVite(createMdxLoader(configLoader));
				const metaLoader = toVite(createMetaLoader(configLoader, { json: "json" }));
				mdxPlugin.transform = {
					filter: { id: mdxLoaderGlob },
					order: "pre",
					handler(code, id) {
						if (id.includes("virtual:vite-rsc")) return null;
						if (!forcedConfig) this.addWatchFile(options.configPath);
						return mdxLoader.transform.call(this, code, id);
					}
				};
				metaPlugin.transform = {
					filter: { id: metaLoaderGlob },
					order: "pre",
					handler(code, id) {
						if (!forcedConfig) this.addWatchFile(options.configPath);
						return metaLoader.transform.call(this, code, id);
					}
				};
				if ("_fumadocs_skipViteConfig" in config && config._fumadocs_skipViteConfig) return;
				if (!options.updateViteConfig) return;
				const { getConfig } = await import("../dist-CE1l7WuI.js");
				return getConfig({ root });
			},
			async buildStart() {
				await core.emit({ write: true });
			},
			async configureServer(server) {
				await core.initServer({ watcher: server.watcher });
				if (!forcedConfig) server.watcher.on("change", async (file) => {
					if (path.resolve(file) === options.configPath) {
						await core.init({ config: buildConfig((await runnerImport(options.configPath)).module, root) });
						await core.emit({ write: true });
					}
				});
			}
		},
		mdxPlugin,
		metaPlugin
	];
}
async function postInstall(pluginOptions = {}) {
	const { loadConfig } = await import("../load-from-file-CVTAyohB.js").then((n) => n.n);
	const core = createViteCore(applyDefaults(process.cwd(), pluginOptions));
	await core.init({ config: loadConfig(core, true) });
	await core.emit({ write: true });
}
function createViteCore({ index, configPath, outDir }) {
	if (index === true) index = {};
	return createCore({
		environment: "vite",
		configPath,
		outDir,
		plugins: [index && indexFile({
			...index,
			target: index.target ?? "vite"
		})]
	});
}
function applyDefaults(root, options) {
	return {
		updateViteConfig: options.updateViteConfig ?? true,
		index: options.index ?? true,
		configPath: path.resolve(root, options.configPath ?? _Defaults.configPath),
		outDir: path.resolve(root, options.outDir ?? _Defaults.outDir)
	};
}
//#endregion
export { mdx as default, postInstall };
