// fumadocs-mdx/dist/node/_loader.js
import { t as createMdxLoader } from "../mdx-wSuDwc6c.js";
import { n as createCore, t as _Defaults } from "../core-Cq6V99iz.js";
import { n as toNode, o as createStandaloneConfigLoader } from "../adapter-DPEpBHhE.js";
import { t as createMetaLoader } from "../meta-UmuO8lue.js";
import { r as metaLoaderGlob, t as mdxLoaderGlob } from "../loaders-CUuRGbiE.js";
//#region src/node/_loader.ts
let cachedLoaders;
const initialize = (options) => {
	const configLoader = createStandaloneConfigLoader({
		core: createCore({
			environment: "node-loader",
			...options,
			configPath: options.configPath ?? _Defaults.configPath,
			outDir: options.outDir ?? _Defaults.outDir
		}),
		buildConfig: true,
		mode: "production"
	});
	cachedLoaders = [toNode(mdxLoaderGlob, createMdxLoader(configLoader))];
	if (!options.disableMetaFile) cachedLoaders.push(toNode(metaLoaderGlob, createMetaLoader(configLoader)));
};
const load = async (url, context, nextLoad) => {
	if (!cachedLoaders) throw new Error("not initialized");
	const hooks = cachedLoaders;
	function run(i, url, context) {
		if (i >= hooks.length) return nextLoad(url, context);
		return hooks[i](url, context, (url, ctx) => run(i + 1, url, {
			...context,
			...ctx
		}));
	}
	return run(0, url, context);
};
//#endregion
export { initialize, load };
