// fumadocs-mdx/dist/node/loader.js
import { t as createMdxLoader } from "../mdx-wSuDwc6c.js";
import { n as createCore, t as _Defaults } from "../core-Cq6V99iz.js";
import { n as toNode, o as createStandaloneConfigLoader } from "../adapter-DPEpBHhE.js";
import { t as createMetaLoader } from "../meta-UmuO8lue.js";
import { r as metaLoaderGlob, t as mdxLoaderGlob } from "../loaders-CUuRGbiE.js";
//#region src/node/loader.ts
const configLoader = createStandaloneConfigLoader({
	core: createCore({
		environment: "node",
		configPath: _Defaults.configPath,
		outDir: _Defaults.outDir
	}),
	buildConfig: true,
	mode: "production"
});
const mdxLoader = toNode(mdxLoaderGlob, createMdxLoader(configLoader));
const metaLoader = toNode(metaLoaderGlob, createMetaLoader(configLoader));
/**
* @deprecated use the `register()` function from `fumadocs-mdx/node` instead.
*/
const load = (url, context, nextLoad) => {
	return mdxLoader(url, context, (v, ctx) => metaLoader(v, {
		...context,
		...ctx
	}, nextLoad));
};
//#endregion
export { load };
