// fumapress/dist/components/blog.js
import { Link } from "./link.js";
import { getCreationDate, getPressContext } from "../lib/shared.js";
import { cn } from "../lib/cn.js";
import { joinPathname } from "../lib/pathname.js";
import { getBlogContext } from "../plugins/blog.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { CornerLeftUpIcon } from "lucide-react";
import { T } from "@fuma-translate/react";
import { buttonVariants } from "fumadocs-ui/components/ui/button";
//#region src/components/blog.tsx
function BlogItem({ page, date }) {
	return /* @__PURE__ */ jsxs(Link, {
		href: page.url,
		className: "flex flex-col bg-fd-card rounded-2xl border shadow-sm p-4 transition-colors hover:bg-fd-accent hover:text-fd-accent-foreground",
		children: [
			/* @__PURE__ */ jsx("p", {
				className: "font-medium",
				children: page.data.title
			}),
			/* @__PURE__ */ jsx("p", {
				className: "text-sm text-fd-muted-foreground",
				children: page.data.description
			}),
			/* @__PURE__ */ jsx("p", {
				className: "mt-auto pt-4 text-xs text-fd-primary",
				children: date.toDateString()
			})
		]
	});
}
async function OrderedBlogGrid({ posts }) {
	const ctx = getPressContext();
	const currentDate = new Date(Date.now());
	const orderedPosts = [];
	for (const page of posts) {
		const date = await getCreationDate(ctx, page);
		orderedPosts.push({
			page,
			date: date ?? currentDate
		});
	}
	orderedPosts.sort((a, b) => b.date.getTime() - a.date.getTime());
	return /* @__PURE__ */ jsx("div", {
		className: "grid grid-cols-1 gap-2 mt-4 md:grid-cols-3 xl:grid-cols-4",
		children: orderedPosts.map(({ page, date }) => /* @__PURE__ */ jsx(BlogItem, {
			page,
			date
		}, page.url))
	});
}
function LinkToHome({ lang }) {
	const { indexPath } = getBlogContext();
	if (!indexPath) return;
	return /* @__PURE__ */ jsxs(Link, {
		href: lang ? joinPathname(lang, indexPath) : indexPath,
		className: cn(buttonVariants({
			variant: "ghost",
			className: "text-fd-muted-foreground gap-2"
		})),
		children: [/* @__PURE__ */ jsx(CornerLeftUpIcon, { className: "size-3.5" }), /* @__PURE__ */ jsx(T, {
			text: "Back to Home",
			note: "blog"
		})]
	});
}
//#endregion
export { BlogItem, LinkToHome, OrderedBlogGrid };
