// fumapress/dist/components/blog-panel.js
"use client";
import { cn } from "../lib/cn.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { ChevronDown, ShareIcon } from "lucide-react";
import { TOCProvider, TOCScrollArea, useTOCItems } from "fumadocs-ui/components/toc";
import { TOCItem, TOCItems } from "fumadocs-ui/components/toc/default";
import { Collapsible, CollapsibleContent, CollapsibleTrigger } from "fumadocs-ui/components/ui/collapsible";
import { cva } from "class-variance-authority";
import { useCopyButton } from "fumadocs-ui/utils/use-copy-button";
import { T } from "@fuma-translate/react";
//#region src/components/blog-panel.tsx
const panelButtonVariants = cva("inline-flex items-center font-medium gap-2 px-3 py-2 transition-all duration-150 rounded-lg hover:text-fd-accent-foreground hover:bg-fd-accent active:scale-95");
function BlogPanel() {
	const items = useTOCItems();
	const [open, setOpen] = useState(false);
	const [isSuccessful, onCopy] = useCopyButton(() => {
		if (navigator.share) return navigator.share({
			title: document.title,
			url: location.href
		});
		else return navigator.clipboard.writeText(location.href);
	});
	return /* @__PURE__ */ jsxs(Collapsible, {
		open,
		onOpenChange: setOpen,
		className: "fixed w-full max-w-[calc(100%---spacing(4))] left-1/2 -translate-x-1/2 bottom-2 border shadow-md text-sm bg-fd-secondary/80 backdrop-blur-sm z-20 p-1 rounded-xl sm:max-w-[400px] sm:bottom-4",
		children: [/* @__PURE__ */ jsx(CollapsibleContent, { children: /* @__PURE__ */ jsx(TOCScrollArea, {
			className: "max-h-[min(600px,calc(100vh---spacing(30)))]",
			children: /* @__PURE__ */ jsx(TOCItems, { children: items.map((item) => /* @__PURE__ */ jsx(TOCItem, {
				item,
				onClick: () => setOpen(false)
			}, item.url)) })
		}) }), /* @__PURE__ */ jsxs("div", {
			className: "flex flex-row gap-2",
			children: [/* @__PURE__ */ jsxs(CollapsibleTrigger, {
				className: cn(panelButtonVariants(), "min-w-0"),
				children: [/* @__PURE__ */ jsx("span", {
					className: "truncate",
					children: /* @__PURE__ */ jsx(T, {
						text: "Table of Contents",
						note: "blog panel"
					})
				}), /* @__PURE__ */ jsx(ChevronDown, { className: cn("size-3.5 shrink-0 text-fd-muted-foreground transition-transform", open && "rotate-180") })]
			}), /* @__PURE__ */ jsxs("button", {
				className: cn(panelButtonVariants(), "ms-auto text-fd-muted-foreground"),
				onClick: onCopy,
				children: [/* @__PURE__ */ jsx(ShareIcon, { className: "size-3.5 shrink-0" }), isSuccessful ? /* @__PURE__ */ jsx(T, {
					text: "Copied",
					note: "blog panel"
				}) : /* @__PURE__ */ jsx(T, {
					text: "Share",
					note: "blog panel"
				})]
			})]
		})]
	});
}
function BlogProvider({ toc, children }) {
	return /* @__PURE__ */ jsx(TOCProvider, {
		toc,
		children
	});
}
//#endregion
export { BlogPanel, BlogProvider };
