// fumapress/dist/components/orama-search-static.js
"use client";
import { resolveBaseUrl } from "../lib/pathname.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { SearchDialog, SearchDialogClose, SearchDialogContent, SearchDialogHeader, SearchDialogIcon, SearchDialogInput, SearchDialogList, SearchDialogOverlay } from "fumadocs-ui/components/dialog/search";
import { useDocsSearch } from "fumadocs-core/search/client";
import { useI18n } from "fumadocs-ui/contexts/i18n";
import { oramaStaticClient } from "fumadocs-core/search/client/orama-static";
import { create } from "@orama/orama";
//#region src/components/orama-search-static.tsx
function initOrama() {
	return create({
		schema: { _: "string" },
		language: "english"
	});
}
function DefaultSearchDialog(props) {
	const { locale } = useI18n();
	const { search, setSearch, query } = useDocsSearch({ client: oramaStaticClient({
		initOrama,
		locale,
		from: resolveBaseUrl(import.meta.env.BASE_URL, "/api/search")
	}) });
	return /* @__PURE__ */ jsxs(SearchDialog, {
		search,
		onSearchChange: setSearch,
		isLoading: query.isLoading,
		...props,
		children: [/* @__PURE__ */ jsx(SearchDialogOverlay, {}), /* @__PURE__ */ jsxs(SearchDialogContent, { children: [/* @__PURE__ */ jsxs(SearchDialogHeader, { children: [
			/* @__PURE__ */ jsx(SearchDialogIcon, {}),
			/* @__PURE__ */ jsx(SearchDialogInput, {}),
			/* @__PURE__ */ jsx(SearchDialogClose, {})
		] }), /* @__PURE__ */ jsx(SearchDialogList, { items: query.data !== "empty" ? query.data : null })] })]
	});
}
//#endregion
export { DefaultSearchDialog as default };
