// fumapress/dist/components/provider.js
"use client";
import { Link } from "./link.js";
import { useRouter } from "../client.js";
import { Image } from "./image.js";
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { FrameworkProvider } from "fumadocs-core/framework";
import { RootProvider } from "fumadocs-ui/provider/base";
//#region src/components/provider.tsx
const framework = {
	useParams() {
		console.warn("[Fumadocs] useParams() is not supported on Fumapress");
		return useMemo(() => ({}), []);
	},
	usePathname() {
		return useRouter().path;
	},
	useRouter() {
		const router = useRouter();
		return useMemo(() => ({
			push: router.push.bind(router),
			refresh: router.reload.bind(router)
		}), [router]);
	},
	Image: ({ priority, ...props }) => /* @__PURE__ */ jsx(Image, {
		fetchPriority: priority ? "high" : void 0,
		loading: priority ? "eager" : void 0,
		...props
	}),
	Link: ({ prefetch = true, ...props }) => /* @__PURE__ */ jsx(Link, {
		unstable_prefetchOnEnter: prefetch,
		...props
	})
};
function PressProvider(props) {
	return /* @__PURE__ */ jsx(FrameworkProvider, {
		...framework,
		children: /* @__PURE__ */ jsx(RootProvider, { ...props })
	});
}
//#endregion
export { PressProvider };
