// fumapress/dist/vite.js
import { crawlFrameworkPkgs } from "./lib/vitefu.js";
import { fileURLToPath } from "node:url";
//#region src/vite.ts
function press(options) {
	return [
		core(options),
		{ name: "waku:vite-plugins:fs-router-typegen" },
		{
			name: "fumapress:internal-flags",
			enforce: "pre",
			config(config) {
				Object.assign(config, { _fumadocs_skipViteConfig: true });
			}
		}
	];
}
function core(options = {}) {
	const { generateViteConfig = true } = options;
	return {
		name: "fumapress:core",
		async config(config, { command }) {
			const out = generateViteConfig ? await crawlFrameworkPkgs({
				root: config.root ?? process.cwd(),
				isBuild: command === "build",
				isFrameworkPkgByName(pkgName) {
					if (pkgName.startsWith("@fumapress/") || pkgName.startsWith("@fumadocs/") || pkgName.startsWith("fumadocs-") || pkgName.startsWith("fumapress-") || pkgName === "fumapress") return true;
					switch (pkgName) {
						case "vite":
						case "waku":
						case "shiki": return false;
					}
				}
			}) : null;
			return {
				ssr: {
					noExternal: out?.ssr.noExternal,
					external: ["sharp"]
				},
				optimizeDeps: out?.optimizeDeps
			};
		},
		async resolveId(source, importer, options) {
			if (source === "virtual:fumapress-core/config") return this.resolve("/press.config", void 0, options);
			if (source === "@fuma-translate/react" && importer && isRscClientReferencesModule(importer)) return fileURLToPath(import.meta.resolve(source));
			if (source.startsWith("virtual:root.css")) {
				const [_id, query = ""] = source.split("?", 2);
				return await this.resolve(`/src/app.css?${query}`) ?? await this.resolve(`fumapress/css/default.css?${query}`);
			}
		},
		async load(id) {
			if (id === "\0virtual:vite-rsc-waku/server-entry-inner") return getManagedServerEntry();
		}
	};
}
function isRscClientReferencesModule(id) {
	if (id.startsWith("\0")) id = id.slice(1);
	return id.startsWith("virtual:vite-rsc/client-references") || id.startsWith("virtual:vite-rsc/client-package-proxy");
}
function getManagedServerEntry() {
	return `import _adapter from 'waku/adapters/default';
import pressConfig from 'virtual:fumapress-core/config';
import { createRouter } from 'fumapress/router';
import { fsRouterFn } from 'fumapress/router/fs';

const modules = import.meta.glob("./pages/**/*.{ts,tsx,js,jsx}", {
  base: '/src',
});

const router = await createRouter(pressConfig);
const pages = router.createPages(fsRouterFn(modules));
const middlewareFns = router.createMiddlewares();
const adapter = router.patchAdapter(_adapter);

export default adapter(pages, { middlewareFns });
`;
}
//#endregion
export { press as default };
