// fumadocs-mdx/dist/next/index.js
import { n as createCore, t as _Defaults } from "../core-Cq6V99iz.js";
import { t as loadConfig } from "../load-from-file-CVTAyohB.js";
import { i as metaLoaderQueryGlob, n as metaLoaderFileGlob, t as mdxLoaderGlob } from "../loaders-CUuRGbiE.js";
import { t as indexFile } from "../index-file-C9Bw7He1.js";
import * as path$1 from "node:path";
//#region src/next/index.ts
const defaultPageExtensions = [
	"mdx",
	"md",
	"jsx",
	"js",
	"tsx",
	"ts"
];
function createMDX(createOptions = {}) {
	const core = createNextCore(applyDefaults(createOptions));
	const isDev = process.env.NODE_ENV === "development";
	if (process.env._FUMADOCS_MDX !== "1") {
		process.env._FUMADOCS_MDX = "1";
		init(isDev, core);
	}
	return (nextConfig = {}) => {
		const { configPath, outDir } = core.getOptions();
		const loaderOptions = {
			configPath,
			outDir,
			absoluteCompiledConfigPath: path$1.resolve(core.getCompiledConfigPath()),
			isDev
		};
		const turbopack = {
			...nextConfig.turbopack,
			rules: {
				...nextConfig.turbopack?.rules,
				"*.{md,mdx}": {
					loaders: [{
						loader: "fumadocs-mdx/webpack/mdx",
						options: loaderOptions
					}],
					as: "*.js"
				},
				"*.json": {
					condition: { query: metaLoaderQueryGlob },
					loaders: [{
						loader: "fumadocs-mdx/webpack/meta",
						options: loaderOptions
					}],
					as: "*.json"
				},
				"*.yaml": {
					condition: { query: metaLoaderQueryGlob },
					loaders: [{
						loader: "fumadocs-mdx/webpack/meta",
						options: loaderOptions
					}],
					as: "*.js"
				}
			}
		};
		return {
			...nextConfig,
			turbopack,
			pageExtensions: nextConfig.pageExtensions ?? defaultPageExtensions,
			webpack: (config, options) => {
				config.resolve ||= {};
				config.module ||= {};
				config.module.rules ||= [];
				config.module.rules.push({
					test: mdxLoaderGlob,
					use: [options.defaultLoaders.babel, {
						loader: "fumadocs-mdx/webpack/mdx",
						options: loaderOptions
					}]
				}, {
					test: metaLoaderFileGlob,
					resourceQuery: metaLoaderQueryGlob,
					enforce: "pre",
					use: [{
						loader: "fumadocs-mdx/webpack/meta",
						options: loaderOptions
					}]
				});
				config.plugins ||= [];
				return nextConfig.webpack?.(config, options) ?? config;
			}
		};
	};
}
async function init(dev, core) {
	async function initOrReload() {
		await core.init({ config: loadConfig(core, true) });
		await core.emit({ write: true });
	}
	async function devServer() {
		const { FSWatcher } = await import("chokidar");
		const { configPath, outDir } = core.getOptions();
		const watcher = new FSWatcher({
			ignoreInitial: true,
			persistent: true,
			ignored: [outDir]
		});
		watcher.add(configPath);
		for (const collection of core.getCollections()) watcher.add(collection.dir);
		for (const workspace of core.getWorkspaces().values()) for (const collection of workspace.getCollections()) watcher.add(collection.dir);
		watcher.on("ready", () => {
			console.log("[MDX] started dev server");
		});
		const absoluteConfigPath = path$1.resolve(configPath);
		watcher.on("all", async (_event, file) => {
			if (path$1.resolve(file) === absoluteConfigPath) {
				watcher.removeAllListeners();
				await watcher.close();
				await initOrReload();
				console.log("[MDX] restarting dev server");
				await devServer();
			}
		});
		process.on("exit", () => {
			if (watcher.closed) return;
			console.log("[MDX] closing dev server");
			watcher.close();
		});
		await core.initServer({ watcher });
	}
	await initOrReload();
	if (dev) await devServer();
}
async function postInstall(options = {}) {
	const core = createNextCore(applyDefaults(options));
	await core.init({ config: loadConfig(core, true) });
	await core.emit({ write: true });
}
function applyDefaults(options) {
	return {
		index: {},
		outDir: options.outDir ?? _Defaults.outDir,
		configPath: options.configPath ?? _Defaults.configPath
	};
}
function createNextCore(options) {
	return createCore({
		environment: "next",
		outDir: options.outDir,
		configPath: options.configPath,
		plugins: [options.index && indexFile(options.index)]
	});
}
//#endregion
export { createMDX, postInstall };
