// fumapress/dist/layouts/blog.tags.js
import { Link } from "../components/link.js";
import { getPressContext } from "../lib/shared.js";
import { getTags, groupTags } from "../lib/shared/blog.js";
import { getBlogContext } from "../plugins/blog.js";
import { LinkToHome, OrderedBlogGrid } from "../components/blog.js";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { NewspaperIcon, TagIcon } from "lucide-react";
import { T } from "@fuma-translate/react";
//#region src/layouts/blog.tags.tsx
function createBlogTagsPage({ heading, description } = {}) {
	return async function BlogTagsPage({ lang }) {
		const ctx = getPressContext();
		const { isBlog } = getBlogContext();
		const grouped = await groupTags(ctx, (await ctx.getLoader()).getPages(lang).filter((page) => isBlog.call(ctx, page)));
		return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsxs("div", {
			className: "flex flex-col items-start gap-4 border-y px-4 pt-3.5 pb-6 bg-fd-card text-fd-card-foreground shadow-inner max-sm:-mx-4 sm:rounded-xl sm:border",
			children: [
				/* @__PURE__ */ jsx(LinkToHome, { lang }),
				/* @__PURE__ */ jsx("h1", {
					className: "font-semibold text-2xl",
					children: heading ?? /* @__PURE__ */ jsx(T, {
						text: "All Tags",
						note: "blog tags page"
					})
				}),
				/* @__PURE__ */ jsx("p", {
					className: "text-fd-muted-foreground empty:hidden",
					children: description ?? /* @__PURE__ */ jsxs("span", {
						className: "flex items-center gap-1",
						children: [/* @__PURE__ */ jsx(TagIcon, { className: "size-3.5 text-fd-primary" }), /* @__PURE__ */ jsx(T, {
							text: "{count} tags in total.",
							note: "blog tags page",
							variables: { count: String(grouped.size) }
						})]
					})
				})
			]
		}), /* @__PURE__ */ jsx("div", {
			className: "grid grid-cols-2 sm:grid-cols-3 gap-2 mt-4 md:grid-cols-4 xl:grid-cols-6",
			children: Array.from(grouped.entries()).sort((a, b) => b[1] - a[1]).map(([tag, count]) => /* @__PURE__ */ jsxs(Link, {
				href: `/blog/tags/${tag}`,
				className: "flex flex-row items-center gap-2 bg-fd-card text-fd-card-foreground border font-mono rounded-lg px-2 py-1 transition-colors hover:bg-fd-accent hover:text-fd-accent-foreground",
				children: [
					/* @__PURE__ */ jsx(TagIcon, { className: "size-3.5 text-fd-muted-foreground" }),
					/* @__PURE__ */ jsx("p", {
						className: "font-medium",
						children: tag
					}),
					/* @__PURE__ */ jsx("p", {
						className: "ms-auto text-sm text-fd-muted-foreground",
						children: count
					})
				]
			}, tag))
		})] });
	};
}
function createBlogTagPage({ heading, description } = {}) {
	return async function BlogTagPage({ lang, tag }) {
		const ctx = getPressContext();
		const { isBlog } = getBlogContext();
		const source = await ctx.getLoader();
		const posts = [];
		for (const page of source.getPages(lang)) {
			if (!isBlog.call(ctx, page)) continue;
			const tags = await getTags(ctx, page);
			if (!tags || !tags.includes(tag)) continue;
			posts.push(page);
		}
		return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsxs("div", {
			className: "flex flex-col items-start gap-4 border-y p-4 pt-3.5 bg-fd-card text-fd-card-foreground shadow-inner max-sm:-mx-4 sm:rounded-xl sm:border",
			children: [
				/* @__PURE__ */ jsx(LinkToHome, { lang }),
				/* @__PURE__ */ jsx("h1", {
					className: "font-semibold text-2xl",
					children: heading ?? /* @__PURE__ */ jsxs("span", {
						className: "inline-flex gap-2 items-center",
						children: [/* @__PURE__ */ jsx(TagIcon, { className: "text-fd-primary size-6" }), /* @__PURE__ */ jsx(T, {
							text: "Tag \"{tag}\"",
							note: "blog tag page",
							variables: { tag }
						})]
					})
				}),
				/* @__PURE__ */ jsx("p", {
					className: "text-fd-muted-foreground empty:hidden",
					children: description ?? /* @__PURE__ */ jsxs("span", {
						className: "inline-flex items-center gap-1",
						children: [/* @__PURE__ */ jsx(NewspaperIcon, { className: "text-fd-primary size-3.5" }), /* @__PURE__ */ jsx(T, {
							text: "{count} matching blog posts.",
							note: "blog tag page",
							variables: { count: String(posts.length) }
						})]
					})
				})
			]
		}), /* @__PURE__ */ jsx(OrderedBlogGrid, { posts })] });
	};
}
//#endregion
export { createBlogTagPage, createBlogTagsPage };
