// fumapress/dist/layouts/blog.js
import { Link } from "../components/link.js";
import { getCreationDate, getPressContext, renderBody, renderPageMeta, renderToc } from "../lib/shared.js";
import { BlogPanel, BlogProvider } from "../components/blog-panel.js";
import { getTags } from "../lib/shared/blog.js";
import { joinPathname } from "../lib/pathname.js";
import { createHomeLayout } from "./home.js";
import { getBlogContext } from "../plugins/blog.js";
import { LinkToHome } from "../components/blog.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { TagIcon } from "lucide-react";
//#region src/layouts/blog.tsx
/** You can use `createHomeLayout()` directly, this is only a wrapper */
function createBlogLayout(options) {
	return createHomeLayout(options);
}
function createBlogLayoutPage(options = {}) {
	const { render } = options;
	return async function BlogLayoutPage({ page, lang }) {
		const ctx = getPressContext();
		const { tagsPath } = getBlogContext();
		const tags = await getTags(ctx, page);
		const _raw = await render?.call(ctx, page);
		const result = {
			body: _raw?.body ?? await renderBody(ctx, page, "[Fumapress] Please specify the `render` option in createBlogLayoutPage()"),
			toc: _raw?.toc ?? await renderToc(ctx, page) ?? [],
			creationDate: _raw?.creationDate ?? await getCreationDate(ctx, page)
		};
		return /* @__PURE__ */ jsxs(BlogProvider, {
			toc: result.toc,
			children: [
				renderPageMeta(page, ctx),
				/* @__PURE__ */ jsxs("div", {
					className: "flex flex-col gap-4 items-center border-y px-4 pt-3.5 pb-6 bg-fd-card text-fd-card-foreground shadow-inner max-sm:-mx-4 sm:rounded-xl sm:border",
					children: [
						/* @__PURE__ */ jsx("div", {
							className: "flex flex-row items-center gap-2 w-full max-w-[900px]",
							children: /* @__PURE__ */ jsx(LinkToHome, { lang })
						}),
						/* @__PURE__ */ jsx("h1", {
							className: "font-semibold text-2xl w-full max-w-[900px]",
							children: page.data.title
						}),
						/* @__PURE__ */ jsx("p", {
							className: "text-fd-muted-foreground w-full max-w-[900px]",
							children: page.data.description
						}),
						tags && tags.length > 0 && /* @__PURE__ */ jsxs("div", {
							className: "flex flex-row items-center gap-2 flex-wrap w-full max-w-[900px] text-sm text-fd-primary-foreground font-mono",
							children: [/* @__PURE__ */ jsx(TagIcon, { className: "size-4 text-fd-muted-foreground" }), tags.map((t) => {
								if (tagsPath !== false) return /* @__PURE__ */ jsx(Link, {
									href: joinPathname(lang ?? "", tagsPath, t),
									className: "px-1.5 py-0.5 rounded-lg bg-fd-primary",
									children: t
								}, t);
								return /* @__PURE__ */ jsx("p", {
									className: "px-1.5 py-0.5 rounded-lg bg-fd-primary",
									children: t
								}, t);
							})]
						})
					]
				}),
				/* @__PURE__ */ jsx("article", {
					className: "prose mt-6 mx-auto w-full max-w-[900px]",
					children: result.body
				}),
				/* @__PURE__ */ jsx("div", { className: "h-12" }),
				/* @__PURE__ */ jsx(BlogPanel, {})
			]
		});
	};
}
//#endregion
export { createBlogLayout, createBlogLayoutPage };
