// fumapress/dist/layouts/blog.index.js
import { Link } from "../components/link.js";
import { getPressContext } from "../lib/shared.js";
import { cn } from "../lib/cn.js";
import { joinPathname } from "../lib/pathname.js";
import { getBlogContext } from "../plugins/blog.js";
import { OrderedBlogGrid } from "../components/blog.js";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { ListIcon } from "lucide-react";
import { T } from "@fuma-translate/react";
import { buttonVariants } from "fumadocs-ui/components/ui/button";
//#region src/layouts/blog.index.tsx
function createBlogIndexPage({ heading, description } = {}) {
	return async function BlogIndexPage({ lang }) {
		const ctx = getPressContext();
		const { tagsPath, isBlog } = getBlogContext();
		const source = await ctx.getLoader();
		return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsxs("div", {
			className: "flex flex-col gap-4 items-start border-2 border-dashed border-fd-primary rounded-xl bg-fd-primary/10 p-4 z-2 md:p-8",
			children: [
				/* @__PURE__ */ jsx("h1", {
					className: "text-3xl font-semibold",
					children: heading ?? /* @__PURE__ */ jsx(T, {
						text: "Blog",
						note: "blog"
					})
				}),
				/* @__PURE__ */ jsx("p", {
					className: "text-fd-primary overline decoration-fd-primary empty:hidden",
					children: description
				}),
				tagsPath !== false && /* @__PURE__ */ jsxs(Link, {
					href: lang ? joinPathname(lang, tagsPath) : tagsPath,
					className: cn(buttonVariants({ variant: "primary" }), "gap-2"),
					children: [/* @__PURE__ */ jsx(ListIcon, { className: "size-4" }), /* @__PURE__ */ jsx(T, {
						text: "All Tags",
						note: "blog tags page"
					})]
				})
			]
		}), /* @__PURE__ */ jsx(OrderedBlogGrid, { posts: source.getPages(lang).filter(isBlog.bind(ctx)) })] });
	};
}
//#endregion
export { createBlogIndexPage };
