// fumapress/dist/layouts/home.js
import { baseLayoutProps, createTransformChildren, getPressContext, mergeLayoutConfigs, renderBody, renderPageMeta } from "../lib/shared.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { HomeLayout } from "fumadocs-ui/layouts/home";
//#region src/layouts/home.tsx
function createHomeLayoutPage({ layout: Container = createHomeLayout(), render } = {}) {
	return async function Layout({ lang, page }) {
		const ctx = getPressContext();
		const _raw = await render?.call(ctx, page);
		const body = _raw?.body ?? await renderBody(ctx, page, "[Fumapress] Please specify the `render` option in createHomeLayoutPage()");
		return /* @__PURE__ */ jsxs(Container, {
			lang,
			layoutProps: _raw?.layoutProps,
			children: [renderPageMeta(page, ctx), body]
		});
	};
}
function createHomeLayout({ layoutProps: getLayoutProps, inherit: { layoutProps: inheritLayoutProps = true } = {} } = {}) {
	const THomeLayout = createTransformChildren(HomeLayout);
	return async function Layout({ lang, layoutProps, children }) {
		const ctx = getPressContext();
		const { layouts, data: { "core:home-layout": layoutData } } = ctx;
		const inherited = inheritLayoutProps ? await layouts.defaultProps?.call(ctx, { lang }) : void 0;
		let result = {
			body: children,
			layoutProps: mergeLayoutConfigs(baseLayoutProps(ctx), inherited, typeof getLayoutProps === "function" ? await getLayoutProps.call(ctx) : getLayoutProps, layoutProps)
		};
		if (layoutData?.renderers) for (const r of layoutData.renderers) result = await r(result);
		return /* @__PURE__ */ jsx(THomeLayout, {
			props: result.layoutProps,
			children: /* @__PURE__ */ jsx("main", {
				"data-fd-home-layout-container": "",
				className: "flex flex-col w-full max-w-[1400px] flex-1 px-4 py-6 mx-auto",
				children: result.body
			})
		});
	};
}
//#endregion
export { createHomeLayout, createHomeLayoutPage };
