// fumapress/dist/layouts/notebook.js
import { baseLayoutProps, createTransformChildren, getGitHubFileUrl, getLastModifiedDate, getPressContext, mergeLayoutConfigs, renderBody, renderPageMeta, renderToc } from "../lib/shared.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { DocsLayout } from "fumadocs-ui/layouts/notebook";
import { DocsBody, DocsDescription, DocsPage, DocsTitle, MarkdownCopyButton, PageLastUpdate, ViewOptionsPopover } from "fumadocs-ui/layouts/notebook/page";
//#region src/layouts/notebook.tsx
function createNotebookLayoutPage({ render, renderBody: renderDocsBody, inherit: { layoutProps: inheritLayoutProps = true } = {} } = {}) {
	const TDocsLayout = createTransformChildren(DocsLayout);
	const TDocsPage = createTransformChildren(DocsPage);
	return async function Layout({ lang, page }) {
		const ctx = getPressContext();
		const { getLoader, layouts, data: { "core:notebook-layout": layoutData } } = ctx;
		const source = await getLoader();
		const inherited = inheritLayoutProps ? await layouts.defaultProps?.call(ctx, { lang }) : void 0;
		const _raw = await render?.call(ctx, page);
		const layoutProps = mergeLayoutConfigs(baseLayoutProps(ctx), inherited, _raw?.layoutProps);
		layoutProps.tree ??= source.getPageTree(lang);
		let result = {
			..._raw,
			lastModified: _raw?.lastModified ?? await getLastModifiedDate(ctx, page),
			pageProps: {
				..._raw?.pageProps,
				toc: _raw?.pageProps?.toc ?? await renderToc(ctx, page)
			},
			body: _raw?.body ?? await renderBody(ctx, page, "[Fumapress] Please specify the `render` option in createNotebookLayoutPage()"),
			layoutProps
		};
		if (layoutData?.renderers) {
			const renderCtx = { page };
			for (const r of layoutData.renderers) result = await r.call(renderCtx, result);
		}
		let dynamicDocsBody = (props) => /* @__PURE__ */ jsx(DocsBody, { ...props });
		if (renderDocsBody) dynamicDocsBody = (props) => renderDocsBody.call(ctx, {
			lang,
			page,
			props,
			default: dynamicDocsBody
		});
		return /* @__PURE__ */ jsxs(TDocsLayout, {
			props: result.layoutProps,
			children: [renderPageMeta(page, ctx), /* @__PURE__ */ jsxs(TDocsPage, {
				props: result.pageProps,
				children: [
					/* @__PURE__ */ jsx(DocsTitle, { children: page.data.title }),
					/* @__PURE__ */ jsx(DocsDescription, {
						className: "mb-0",
						children: page.data.description
					}),
					/* @__PURE__ */ jsxs("div", {
						className: "flex flex-row gap-2 items-center border-b pt-2 pb-6",
						children: [result.markdownUrl && /* @__PURE__ */ jsx(MarkdownCopyButton, { markdownUrl: result.markdownUrl }), /* @__PURE__ */ jsx(ViewOptionsPopover, {
							markdownUrl: result.markdownUrl,
							githubUrl: page.absolutePath ? getGitHubFileUrl(ctx, page.absolutePath) : void 0
						})]
					}),
					dynamicDocsBody({ children: result.body }),
					result.lastModified && /* @__PURE__ */ jsx(PageLastUpdate, { date: result.lastModified })
				]
			})]
		});
	};
}
//#endregion
export { createNotebookLayoutPage };
