// fumapress/dist/layouts/root.js
import { getPressContext, renderRootMeta } from "../lib/shared.js";
import { PressProvider } from "../components/provider.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { i18nProvider, uiTranslations } from "fumadocs-ui/i18n";
import stylesInline from "virtual:root.css?inline";
import stylesHref from "virtual:root.css?url";
//#region src/layouts/root.tsx
let styleTag;
if (import.meta.env.DEV) styleTag = /* @__PURE__ */ jsx("style", { children: stylesInline });
else styleTag = /* @__PURE__ */ jsx("link", {
	rel: "stylesheet",
	href: stylesHref
});
function createRootLayout(options) {
	return async function({ lang, children }) {
		const ctx = getPressContext();
		const hooks = ctx.data["core:provider"];
		let providerProps = {
			...options?.providerProps,
			children
		};
		if (ctx.translationsConfig && "config" in ctx.translationsConfig) providerProps.i18n ??= i18nProvider(ctx.translationsConfig.extend(uiTranslations()), lang);
		else if (ctx.translationsConfig) providerProps.i18n ??= i18nProvider(ctx.translationsConfig.extend(uiTranslations()));
		if (hooks) for (const hook of hooks) providerProps = await hook(providerProps);
		return /* @__PURE__ */ jsxs("html", {
			lang: lang ?? "en",
			suppressHydrationWarning: true,
			children: [/* @__PURE__ */ jsxs("head", { children: [styleTag, renderRootMeta(ctx)] }), /* @__PURE__ */ jsx("body", {
				"data-version": "1.0",
				className: "flex flex-col min-h-screen",
				children: /* @__PURE__ */ jsx(PressProvider, { ...providerProps })
			})]
		});
	};
}
//#endregion
export { createRootLayout };
