// fumadocs-mdx/dist/build-default-DCQTM_P7.js
import { n as flattenNode, r as applyMdxPreset, t as remarkInclude } from "./remark-include-B6rmf4sh.js";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import { createProcessor } from "@mdx-js/mdx";
import { valueToEstree } from "estree-util-value-to-estree";
import { removePosition } from "unist-util-remove-position";
import { remarkLLMs } from "fumadocs-core/mdx-plugins/remark-llms";
//#region src/loaders/mdx/remark-postprocess.ts
/**
* - collect references
* - write frontmatter (auto-title & description)
*/
function remarkPostprocess({ includeProcessedMarkdown = false, includeMDAST = false, extractLinkReferences = false, valueToExport = [] }) {
	return (tree, file) => {
		const frontmatter = file.data.frontmatter ??= {};
		if (!frontmatter.title) visit(tree, "heading", (node) => {
			if (node.depth === 1) {
				frontmatter.title = flattenNode(node);
				return false;
			}
		});
		file.data["mdx-export"] ??= [];
		file.data["mdx-export"].push({
			name: "frontmatter",
			value: frontmatter
		});
		if (extractLinkReferences) {
			const urls = [];
			visit(tree, "link", (node) => {
				urls.push({ href: node.url });
				return "skip";
			});
			file.data["mdx-export"].push({
				name: "extractedReferences",
				value: urls
			});
		}
		if (includeProcessedMarkdown) remarkLLMs.call(this, typeof includeProcessedMarkdown === "object" ? includeProcessedMarkdown : void 0)(tree, file, () => void 0);
		if (includeMDAST) {
			const mdast = JSON.stringify((includeMDAST === true ? {} : includeMDAST).removePosition ? removePosition(structuredClone(tree)) : tree);
			file.data["mdx-export"].push({
				name: "_mdast",
				value: mdast
			});
		}
		for (const { name, value } of file.data["mdx-export"]) tree.children.unshift(getMdastExport(name, value));
		file.data["mdx-export"] = [];
		for (const name of valueToExport) {
			if (!(name in file.data)) continue;
			tree.children.unshift(getMdastExport(name, file.data[name]));
		}
	};
}
/**
* MDX.js first converts javascript (with esm support) into mdast nodes with remark-mdx, then handle the other remark plugins
*
* Therefore, if we want to inject an export, we must convert the object into AST, then add the mdast node
*/
function getMdastExport(name, value) {
	return {
		type: "mdxjsEsm",
		value: "",
		data: { estree: {
			type: "Program",
			sourceType: "module",
			body: [{
				type: "ExportNamedDeclaration",
				attributes: [],
				specifiers: [],
				source: null,
				declaration: {
					type: "VariableDeclaration",
					kind: "let",
					declarations: [{
						type: "VariableDeclarator",
						id: {
							type: "Identifier",
							name
						},
						init: valueToEstree(value)
					}]
				}
			}]
		} }
	};
}
//#endregion
//#region src/loaders/mdx/build-default.ts
async function buildJSMDX(core, collection, { filePath, frontmatter, source, _compiler, environment, isDevelopment }) {
	const mdxOptions = await getMDXOptions(core.getConfig(), collection, environment);
	function getProcessor(format) {
		const cache = core.cache;
		const key = `build-mdx:${collection?.name ?? "global"}:${format}`;
		let processor = cache.get(key);
		if (!processor) {
			const postprocessOptions = {
				_format: format,
				...collection?.postprocess
			};
			processor = createProcessor({
				outputFormat: "program",
				development: isDevelopment,
				...mdxOptions,
				remarkPlugins: [
					remarkInclude,
					...mdxOptions.remarkPlugins ?? [],
					[remarkPostprocess, postprocessOptions]
				],
				format
			});
			cache.set(key, processor);
		}
		return processor;
	}
	let vfile = new VFile({
		value: source,
		path: filePath,
		cwd: collection?.cwd,
		data: {
			frontmatter,
			_compiler,
			_getProcessor: getProcessor
		}
	});
	if (collection) vfile = await core.transformVFile({
		collection,
		filePath,
		source
	}, vfile);
	const out = await getProcessor(filePath.endsWith(".mdx") ? "mdx" : "md").process(vfile);
	return {
		code: String(out.value),
		map: out.map
	};
}
const mdxOptionsCache = /* @__PURE__ */ new WeakMap();
function getMDXOptions(config, collection, environment = "bundler") {
	let cacheMap = mdxOptionsCache.get(config);
	if (!cacheMap) {
		cacheMap = /* @__PURE__ */ new Map();
		mdxOptionsCache.set(config, cacheMap);
	}
	const key = collection ? `${environment}:${collection.name}` : environment;
	const cached = cacheMap.get(key);
	if (cached) return cached;
	let result;
	if (collection?.mdxOptions) {
		const optionsFn = collection.mdxOptions;
		result = typeof optionsFn === "function" ? optionsFn(environment) : optionsFn;
	} else result = (async () => {
		const optionsFn = config.global.mdxOptions;
		return applyMdxPreset(typeof optionsFn === "function" ? await optionsFn() : optionsFn)(environment);
	})();
	cacheMap.set(key, result);
	return result;
}
//#endregion
export { buildJSMDX };
