// fumadocs-mdx/dist/build-satteri-Mk7OLtzu.js
import { pathToFileURL } from "node:url";
//#region src/loaders/mdx/build-satteri.ts
async function buildSatteriMDX(core, collection, { filePath, frontmatter, source, _compiler, environment, isDevelopment }) {
	const satteriOptions = await getSatteriOptions(core.getConfig(), collection, environment);
	const postprocess = {
		_format: filePath.endsWith(".mdx") ? "mdx" : "md",
		...collection?.postprocess
	};
	const [{ compileMdx }, { remarkLlms }, { remarkInclude }] = await Promise.all([
		import(
			/* turbopackOptional: true */
			"@fumadocs/satteri/compile"
),
		import(
			/* turbopackOptional: true */
			"@fumadocs/satteri/remark-llms"
),
		import(
			/* turbopackOptional: true */
			"@fumadocs/satteri/remark-include"
)
	]);
	const data = {
		frontmatter,
		_compiler,
		_cwd: collection?.cwd,
		_valueToExport: collection?.postprocess?.valueToExport
	};
	const mdastPlugins = [remarkInclude({ cwd: collection?.cwd })];
	if (satteriOptions.mdastPlugins) mdastPlugins.push(...satteriOptions.mdastPlugins);
	if (postprocess.includeProcessedMarkdown) mdastPlugins.push(remarkLlms(typeof postprocess.includeProcessedMarkdown === "object" ? postprocess.includeProcessedMarkdown : void 0));
	mdastPlugins.push(postprocessPlugin(postprocess));
	return { code: (await compileMdx({
		source,
		filePath,
		frontmatter,
		isDevelopment,
		environment,
		options: {
			...satteriOptions,
			fileURL: pathToFileURL(filePath),
			data,
			mdastPlugins
		}
	})).code };
}
function postprocessPlugin({ extractLinkReferences = false }) {
	return {
		name: "remark-postprocess",
		heading(node, ctx) {
			const frontmatter = ctx.data.frontmatter ??= {};
			if (!frontmatter.title && node.depth === 1) frontmatter.title = ctx.textContent(node);
		},
		link(node, ctx) {
			if (extractLinkReferences) (ctx.data.extractedReferences ??= []).push({ href: node.url });
		},
		afterToJs({ result }) {
			if (extractLinkReferences) {
				result.data.extractedReferences ??= [];
				result.code += `\nexport const extractedReferences = ${JSON.stringify(result.data.extractedReferences)};`;
			}
		}
	};
}
const satteriOptionsCache = /* @__PURE__ */ new WeakMap();
async function resolvePresetOptions(input, environment) {
	const options = typeof input === "function" ? await input(environment) : input;
	const { applySatteriPreset } = await import(
		/* turbopackOptional: true */
		"@fumadocs/satteri/preset"
);
	return applySatteriPreset(options)(environment);
}
function getSatteriOptions(config, collection, environment = "bundler") {
	let cache = satteriOptionsCache.get(config);
	if (!cache) {
		cache = /* @__PURE__ */ new Map();
		satteriOptionsCache.set(config, cache);
	}
	const key = collection ? `${environment}:${collection.name}` : environment;
	const cached = cache.get(key);
	if (cached) return cached;
	let result;
	if (collection?.compiler === "satteri") result = resolvePresetOptions(collection.satteriOptions ?? config.global.satteriOptions, environment);
	else if (collection) throw new Error(`Collection "${collection.name}" uses the default MDX compiler. Use getMDXOptions() instead of getSatteriOptions().`);
	else result = resolvePresetOptions(config.global.satteriOptions, environment);
	cache.set(key, result);
	return result;
}
//#endregion
export { buildSatteriMDX };
