// fumadocs-mdx/dist/dist-CE1l7WuI.js
import fs from "node:fs/promises";
import path from "node:path";
//#region ../vite/dist/index.js
async function findClosestPkgJsonPath(dir, predicate) {
	let currentDir = dir.endsWith("package.json") ? path.dirname(dir) : dir;
	while (currentDir) {
		const pkg = path.join(currentDir, "package.json");
		try {
			if ((await fs.stat(pkg)).isFile() && (!predicate || await predicate(pkg))) return pkg;
		} catch {}
		const nextDir = path.dirname(currentDir);
		if (nextDir === currentDir) break;
		currentDir = nextDir;
	}
}
async function readJson(pkgJsonPath) {
	return JSON.parse(await fs.readFile(pkgJsonPath, "utf8"));
}
var generated_default = {
	optimizeDeps: {
		"include": [
			"@fumadocs/local-md > estree-util-to-js > source-map",
			"@fumadocs/local-md > hast-util-to-jsx-runtime > mdast-util-mdx-expression > mdast-util-from-markdown > micromark > debug",
			"@fumadocs/local-md > hast-util-to-jsx-runtime > style-to-js",
			"@fumadocs/local-md > picomatch",
			"@fumadocs/local-md > remark > unified > extend",
			"@fumadocs/mdx-remote > @mdx-js/mdx > hast-util-to-jsx-runtime > style-to-js",
			"@fumadocs/mdx-remote > @mdx-js/mdx > remark-parse > mdast-util-from-markdown > micromark > debug",
			"@fumadocs/mdx-remote > @mdx-js/mdx > source-map",
			"@fumadocs/mdx-remote > js-yaml > argparse",
			"@fumadocs/mdx-remote > unified > extend",
			"@fumadocs/story > ts-morph",
			"fumadocs-core > hast-util-to-estree > style-to-js",
			"fumadocs-core > js-yaml > argparse",
			"fumadocs-core > mdast-util-mdx > mdast-util-from-markdown > micromark > debug",
			"fumadocs-core > tinyglobby > picomatch",
			"fumadocs-core > unified > extend",
			"fumadocs-mdx > @mdx-js/mdx > hast-util-to-jsx-runtime > style-to-js",
			"fumadocs-mdx > @mdx-js/mdx > source-map",
			"fumadocs-mdx > esbuild",
			"fumadocs-mdx > js-yaml > argparse",
			"fumadocs-mdx > mdast-util-mdx > mdast-util-from-markdown > micromark > debug",
			"fumadocs-mdx > picocolors",
			"fumadocs-mdx > picomatch",
			"fumadocs-mdx > unified > extend",
			"fumadocs-twoslash > mdast-util-from-markdown > micromark > debug",
			"fumadocs-typescript > hast-util-to-estree > mdast-util-mdx-jsx > mdast-util-from-markdown > micromark > debug",
			"fumadocs-typescript > hast-util-to-estree > style-to-js",
			"fumadocs-typescript > ts-morph",
			"fumadocs-typescript > unified > extend"
		],
		"exclude": [
			"@fumadocs/basehub",
			"@fumadocs/local-md",
			"@fumadocs/mdx-remote",
			"@fumadocs/sanity",
			"@fumadocs/story",
			"@fumadocs/tailwind",
			"fumadocs-core",
			"fumadocs-mdx",
			"fumadocs-twoslash",
			"fumadocs-typescript",
			"fumadocs-ui"
		]
	},
	ssr: { "noExternal": [
		"@fumadocs/basehub",
		"@fumadocs/local-md",
		"@fumadocs/mdx-remote",
		"@fumadocs/sanity",
		"@fumadocs/story",
		"@fumadocs/tailwind",
		"fumadocs-core",
		"fumadocs-mdx",
		"fumadocs-twoslash",
		"fumadocs-typescript",
		"fumadocs-ui"
	] }
};
async function getConfig(options) {
	const pkgJsonPath = await findClosestPkgJsonPath(options.root);
	if (!pkgJsonPath) return {
		optimizeDeps: {
			include: [],
			exclude: []
		},
		ssr: { noExternal: [] }
	};
	const pkgJson = await readJson(pkgJsonPath).catch((e) => {
		throw new Error(`Unable to read ${pkgJsonPath}`, { cause: e });
	});
	const out = structuredClone(generated_default);
	const deps = /* @__PURE__ */ new Set();
	if (pkgJson.dependencies) for (const name in pkgJson.dependencies) deps.add(name);
	if (pkgJson.devDependencies) for (const name in pkgJson.devDependencies) deps.add(name);
	out.optimizeDeps.include = out.optimizeDeps.include.filter((item) => {
		const first = item.slice(0, item.indexOf(" > "));
		return deps.has(first);
	});
	out.optimizeDeps.exclude = out.optimizeDeps.exclude.filter((item) => deps.has(item));
	out.ssr.noExternal = out.ssr.noExternal.filter((item) => deps.has(item));
	return out;
}
//#endregion
export { getConfig };
