// fumadocs-mdx/dist/load-from-file-CVTAyohB.js
import { t as __exportAll } from "./rolldown-runtime-D7D4PA-g.js";
import { i as buildConfig } from "./core-Cq6V99iz.js";
import { pathToFileURL } from "node:url";
//#region src/config/load-from-file.ts
var load_from_file_exports = /* @__PURE__ */ __exportAll({ loadConfig: () => loadConfig });
async function compileConfig(core) {
	const { build } = await import("esbuild");
	const { configPath, outDir } = core.getOptions();
	if ((await build({
		entryPoints: [{
			in: configPath,
			out: "source.config"
		}],
		bundle: true,
		outdir: outDir,
		target: "node22",
		write: true,
		platform: "node",
		format: "esm",
		packages: "external",
		outExtension: { ".js": ".mjs" },
		allowOverwrite: true
	})).errors.length > 0) throw new Error("failed to compile configuration file");
}
/**
* Load config
*
* @param build - By default, it assumes the config file has been compiled. Set this `true` to compile the config first.
*/
async function loadConfig(core, build = false) {
	if (build) await compileConfig(core);
	const url = pathToFileURL(core.getCompiledConfigPath());
	url.searchParams.set("hash", Date.now().toString());
	return buildConfig(await import(url.href), process.cwd());
}
//#endregion
export { load_from_file_exports as n, loadConfig as t };
