// fumapress/dist/router/fs.js
import { joinPathname } from "../lib/pathname.js";
//#region src/router/fs.tsx
const Methods = [
	"GET",
	"POST",
	"HEAD",
	"PUT",
	"DELETE",
	"PATCH",
	"OPTIONS"
];
const ValidMethods = new Set(Methods);
const IGNORED_PATH_PARTS = /* @__PURE__ */ new Set(["_components", "_hooks"]);
const SPECIAL_BASENAME = /* @__PURE__ */ new Set([
	"_layout",
	"index",
	"_root"
]);
/** Ignore paths like `_components` and `_hooks` in pages dir */
const isIgnoredPath = (paths) => paths.some((p) => IGNORED_PATH_PARTS.has(p));
function fsRouterFn(modules, options = {}) {
	return async function(fns) {
		const { createPage, createLayout, createRoot, createApi, createSlice } = fns;
		const { pagesDir = "pages", apiDir = "_api", slicesDir = "_slices" } = options;
		const pagesDirPrefix = pagesDir + "/";
		for (const file in modules) {
			const srcPath = new URL(file, "http://localhost:3000").pathname.slice(1);
			if (!srcPath.startsWith(pagesDirPrefix)) continue;
			const pathItems = srcPath.slice(pagesDirPrefix.length).replace(/\.\w+$/, "").split("/").filter(Boolean);
			if (isIgnoredPath(pathItems)) continue;
			const path = "/" + (SPECIAL_BASENAME.has(pathItems.at(-1)) ? pathItems.slice(0, -1) : pathItems).join("/");
			const mod = await modules[file]();
			const config = await mod.getConfig?.call(this);
			if (pathItems.at(-1) === "[path]") throw new Error("Page file cannot be named [path]. This will conflict with the path prop of the page component.");
			if (pathItems[0] === apiDir) {
				const apiPath = "/" + pathItems.slice(1).join("/");
				const renderMode = config?.render ?? (this.mode === "default" ? "dynamic" : this.mode);
				if (renderMode === "static") {
					if (Object.keys(mod).length !== 2 || !mod.GET) console.warn(`API ${path} is invalid. For static API routes, only a single GET handler is supported.`);
					createApi({
						path: apiPath,
						render: renderMode,
						method: "GET",
						handler: mod.GET.bind(this),
						unstable_sourceFile: srcPath
					});
				} else {
					const entries = [];
					for (const [exportName, handler] of Object.entries(mod)) {
						if (!(exportName === "getConfig" || exportName === "default" || ValidMethods.has(exportName))) {
							console.warn(`API ${path} has an invalid export: ${exportName}. Valid exports are: ${Methods.join(", ")}`);
							continue;
						}
						if (exportName === "default") entries.push(["all", handler.bind(this)]);
						else entries.push([exportName, handler.bind(this)]);
					}
					createApi({
						path: apiPath,
						render: renderMode,
						handlers: Object.fromEntries(entries),
						unstable_sourceFile: srcPath
					});
				}
				continue;
			}
			const component = mod.default;
			const renderMode = config?.render ?? (this.mode === "default" ? "static" : this.mode);
			if (pathItems[0] === slicesDir) {
				createSlice({
					component,
					render: renderMode,
					id: pathItems.slice(1).join("/"),
					unstable_sourceFile: srcPath
				});
				continue;
			}
			if (pathItems.at(-1) === "_root") {
				createRoot({
					component,
					render: renderMode,
					unstable_sourceFile: srcPath
				});
				continue;
			}
			const autoI18n = config?.autoI18n ?? true;
			const routePath = this.i18nConfig && autoI18n ? joinPathname("[lang]/(fs)", path) : joinPathname("(fs)", path);
			if (pathItems.at(-1) === "_layout") {
				createLayout({
					path: routePath,
					component,
					render: renderMode,
					unstable_sourceFile: srcPath
				});
				continue;
			}
			createPage({
				path: routePath,
				component,
				render: renderMode,
				unstable_sourceFile: srcPath
			});
		}
	};
}
//#endregion
export { fsRouterFn };
