// fumapress/dist/router/index.js
import { appContext, initApp } from "../lib/shared.js";
import { jsx } from "react/jsx-runtime";
import { Fragment as Fragment$1 } from "react";
import { createPages } from "waku";
import { unstable_notFound, unstable_notFound as notFound, unstable_redirect, unstable_redirect as redirect } from "waku/router/server";
//#region src/router/index.tsx
async function createRouter(userConfig) {
	const context = await initApp(userConfig);
	function createPages$2(base, createPagesOptions) {
		const result = createPages(async (_fns) => {
			const layouts = context.layouts;
			const fns = {
				..._fns,
				unstable_getCreated() {
					return result;
				},
				createApiIsomorphic(config) {
					if (config.render === "static") _fns.createApi({
						render: "static",
						method: "GET",
						staticPaths: config.staticPaths,
						path: config.path,
						handler: config.handler
					});
					else _fns.createApi({
						render: "dynamic",
						path: config.path,
						handlers: { GET: config.handler }
					});
				}
			};
			async function resolvePage(slugs, lang) {
				let page = (await context.getLoader()).getPage(slugs, lang);
				if (!page) unstable_notFound();
				for (const plugin of context.plugins) {
					const resolved = await plugin.resolvePage?.call(context, page);
					if (typeof resolved === "object") page = resolved;
					else if (resolved === false) unstable_notFound();
				}
				return page;
			}
			fns.createInterceptor((next) => appContext.run(context, next));
			await base?.call(context, fns);
			for (const plugin of context.plugins) await plugin.createPages?.call(context, fns);
			const staticPaths = [];
			const defaultRenderMode = context.mode === "default" ? "static" : context.mode;
			outer: for (const page of (await context.getLoader()).getPages()) {
				for (const plugin of context.plugins) if (await plugin.resolvePage?.call(context, page) === false) continue outer;
				staticPaths.push(page.locale ? [page.locale, ...page.slugs] : page.slugs);
			}
			if (context.i18nConfig) {
				fns.createRoot({
					render: defaultRenderMode,
					component: Fragment$1
				});
				fns.createLayout({
					render: defaultRenderMode,
					path: "/[lang]",
					component: layouts.root
				});
				fns.createPage({
					render: defaultRenderMode,
					path: "/[lang]/[...slugs]",
					staticPaths,
					async component({ slugs, lang }) {
						const page = await resolvePage(slugs, lang);
						let fallback = /* @__PURE__ */ jsx(layouts.page, {
							lang,
							slugs,
							page
						});
						for (const plugin of context.plugins) {
							const res = await plugin.renderPage?.call(context, {
								fallback,
								page,
								slugs,
								lang
							});
							if (res !== void 0) fallback = res;
						}
						return fallback;
					}
				});
				fns.createPage({
					render: defaultRenderMode,
					path: "/[lang]/404",
					staticPaths: context.i18nConfig.languages,
					component: layouts.notFound
				});
				if (context.mode !== "static") fns.createPage({
					render: "dynamic",
					path: "/404",
					component: () => unstable_redirect(`/${context.i18nConfig.defaultLanguage}`)
				});
			} else {
				fns.createRoot({
					render: defaultRenderMode,
					component: layouts.root
				});
				fns.createPage({
					render: defaultRenderMode,
					path: "/[...slugs]",
					staticPaths,
					async component({ slugs }) {
						const page = await resolvePage(slugs);
						let fallback = /* @__PURE__ */ jsx(layouts.page, {
							slugs,
							page
						});
						for (const plugin of context.plugins) {
							const res = await plugin.renderPage?.call(context, {
								fallback,
								page,
								slugs
							});
							if (res !== void 0) fallback = res;
						}
						return fallback;
					}
				});
				fns.createPage({
					render: defaultRenderMode,
					staticPaths: [],
					path: "/404",
					component: layouts.notFound
				});
			}
			return null;
		}, createPagesOptions);
		return result;
	}
	function pluginsMiddleware(opts) {
		async function init() {
			const out = [];
			const resolved = await Promise.all(context.plugins.map((plugin) => plugin.createMiddlewares?.call(context, opts)));
			for (const v of resolved) if (v) out.push(...v);
			return out;
		}
		const middlewaresPromise = init();
		return async (c, next) => {
			const middlewares = await middlewaresPromise;
			if (middlewares.length === 0) return next();
			let response;
			const run = async (index) => {
				const handler = middlewares[index];
				if (handler) {
					const result = await handler(c, () => run(index + 1));
					if (result && !response) response = result;
				} else await next();
			};
			await run(0);
			return response;
		};
	}
	function patchAdapter(adapter) {
		return (handlers, options) => {
			let entry = adapter(handlers, options);
			for (const plugin of context.plugins) if (plugin.unstable_onServerEntry) entry = plugin.unstable_onServerEntry(entry);
			return entry;
		};
	}
	return {
		createPages: createPages$2,
		patchAdapter,
		createMiddlewares() {
			return [pluginsMiddleware];
		}
	};
}
//#endregion
export { createRouter, notFound, redirect };
