// fumadocs-mdx/dist/runtime/dynamic.js
import { i as buildConfig, n as createCore } from "../core-Cq6V99iz.js";
import { t as buildMDX } from "../build-BnKCFeb6.js";
import { server } from "./server.js";
import { frontmatter } from "fumadocs-core/content/md/frontmatter";
import fs from "node:fs/promises";
import { pathToFileURL } from "node:url";
import jsxRuntimeDefault from "react/jsx-runtime";
//#region src/runtime/dynamic.ts
const AsyncFunction = Object.getPrototypeOf(executeMdx).constructor;
/** we always assume the input code is safe and execute it directly */
async function executeMdx(compiled, options = {}) {
	const { opts: scopeOpts, ...scope } = options.scope ?? {};
	const fullScope = {
		opts: {
			...scopeOpts,
			...options.jsxRuntime ?? jsxRuntimeDefault,
			baseUrl: options.baseUrl
		},
		...scope
	};
	const hydrateFn = new AsyncFunction(...Object.keys(fullScope), compiled);
	return await hydrateFn.apply(hydrateFn, Object.values(fullScope));
}
async function dynamic(configExports, coreOptions, serverOptions) {
	const core = createCore(coreOptions);
	await core.init({ config: buildConfig(configExports, process.cwd()) });
	const create = server(serverOptions);
	function getDocCollection(name) {
		const collection = core.getCollection(name);
		if (!collection) return;
		if (collection.type === "docs") return collection.docs;
		else if (collection.type === "doc") return collection;
	}
	function convertLazyEntries(collection, entries) {
		const head = {};
		const body = {};
		async function compile({ info, data }) {
			let content = await fs.readFile(info.fullPath, "utf-8");
			content = frontmatter(content).content;
			return await executeMdx((await buildMDX(core, collection, {
				filePath: info.fullPath,
				source: content,
				frontmatter: data,
				isDevelopment: false,
				environment: "runtime"
			})).code, { baseUrl: pathToFileURL(info.fullPath) });
		}
		for (const entry of entries) {
			head[entry.info.path] = () => entry.data;
			let cachedResult;
			body[entry.info.path] = () => cachedResult ??= compile(entry);
		}
		return {
			head,
			body
		};
	}
	return {
		async doc(name, base, entries) {
			const collection = getDocCollection(name);
			if (!collection) throw new Error(`the doc collection ${name} doesn't exist.`);
			const { head, body } = convertLazyEntries(collection, entries);
			return create.docLazy(name, base, head, body);
		},
		async docs(name, base, meta, entries) {
			const collection = getDocCollection(name);
			if (!collection) throw new Error(`the doc collection ${name} doesn't exist.`);
			const docs = convertLazyEntries(collection, entries);
			return create.docsLazy(name, base, meta, docs.head, docs.body);
		}
	};
}
//#endregion
export { dynamic };
