// waku/dist/router/client.js
'use client';
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { Component, createContext, use, useCallback, useContext, useEffect, useLayoutEffect, useMemo, useRef, useState, useTransition } from 'react';
import { preloadModule } from 'react-dom';
import { getErrorInfo } from '../lib/utils/custom-errors.js';
import { addBase, removeBase } from '../lib/utils/path.js';
import { Root, Slot, unstable_prefetchRsc as prefetchRsc, unstable_registerCallServerElementsListener as registerCallServerElementsListener, unstable_registerFetchEnhancer as registerFetchEnhancer, useElementsPromise_UNSTABLE as useElementsPromise, useRefetch } from '../minimal/client.js';
import { buildRouteHref } from './isomorphic-utils/build-route-href.js';
import { matchRouteParams } from './isomorphic-utils/match-route-params.js';
import { ETAG_ID_PREFIX, HAS404_ID, IS_STATIC_ID, ROUTE_ID, SKIP_HEADER, STATIC_ETAG, encodeRoutePath, encodeSliceId, pathnameToRoutePath } from './isomorphic-utils/route-path.js';
import { getRouteSearchCodecId, isCodec } from './isomorphic-utils/search-codec-registry.js';
const pathnameToCurrentRoutePath = (pathname)=>pathnameToRoutePath(removeBase(pathname, import.meta.env.WAKU_CONFIG_BASE_PATH));
const parseRoute = (url)=>{
    const { pathname, searchParams, hash } = url;
    return {
        path: pathnameToCurrentRoutePath(pathname),
        query: searchParams.toString(),
        hash
    };
};
const getRouteUrl = (route)=>{
    const nextUrl = new URL(window.location.href);
    nextUrl.pathname = route.path;
    nextUrl.search = route.query;
    nextUrl.hash = route.hash;
    return nextUrl;
};
const parseRouteFromLocation = ()=>{
    return parseRoute(new URL(window.location.href));
};
const getRouteFromElements = (elements)=>{
    const routeData = elements[ROUTE_ID];
    if (routeData) {
        const [path, query] = routeData;
        return {
            path,
            query,
            hash: ''
        };
    }
    return undefined;
};
const isMetaKey = (key)=>key === ROUTE_ID || key === HAS404_ID || key.startsWith(IS_STATIC_ID) || key.startsWith(ETAG_ID_PREFIX);
const shouldScrollByDefault = (url)=>pathnameToCurrentRoutePath(url.pathname) !== pathnameToCurrentRoutePath(window.location.pathname) || url.hash !== window.location.hash;
const isPathChange = (next, prev)=>next.path !== prev.path;
const isHashChange = (next, prev)=>next.hash !== prev.hash;
const isSameRoute = (next, prev)=>next.path === prev.path && next.query === prev.query && next.hash === prev.hash;
const shouldScrollForRouteChange = (next, prev)=>isPathChange(next, prev) || isHashChange(next, prev);
const isAltClick = (event)=>event.button !== 0 || !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
let savedRscParams;
const createRscParams = (query)=>{
    if (savedRscParams && savedRscParams[0] === query) {
        return savedRscParams[1];
    }
    const rscParams = new URLSearchParams({
        query
    });
    savedRscParams = [
        query,
        rscParams
    ];
    return rscParams;
};
const createSkipHeaderEnhancer = (cachedEtagsRef)=>(fetchFn)=>(input, init = {})=>{
            const skipStr = JSON.stringify(cachedEtagsRef.current);
            const headers = init.headers ||= {};
            if (Array.isArray(headers)) {
                headers.push([
                    SKIP_HEADER,
                    skipStr
                ]);
            } else if (headers instanceof Headers) {
                headers.set(SKIP_HEADER, skipStr);
            } else {
                headers[SKIP_HEADER] = skipStr;
            }
            return fetchFn(input, init);
        };
const createRouteChangeListeners = ()=>{
    const listeners = {
        start: new Set(),
        complete: new Set()
    };
    const emit = (event, route)=>{
        const eventListenersSet = listeners[event];
        if (!eventListenersSet.size) {
            return;
        }
        for (const listener of eventListenersSet){
            listener(route);
        }
    };
    return [
        {
            on: (event, handler)=>{
                listeners[event].add(handler);
            },
            off: (event, handler)=>{
                listeners[event].delete(handler);
            }
        },
        emit
    ];
};
// This is an internal thing, not a public API
const RouterContext = /*#__PURE__*/ createContext(null);
const SearchCodecsContext = /*#__PURE__*/ createContext(new Map());
const useResolveSearchCodec = ()=>{
    const codecs = useContext(SearchCodecsContext);
    return useCallback((routePath)=>{
        const id = getRouteSearchCodecId(routePath);
        return id !== undefined ? codecs.get(id) : undefined;
    }, [
        codecs
    ]);
};
export function useRouter() {
    const router = useContext(RouterContext);
    if (!router) {
        throw new Error('Missing Router');
    }
    const { route, changeRoute, prefetchRoute } = router;
    const resolveCodec = useResolveSearchCodec();
    const push = useCallback(async (to, options)=>{
        const href = typeof to === 'string' ? to : buildRouteHref(to, resolveCodec);
        const url = new URL(addBase(href, import.meta.env.WAKU_CONFIG_BASE_PATH), window.location.href);
        await changeRoute(parseRoute(url), {
            shouldScroll: options?.scroll ?? shouldScrollByDefault(url),
            mode: 'push',
            url,
            instant: options?.unstable_instant
        });
    }, [
        changeRoute,
        resolveCodec
    ]);
    const replace = useCallback(async (to, options)=>{
        const href = typeof to === 'string' ? to : buildRouteHref(to, resolveCodec);
        const url = new URL(addBase(href, import.meta.env.WAKU_CONFIG_BASE_PATH), window.location.href);
        await changeRoute(parseRoute(url), {
            shouldScroll: options?.scroll ?? shouldScrollByDefault(url),
            mode: 'replace',
            url,
            instant: options?.unstable_instant
        });
    }, [
        changeRoute,
        resolveCodec
    ]);
    const reload = useCallback(async ()=>{
        const url = new URL(window.location.href);
        await changeRoute(parseRoute(url), {
            shouldScroll: true,
            refetch: true
        });
    }, [
        changeRoute
    ]);
    const back = useCallback(()=>{
        // FIXME is this correct?
        window.history.back();
    }, []);
    const forward = useCallback(()=>{
        // FIXME is this correct?
        window.history.forward();
    }, []);
    const prefetch = useCallback((to)=>{
        const href = typeof to === 'string' ? to : buildRouteHref(to, resolveCodec);
        const url = new URL(addBase(href, import.meta.env.WAKU_CONFIG_BASE_PATH), window.location.href);
        prefetchRoute(parseRoute(url));
    }, [
        prefetchRoute,
        resolveCodec
    ]);
    return {
        ...route,
        push,
        replace,
        reload,
        back,
        forward,
        prefetch,
        unstable_events: router.routeChangeEvents
    };
}
/**
 * Read the current route's params, typed from the `from` path, or null when
 * the current path does not match it. Re-renders when the route path changes.
 * The result is memoized by path, so its identity changes on navigation to a
 * different path; read its fields rather than using the object itself as an
 * effect dependency.
 */ export function useParams_UNSTABLE({ from }) {
    const { path } = useRouter();
    return useMemo(()=>matchRouteParams(from, path), [
        from,
        path
    ]);
}
/**
 * Provide search codecs to `useSearch_UNSTABLE`, `useSetSearch_UNSTABLE`,
 * `push`, and `Link`. Render it in your root layout so the codecs are present in
 * both the SSR render and the browser. Pass only search codecs: a codec-only
 * module (via `import * as`), a record, or an array. A value that is not a codec
 * is ignored with a warning, so keep helpers and constants out of the module you
 * pass (or list the codecs explicitly).
 */ export function Unstable_SearchCodecsProvider({ searchCodecs, children }) {
    const codecs = useMemo(()=>{
        const map = new Map();
        const values = Array.isArray(searchCodecs) ? searchCodecs : Object.values(searchCodecs);
        for (const value of values){
            if (!isCodec(value)) {
                console.warn('Unstable_SearchCodecsProvider ignored a value that is not a search codec; pass only codecs (a codec-only module or an explicit array).', value);
                continue;
            }
            const existing = map.get(value.id);
            if (existing && existing !== value) {
                throw new Error(`Duplicate search codec id: "${value.id}"`);
            }
            map.set(value.id, value);
        }
        return map;
    }, [
        searchCodecs
    ]);
    return /*#__PURE__*/ _jsx(SearchCodecsContext, {
        value: codecs,
        children: children
    });
}
/**
 * Read the current route's typed `search`, parsed client-side with the route's
 * codec (provided via `Unstable_SearchCodecsProvider`), or null when the current
 * path does not match `from` or the route has no codec. Re-renders when the
 * query changes.
 */ export function useSearch_UNSTABLE({ from }) {
    const { path, query } = useRouter();
    const codecs = useContext(SearchCodecsContext);
    return useMemo(()=>{
        if (matchRouteParams(from, path) === null) {
            return null;
        }
        const codecId = getRouteSearchCodecId(from);
        const codec = codecId !== undefined ? codecs.get(codecId) : undefined;
        return codec ? codec.parse(query) : null;
    }, [
        from,
        path,
        query,
        codecs
    ]);
}
/**
 * Returns a setter for the current route's `search`, serialized client-side with
 * the route's codec (provided via `Unstable_SearchCodecsProvider`). Accepts a
 * partial or an updater of the current search and navigates (push by default, or
 * replace) to the same path. A no-op when the current path does not match `from`
 * or has no codec.
 */ export function useSetSearch_UNSTABLE({ from }) {
    const router = useContext(RouterContext);
    if (!router) {
        throw new Error('Missing Router');
    }
    const { route, changeRoute } = router;
    const codecs = useContext(SearchCodecsContext);
    return useCallback(async (update, options)=>{
        if (matchRouteParams(from, route.path) === null) {
            return;
        }
        const codecId = getRouteSearchCodecId(from);
        const codec = codecId !== undefined ? codecs.get(codecId) : undefined;
        if (!codec) {
            return;
        }
        const prev = codec.parse(route.query);
        const partial = typeof update === 'function' ? update(prev) : update;
        const nextQuery = codec.serialize({
            ...prev,
            ...partial
        });
        const url = new URL(window.location.href);
        url.search = nextQuery;
        await changeRoute(parseRoute(url), {
            shouldScroll: options?.scroll ?? false,
            mode: options?.history ?? 'push',
            url
        });
    }, [
        from,
        route.path,
        route.query,
        codecs,
        changeRoute
    ]);
}
function useSharedRef(ref) {
    const managedRef = useRef(null);
    const handleRef = useCallback(// eslint-disable-next-line react-hooks/immutability
    (node)=>{
        managedRef.current = node;
        const isRefCallback = typeof ref === 'function';
        let cleanup;
        if (isRefCallback) {
            cleanup = ref(node);
        } else if (ref) {
            // TODO is this a false positive?
            // eslint-disable-next-line react-hooks/immutability
            ref.current = node;
        }
        return ()=>{
            managedRef.current = null;
            if (isRefCallback) {
                if (cleanup) {
                    cleanup();
                } else {
                    ref(null);
                }
            } else if (ref) {
                ref.current = null;
            }
        };
    }, [
        ref
    ]);
    return [
        managedRef,
        handleRef
    ];
}
const NavigationStatusContext = /*#__PURE__*/ createContext({});
/**
 * Returns the navigation status of the enclosing `Link`, like React's
 * `useFormStatus`. `pending` is `true` while the navigation transition is in
 * flight, until the destination route's async components resolve. Returns an
 * empty object when called outside a `Link`.
 */ export const useNavigationStatus_UNSTABLE = ()=>useContext(NavigationStatusContext);
export function Link({ to, children, scroll, unstable_instant, unstable_prefetchOnEnter, unstable_prefetchOnView, unstable_startTransition, ref: refProp, ...props }) {
    const resolveCodec = useResolveSearchCodec();
    const href = typeof to === 'string' ? to : buildRouteHref(to, resolveCodec);
    const resolvedTo = addBase(href, import.meta.env.WAKU_CONFIG_BASE_PATH);
    const router = useContext(RouterContext);
    const changeRoute = router ? router.changeRoute : ()=>{
        throw new Error('Missing Router');
    };
    const prefetchRoute = router ? router.prefetchRoute : ()=>{
        throw new Error('Missing Router');
    };
    const [isPending, startTransition] = useTransition();
    const startTransitionFn = unstable_startTransition || startTransition;
    const [ref, setRef] = useSharedRef(refProp);
    useEffect(()=>{
        if (!unstable_prefetchOnView || !ref.current) {
            return;
        }
        const observer = new IntersectionObserver((entries)=>{
            entries.forEach((entry)=>{
                if (entry.isIntersecting) {
                    const url = new URL(resolvedTo, window.location.href);
                    if (router && url.href !== window.location.href) {
                        router.prefetchRoute(parseRoute(url));
                    }
                }
            });
        }, {
            threshold: 0.1
        });
        observer.observe(ref.current);
        return ()=>{
            observer.disconnect();
        };
    }, [
        unstable_prefetchOnView,
        router,
        resolvedTo,
        ref
    ]);
    const internalOnClick = ()=>{
        const url = new URL(resolvedTo, window.location.href);
        if (url.href !== window.location.href) {
            const route = parseRoute(url);
            prefetchRoute(route);
            if (unstable_instant) {
                changeRoute(route, {
                    shouldScroll: scroll ?? shouldScrollByDefault(url),
                    mode: 'push',
                    url,
                    instant: true
                }).catch(()=>{});
            } else {
                startTransitionFn(async ()=>{
                    await changeRoute(route, {
                        shouldScroll: scroll ?? shouldScrollByDefault(url),
                        mode: 'push',
                        url,
                        startTransition: startTransitionFn
                    });
                });
            }
        } else if (url.hash && scroll !== false) {
            scrollToRoute(parseRoute(url), 'auto', false);
        }
    };
    const onClick = (event)=>{
        props.onClick?.(event);
        if (event.defaultPrevented || isAltClick(event)) {
            return;
        }
        if (props.target && props.target.toLowerCase() !== '_self') {
            console.warn('[Link] `target` is discouraged. Use `<a>` for this case.');
        }
        if (props.download !== undefined && props.download !== null && props.download !== false) {
            console.warn('[Link] `download` is discouraged. Use `<a>` for this case.');
        }
        event.preventDefault();
        internalOnClick();
    };
    const onMouseEnter = unstable_prefetchOnEnter ? (event)=>{
        const url = new URL(resolvedTo, window.location.href);
        if (url.href !== window.location.href) {
            prefetchRoute(parseRoute(url));
        }
        props.onMouseEnter?.(event);
    } : props.onMouseEnter;
    const navigationStatus = useMemo(()=>({
            pending: isPending
        }), [
        isPending
    ]);
    return /*#__PURE__*/ _jsx(NavigationStatusContext, {
        value: navigationStatus,
        children: /*#__PURE__*/ _jsx("a", {
            ...props,
            href: resolvedTo,
            onClick: onClick,
            onMouseEnter: onMouseEnter,
            ref: setRef,
            children: children
        })
    });
}
const notAvailableInServer = (name)=>()=>{
        throw new Error(`${name} is not in the server`);
    };
function renderError(message) {
    return /*#__PURE__*/ _jsxs("html", {
        children: [
            /*#__PURE__*/ _jsx("head", {
                children: /*#__PURE__*/ _jsx("title", {
                    children: "Unhandled Error"
                })
            }),
            /*#__PURE__*/ _jsxs("body", {
                style: {
                    height: '100vh',
                    display: 'flex',
                    flexDirection: 'column',
                    placeContent: 'center',
                    placeItems: 'center',
                    fontSize: '16px',
                    margin: 0
                },
                children: [
                    /*#__PURE__*/ _jsx("h1", {
                        children: "Caught an unexpected error"
                    }),
                    /*#__PURE__*/ _jsxs("p", {
                        children: [
                            "Error: ",
                            message
                        ]
                    })
                ]
            })
        ]
    });
}
export class ErrorBoundary extends Component {
    constructor(props){
        super(props);
        this.state = {};
    }
    static getDerivedStateFromError(error) {
        return {
            error
        };
    }
    render() {
        if ('error' in this.state) {
            if (this.state.error instanceof Error) {
                return renderError(this.state.error.message);
            }
            return renderError(String(this.state.error));
        }
        return this.props.children;
    }
}
const NotFound = ({ error, has404, reset, handledErrorSet })=>{
    const router = useContext(RouterContext);
    if (!router) {
        throw new Error('Missing Router');
    }
    const { changeRoute } = router;
    useEffect(()=>{
        if (has404) {
            if (handledErrorSet.has(error)) {
                return;
            }
            handledErrorSet.add(error);
            const url = new URL('/404', window.location.href);
            changeRoute(parseRoute(url), {
                shouldScroll: true
            }).then(()=>{
                reset();
            }).catch((err)=>{
                console.log('Error while navigating to 404:', err);
            });
        }
    }, [
        error,
        has404,
        reset,
        changeRoute,
        handledErrorSet
    ]);
    return has404 ? null : /*#__PURE__*/ _jsx("h1", {
        children: "Not Found"
    });
};
const Redirect = ({ error, to, reset, handledErrorSet })=>{
    const router = useContext(RouterContext);
    if (!router) {
        throw new Error('Missing Router');
    }
    const { changeRoute } = router;
    useEffect(()=>{
        // ensure single re-fetch per server redirection error on StrictMode
        // https://github.com/wakujs/waku/pull/1512
        if (handledErrorSet.has(error)) {
            return;
        }
        handledErrorSet.add(error);
        const url = new URL(to, window.location.href);
        if (url.protocol !== 'http:' && url.protocol !== 'https:') {
            return;
        }
        if (url.hostname !== window.location.hostname) {
            window.location.replace(url.href);
            return;
        }
        const currentPath = window.location.pathname;
        const newPath = url.pathname !== currentPath;
        const historyUrl = url.origin === window.location.origin ? url : undefined;
        changeRoute(parseRoute(url), {
            shouldScroll: newPath,
            mode: 'replace',
            url: historyUrl
        }).then(()=>{
            handledErrorSet.delete(error);
            // FIXME: As we understand it, we should have a proper solution.
            setTimeout(()=>{
                reset();
            }, 1);
        }).catch((err)=>{
            handledErrorSet.delete(error);
            console.log('Error while navigating to redirect:', err);
        });
    }, [
        error,
        to,
        reset,
        changeRoute,
        handledErrorSet
    ]);
    return null;
};
class CustomErrorHandler extends Component {
    handledErrorSet = new WeakSet();
    constructor(props){
        super(props);
        this.state = {
            error: null
        };
        this.reset = this.reset.bind(this);
    }
    static getDerivedStateFromError(error) {
        return {
            error
        };
    }
    reset() {
        this.setState({
            error: null
        });
    }
    render() {
        const { error } = this.state;
        if (error !== null) {
            const info = getErrorInfo(error);
            if (info?.status === 404) {
                return /*#__PURE__*/ _jsx(NotFound, {
                    error: error,
                    has404: this.props.has404,
                    reset: this.reset,
                    handledErrorSet: this.handledErrorSet
                });
            }
            if (info?.location) {
                return /*#__PURE__*/ _jsx(Redirect, {
                    error: error,
                    to: info.location,
                    reset: this.reset,
                    handledErrorSet: this.handledErrorSet
                });
            }
            throw error;
        }
        return this.props.children;
    }
}
const ThrowError = ({ error })=>{
    throw error;
};
const getRouteSlotId = (path)=>'route:' + path;
const getSliceSlotId = (id)=>'slice:' + id;
export function Slice({ id, children, ...props }) {
    const router = useContext(RouterContext);
    if (!router) {
        throw new Error('Missing Router');
    }
    const { fetchingSlices } = router;
    const refetch = useRefetch();
    const slotId = getSliceSlotId(id);
    const elementsPromise = useElementsPromise();
    const elements = use(elementsPromise);
    const needsToFetchSlice = props.lazy && (!(slotId in elements) || elements[ETAG_ID_PREFIX + slotId] !== STATIC_ETAG);
    useEffect(()=>{
        // FIXME this works because of subtle timing behavior.
        if (needsToFetchSlice && !fetchingSlices.has(id)) {
            fetchingSlices.add(id);
            const rscPath = encodeSliceId(id);
            refetch(rscPath).catch((e)=>{
                console.error('Failed to fetch slice:', e);
            }).finally(()=>{
                fetchingSlices.delete(id);
            });
        }
    }, [
        fetchingSlices,
        refetch,
        id,
        needsToFetchSlice
    ]);
    if (props.lazy && !(slotId in elements)) {
        // FIXME the fallback doesn't show on refetch after the first one.
        return props.fallback;
    }
    return /*#__PURE__*/ _jsx(Slot, {
        id: slotId,
        children: children
    });
}
const getHashElement = (hash)=>{
    const raw = hash.slice(1);
    const rawElement = document.getElementById(raw);
    if (rawElement) {
        return rawElement;
    }
    try {
        return document.getElementById(decodeURIComponent(raw));
    } catch  {
        return null;
    }
};
const scrollToRoute = (route, behavior, scrollTopForMissingHash)=>{
    if (route.hash) {
        const element = getHashElement(route.hash);
        if (!element) {
            if (!scrollTopForMissingHash) {
                return;
            }
            window.scrollTo({
                left: 0,
                top: 0,
                behavior
            });
            return;
        }
        const scrollMarginTop = Number.parseFloat(window.getComputedStyle(element).scrollMarginTop) || 0;
        window.scrollTo({
            left: 0,
            top: element.getBoundingClientRect().top + window.scrollY - scrollMarginTop,
            behavior
        });
        return;
    }
    window.scrollTo({
        left: 0,
        top: 0,
        behavior
    });
};
const writeUrlToHistory = (mode, url)=>{
    if (mode === 'push') {
        window.history.pushState(window.history.state, '', url);
    } else {
        window.history.replaceState(window.history.state, '', url);
    }
};
const useElementsMetadata = (elementsPromise)=>{
    const [has404, setHas404] = useState(false);
    const staticPathSetRef = useRef(new Set());
    const cachedEtagsRef = useRef({});
    useEffect(()=>{
        elementsPromise.then((elements)=>{
            // Under `unstable_instant` the eager merge serves static etags and
            // ROUTE_ID from the previous bag; a dynamic slot's etag stays a hole,
            // so it is skipped here until it resolves rather than going stale.
            const { [ROUTE_ID]: routeData, [IS_STATIC_ID]: isStatic, [HAS404_ID]: has404FromElements } = elements;
            if (has404FromElements) {
                setHas404(true);
            }
            if (routeData) {
                const [path, _query] = routeData;
                if (isStatic) {
                    staticPathSetRef.current.add(path);
                }
            }
            const etags = {};
            for (const [key, value] of Object.entries(elements)){
                // Keep the static sentinel; for string tags drop empty (clear
                // signal) and non-Latin1 (breaks fetch).
                if (key.startsWith(ETAG_ID_PREFIX) && (value === STATIC_ETAG || typeof value === 'string' && /^[\u0020-\u00ff]+$/.test(value))) {
                    etags[key.slice(ETAG_ID_PREFIX.length)] = value;
                }
            }
            cachedEtagsRef.current = etags;
        }, ()=>{});
    }, [
        elementsPromise
    ]);
    return {
        has404,
        staticPathSetRef,
        cachedEtagsRef
    };
};
const defaultRouteInterceptor = (route)=>route;
const InnerRouter = ({ fallbackRoute, routeInterceptor = defaultRouteInterceptor })=>{
    const elementsPromise = useElementsPromise();
    const elements = use(elementsPromise);
    const routeFromElements = getRouteFromElements(elements);
    const resolvedRoute = routeFromElements && routeFromElements.path !== fallbackRoute.path ? {
        ...routeFromElements,
        hash: fallbackRoute.hash
    } : fallbackRoute;
    const initialRouteRef = useRef(resolvedRoute);
    if (import.meta.hot) {
        const refetchRoute = ()=>{
            staticPathSetRef.current.clear();
            cachedEtagsRef.current = {};
            const rscPath = encodeRoutePath(route.path);
            const rscParams = createRscParams(route.query);
            // eslint-disable-next-line @typescript-eslint/no-floating-promises
            refetch(rscPath, rscParams);
        };
        globalThis.__WAKU_RSC_RELOAD_LISTENERS__ ||= [];
        const index = globalThis.__WAKU_RSC_RELOAD_LISTENERS__.indexOf(globalThis.__WAKU_REFETCH_ROUTE__);
        if (index !== -1) {
            globalThis.__WAKU_RSC_RELOAD_LISTENERS__.splice(index, 1, refetchRoute);
        } else {
            globalThis.__WAKU_RSC_RELOAD_LISTENERS__.unshift(refetchRoute);
        }
        globalThis.__WAKU_REFETCH_ROUTE__ = refetchRoute;
    }
    const { has404, staticPathSetRef, cachedEtagsRef } = useElementsMetadata(elementsPromise);
    // FIXME this "fetchingSlices" hack feels suboptimal.
    const fetchingSlicesRef = useRef(new Set());
    const refetch = useRefetch();
    const [route, setRoute] = useState(()=>({
            // This is the first initialization of the route, and it has
            // to ignore the hash, because on server side there is none.
            // Otherwise there will be a hydration error.
            // The client side route, including the hash, will be updated in the effect below.
            ...initialRouteRef.current,
            hash: ''
        }));
    const routeChangeListenersRef = useRef(null);
    if (routeChangeListenersRef.current === null) {
        routeChangeListenersRef.current = createRouteChangeListeners();
    }
    // Update the route post-load to include the current hash.
    const routeRef = useRef(route);
    useEffect(()=>{
        const route = {
            ...initialRouteRef.current,
            hash: window.location.hash || initialRouteRef.current.hash
        };
        routeRef.current = route;
        setRoute((prev)=>isSameRoute(prev, route) ? prev : route);
        setErr(null);
        setPendingScroll(null);
        setPendingHistory(null);
    }, []);
    const [err, setErr] = useState(null);
    const [pendingHistory, setPendingHistory] = useState(null);
    useLayoutEffect(()=>{
        if (pendingHistory) {
            const { mode, url } = pendingHistory;
            const urlToWrite = url || getRouteUrl(route);
            writeUrlToHistory(mode, urlToWrite);
        }
    }, [
        route,
        pendingHistory
    ]);
    const [pendingScroll, setPendingScroll] = useState(null);
    useLayoutEffect(()=>{
        if (pendingScroll) {
            const { pathChanged } = pendingScroll;
            const scrollBehavior = pathChanged ? 'instant' : 'auto';
            scrollToRoute(route, scrollBehavior, pathChanged);
        }
    }, [
        route,
        pendingScroll
    ]);
    // TODO(daishi): consider combining three or four useState hooks above.
    const [routeChangeEvents, emitRouteChangeEvent] = routeChangeListenersRef.current;
    const routeChangeAbortRef = useRef(null);
    const changeRoute = useCallback(async (nextRoute, options)=>{
        routeChangeAbortRef.current?.abort();
        const abortController = new AbortController();
        routeChangeAbortRef.current = abortController;
        const isAborted = ()=>abortController.signal.aborted;
        emitRouteChangeEvent('start', nextRoute);
        const startTransitionFn = options.startTransition || ((fn)=>fn());
        const prevPathname = window.location.pathname;
        let { mode, url } = options;
        const routeBeforeChange = routeRef.current;
        const shouldRefetch = options.refetch ?? !isSameRoute(nextRoute, routeBeforeChange);
        const pathChanged = isPathChange(nextRoute, routeBeforeChange);
        const isStaticSlot = (key)=>cachedEtagsRef.current[key] === STATIC_ETAG;
        const commitRoute = (route)=>{
            routeRef.current = route;
            setRoute(route);
            setErr(null);
            setPendingScroll(options.shouldScroll ? {
                pathChanged
            } : null);
            setPendingHistory(mode ? {
                mode,
                url
            } : null);
        };
        const getRedirect = (elements)=>{
            const serverRoute = getRouteFromElements(elements);
            const isStatic = elements[IS_STATIC_ID];
            if (serverRoute && (serverRoute.path !== nextRoute.path || !isStatic && serverRoute.query !== nextRoute.query)) {
                return serverRoute;
            }
            return undefined;
        };
        const commitInTransition = ()=>{
            if (isAborted()) {
                return;
            }
            startTransitionFn(()=>{
                commitRoute(nextRoute);
                routeChangeAbortRef.current = null;
                emitRouteChangeEvent('complete', nextRoute);
            });
        };
        if (staticPathSetRef.current.has(nextRoute.path) || !shouldRefetch) {
            commitInTransition();
            return;
        }
        const rscPath = encodeRoutePath(nextRoute.path);
        const rscParams = createRscParams(nextRoute.query);
        const targetUrl = url || getRouteUrl(nextRoute);
        const onBuildIdMismatch = ()=>{
            window.history.pushState(window.history.state, '', targetUrl);
            window.location.reload();
        };
        if (options.instant && isStaticSlot(getRouteSlotId(nextRoute.path))) {
            const isEager = (key)=>key.startsWith(ETAG_ID_PREFIX) ? isStaticSlot(key.slice(ETAG_ID_PREFIX.length)) : isMetaKey(key) || isStaticSlot(key);
            const dataPromise = refetch(rscPath, rscParams, {
                signal: abortController.signal,
                unstable_isEager: isEager,
                onBuildIdMismatch
            });
            commitRoute(nextRoute);
            return dataPromise.then((elements)=>{
                if (isAborted()) {
                    return;
                }
                routeChangeAbortRef.current = null;
                const redirect = getRedirect(elements);
                if (redirect) {
                    routeRef.current = redirect;
                    setRoute(redirect);
                    if (redirect.path !== '/404') {
                        setPendingHistory({
                            mode: 'replace',
                            url: getRouteUrl(redirect)
                        });
                    }
                }
                emitRouteChangeEvent('complete', redirect ?? nextRoute);
            }, (e)=>{
                if (isAborted()) {
                    return;
                }
                routeChangeAbortRef.current = null;
                setErr(e);
                throw e;
            });
        }
        try {
            const elements = await refetch(rscPath, rscParams, {
                signal: abortController.signal,
                onBuildIdMismatch
            });
            const redirect = getRedirect(elements);
            if (redirect) {
                nextRoute = redirect;
                if (mode) {
                    mode = redirect.path === '/404' ? undefined : 'push';
                    url = undefined;
                }
            }
        } catch (e) {
            if (isAborted()) {
                return;
            }
            routeChangeAbortRef.current = null;
            // Write URL synchronously
            // React may rollback transition state updates when the render throws
            if (mode && window.location.pathname === prevPathname) {
                writeUrlToHistory(mode, targetUrl);
            }
            setErr(e);
            throw e;
        }
        commitInTransition();
    }, [
        emitRouteChangeEvent,
        refetch,
        staticPathSetRef,
        cachedEtagsRef
    ]);
    const applyChangeRouteData = useCallback(async (routeData, isStatic)=>{
        if (!routeData) {
            return;
        }
        const [path, query] = routeData;
        const currentRoute = routeRef.current;
        if (currentRoute.path === path && (isStatic || currentRoute.query === query)) {
            return;
        }
        const url = new URL(window.location.href);
        url.pathname = path;
        url.search = query;
        url.hash = '';
        await changeRoute(parseRoute(url), {
            refetch: false,
            shouldScroll: false,
            mode: path === '/404' ? undefined : 'push',
            url
        });
    }, [
        changeRoute
    ]);
    useEffect(()=>registerFetchEnhancer(createSkipHeaderEnhancer(cachedEtagsRef)), [
        cachedEtagsRef
    ]);
    useEffect(()=>{
        const listener = (elements)=>{
            const { [ROUTE_ID]: routeData, [IS_STATIC_ID]: isStatic } = elements;
            applyChangeRouteData(routeData, isStatic).catch((err)=>{
                console.log('Error while handling route updates:', err);
            });
        };
        return registerCallServerElementsListener(listener);
    }, [
        applyChangeRouteData
    ]);
    const prefetchRoute = useCallback((route)=>{
        if (staticPathSetRef.current.has(route.path)) {
            return;
        }
        const rscPath = encodeRoutePath(route.path);
        const rscParams = createRscParams(route.query);
        prefetchRsc(rscPath, rscParams);
        globalThis.__WAKU_ROUTER_PREFETCH__?.(route.path, (id)=>{
            preloadModule(id, {
                as: 'script'
            });
        });
    }, [
        staticPathSetRef
    ]);
    useEffect(()=>{
        const callback = ()=>{
            const nextRoute = routeInterceptor(parseRoute(new URL(window.location.href)));
            if (!nextRoute) {
                return;
            }
            changeRoute(nextRoute, {
                shouldScroll: shouldScrollForRouteChange(nextRoute, routeRef.current)
            }).catch((err)=>{
                console.log('Error while navigating back:', err);
            });
        };
        window.addEventListener('popstate', callback);
        return ()=>{
            window.removeEventListener('popstate', callback);
        };
    }, [
        changeRoute,
        routeInterceptor
    ]);
    const routeElement = err !== null ? /*#__PURE__*/ _jsx(ThrowError, {
        error: err
    }) : /*#__PURE__*/ _jsx(Slot, {
        id: getRouteSlotId(route.path)
    });
    const rootElement = /*#__PURE__*/ _jsx(Slot, {
        id: "root",
        children: /*#__PURE__*/ _jsx(CustomErrorHandler, {
            has404: has404,
            children: routeElement
        })
    });
    return /*#__PURE__*/ _jsx(RouterContext, {
        value: {
            route,
            changeRoute,
            prefetchRoute,
            routeChangeEvents,
            fetchingSlices: fetchingSlicesRef.current
        },
        children: rootElement
    });
};
export function Router({ initialRoute = parseRouteFromLocation(), unstable_routeInterceptor }) {
    const initialRscPath = encodeRoutePath(initialRoute.path);
    const initialRscParams = createRscParams(initialRoute.query);
    return /*#__PURE__*/ _jsx(Root, {
        initialRscPath: initialRscPath,
        initialRscParams: initialRscParams,
        children: /*#__PURE__*/ _jsx(InnerRouter, {
            fallbackRoute: initialRoute,
            routeInterceptor: unstable_routeInterceptor
        })
    });
}
const MOCK_ROUTE_CHANGE_LISTENER = {
    on: ()=>notAvailableInServer('routeChange:on'),
    off: ()=>notAvailableInServer('routeChange:off')
};
/**
 * ServerRouter for SSR
 * This is not a public API.
 */ export function INTERNAL_ServerRouter({ route }) {
    const routeElement = /*#__PURE__*/ _jsx(Slot, {
        id: getRouteSlotId(route.path)
    });
    const rootElement = /*#__PURE__*/ _jsx(Slot, {
        id: "root",
        children: routeElement
    });
    return /*#__PURE__*/ _jsx(_Fragment, {
        children: /*#__PURE__*/ _jsx(RouterContext, {
            value: {
                route,
                changeRoute: notAvailableInServer('changeRoute'),
                prefetchRoute: notAvailableInServer('prefetchRoute'),
                routeChangeEvents: MOCK_ROUTE_CHANGE_LISTENER,
                fetchingSlices: new Set()
            },
            children: rootElement
        })
    });
}
export const unstable_HAS404_ID = HAS404_ID;
export const unstable_IS_STATIC_ID = IS_STATIC_ID;
export const unstable_ROUTE_ID = ROUTE_ID;
export const unstable_SKIP_HEADER = SKIP_HEADER;
export const unstable_encodeRoutePath = encodeRoutePath;
export const unstable_encodeSliceId = encodeSliceId;
export const unstable_getRouteSlotId = getRouteSlotId;
export const unstable_getSliceSlotId = getSliceSlotId;
export const unstable_getErrorInfo = getErrorInfo;
export const unstable_addBase = addBase;
export const unstable_removeBase = removeBase;
export const unstable_RouterContext = RouterContext;
export const unstable_parseRoute = parseRoute;
