// waku/dist/router/create-pages.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { createElement } from 'react';
import { getGrouplessPath } from '../lib/utils/create-pages.js';
import { createCustomError } from '../lib/utils/custom-errors.js';
import { countSlugsAndWildcards, getPathMapping, joinPath, parseExactPath, parsePathWithSlug, pathSpecAsString } from '../lib/utils/path.js';
import { Children, Slot } from '../minimal/client.js';
import { ErrorBoundary } from '../router/client.js';
import { unstable_defineRouter } from './define-router.js';
import { pathnameToRoutePath } from './isomorphic-utils/route-path.js';
// https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
export const METHODS = [
    'GET',
    'HEAD',
    'POST',
    'PUT',
    'DELETE',
    'CONNECT',
    'OPTIONS',
    'TRACE',
    'PATCH'
];
export const pathMappingWithoutGroups = (pathSpec, pathname)=>{
    const cleanPathSpec = pathSpec.filter((spec)=>!(spec.type === 'literal' && spec.name.startsWith('(')));
    return getPathMapping(cleanPathSpec, pathname);
};
const sanitizeSlug = (slug)=>slug.replace(/ /g, '-');
const normalizeStaticPaths = (staticPaths)=>staticPaths.map((item)=>(Array.isArray(item) ? item : [
            item
        ]).map(sanitizeSlug));
const assertStaticPathArity = (staticSegments, slugCount, wildcardCount)=>{
    if (staticSegments.length !== slugCount && wildcardCount === 0) {
        throw new Error('staticPaths does not match with slug pattern');
    }
};
const getPageSlotId = (routePath)=>`page:${routePath}`;
const getLayoutSlotId = (layoutIdPath)=>`layout:${layoutIdPath}`;
const parseSearchOrThrow = (codec, query)=>{
    try {
        return codec.parse(query);
    } catch (cause) {
        const err = createCustomError('Bad Request', {
            status: 400
        });
        err.cause = cause;
        throw err;
    }
};
const forEachConcreteStaticPath = (routePathSpec, staticPathsInput, fn)=>{
    const { numSlugs, numWildcards } = countSlugsAndWildcards(routePathSpec);
    for (const staticSegments of normalizeStaticPaths(staticPathsInput)){
        assertStaticPathArity(staticSegments, numSlugs, numWildcards);
        fn(expandStaticRoutePath(routePathSpec, staticSegments));
    }
};
/**
 * Root component for all pages
 * ```tsx
 *   <html>
 *     <head></head>
 *     <body>{children}</body>
 *   </html>
 * ```
 */ const DefaultRoot = ({ children })=>/*#__PURE__*/ _jsx(ErrorBoundary, {
        children: /*#__PURE__*/ _jsxs("html", {
            children: [
                /*#__PURE__*/ _jsx("head", {}),
                /*#__PURE__*/ _jsx("body", {
                    children: children
                })
            ]
        })
    });
const createNestedElements = (elements, children)=>elements.reduceRight((result, element)=>/*#__PURE__*/ createElement(element.component, element.props, result), children);
const routePriorityComparator = (a, b)=>{
    const aPath = a.path;
    const bPath = b.path;
    const aPathLength = aPath.length;
    const bPathLength = bPath.length;
    const aHasWildcard = aPath.at(-1)?.type === 'wildcard';
    const bHasWildcard = bPath.at(-1)?.type === 'wildcard';
    // Special case: root route (length 0) should come before wildcard routes
    // This ensures exact matches like "/" are checked before catch-all routes like "/[...notFound]"
    if (aPathLength === 0 && bHasWildcard) {
        return -1;
    }
    if (bPathLength === 0 && aHasWildcard) {
        return 1;
    }
    // Compare path lengths first (longer paths are more specific)
    if (aPathLength !== bPathLength) {
        return aPathLength > bPathLength ? -1 : 1;
    }
    // If path lengths are equal, literal segments take priority over dynamic segments
    const minLength = Math.min(aPathLength, bPathLength);
    for(let i = 0; i < minLength; i++){
        const aIsLiteral = aPath[i]?.type === 'literal';
        const bIsLiteral = bPath[i]?.type === 'literal';
        if (aIsLiteral !== bIsLiteral) {
            return aIsLiteral ? -1 : 1;
        }
    }
    // If path lengths are equal, compare wildcard presence
    // sort the route without the wildcard higher, to check it earlier
    if (aHasWildcard !== bHasWildcard) {
        return aHasWildcard ? 1 : -1;
    }
    // If all else is equal, routes have the same priority
    return 0;
};
export const createPages = (fn, options)=>{
    let configured = false;
    // layout lookups retain (group) path and pathMaps store without group
    // paths are stored without groups to easily detect duplicates
    const groupedRoutePathByRoutePath = new Map();
    const staticPageEntryByRoutePath = new Map();
    const dynamicPageEntryByRoutePath = new Map();
    const wildcardPageEntryByRoutePath = new Map();
    const dynamicLayoutEntryByRoutePath = new Map();
    const apiEntryByRoutePath = new Map();
    const staticComponentById = new Map();
    const getStaticComponentId = (routePath, kind)=>joinPath(routePath, kind).slice(1);
    const getStaticLayout = (id)=>staticComponentById.get(getStaticComponentId(id, 'layout'))?.component;
    const sliceIdsByRoutePath = new Map();
    const sliceEntryById = new Map();
    let rootItem = undefined;
    const pagePathExists = (path)=>apiEntryByRoutePath.has(path) || staticPageEntryByRoutePath.has(path) || dynamicPageEntryByRoutePath.has(path) || wildcardPageEntryByRoutePath.has(path);
    /** Creates a function to map pathname to component props */ const createPathPropsMapper = (path)=>{
        const layoutMatchPath = groupedRoutePathByRoutePath.get(path) ?? path;
        const routePathSpec = parsePathWithSlug(layoutMatchPath);
        return (pathname)=>pathMappingWithoutGroups(routePathSpec, pathname);
    };
    const createLayoutPropsMapper = (layoutPath)=>{
        const routePathSpec = parsePathWithSlug(layoutPath);
        const numSegments = routePathSpec.filter((segment)=>!(segment.type === 'literal' && segment.name.startsWith('('))).length;
        return (routePath)=>{
            const layoutRoutePath = '/' + routePath.split('/').filter(Boolean).slice(0, numSegments).join('/');
            return pathMappingWithoutGroups(routePathSpec, layoutRoutePath) ?? {};
        };
    };
    const getLayoutIdPath = (layoutPath, routePath)=>{
        const numSegments = parsePathWithSlug(layoutPath).length;
        return '/' + routePath.split('/').filter(Boolean).slice(0, numSegments).join('/');
    };
    /** Builds the routeElement renderer from layouts and page slots */ const buildRouteElement = (layouts, path)=>{
        const layoutElements = layouts.map(({ layoutIdPath })=>({
                component: Slot,
                props: {
                    id: getLayoutSlotId(layoutIdPath)
                }
            }));
        return ()=>createNestedElements(layoutElements, /*#__PURE__*/ _jsx(Slot, {
                id: getPageSlotId(path)
            }));
    };
    /** Renders the root component */ const renderRoot = ()=>/*#__PURE__*/ createElement(rootItem ? rootItem.component : DefaultRoot, null, /*#__PURE__*/ _jsx(Children, {}));
    const registerStaticComponent = (id, component, sourceFile)=>{
        const existing = staticComponentById.get(id);
        if (existing && existing.component !== component) {
            throw new Error(`Duplicated component for: ${id}`);
        }
        staticComponentById.set(id, {
            component,
            sourceFile
        });
    };
    const isAllElementsStatic = (elements)=>Object.values(elements).every((element)=>element.isStatic);
    const isAllSlicesStatic = (path)=>sliceIdsByRoutePath.get(path).every((sliceId)=>sliceEntryById.get(sliceId)?.isStatic);
    const createPage = (page)=>{
        if (configured) {
            throw new Error('createPage no longer available');
        }
        if (!page.component) {
            return page;
        }
        const pageRoutePath = pathnameToRoutePath(page.path);
        if (pagePathExists(pageRoutePath)) {
            throw new Error(`Duplicated path: ${page.path}`);
        }
        const routePathSpec = parsePathWithSlug(pageRoutePath);
        const { numSlugs, numWildcards } = countSlugsAndWildcards(routePathSpec);
        const noSsr = page.unstable_disableSSR ?? false;
        const slices = page.slices || [];
        const sourceFile = page.unstable_sourceFile;
        const getEtag = page.unstable_getEtag;
        const searchCodec = page.unstable_searchCodec;
        if (searchCodec && page.render === 'static') {
            throw new Error(`unstable_searchCodec is not supported on a static route (${page.path}); search params need a per-request query, so use render: 'dynamic'.`);
        }
        const registerPageWithExactPath = ()=>{
            const routePath = pageRoutePath;
            const spec = parseExactPath(routePath);
            if (page.render === 'static') {
                staticPageEntryByRoutePath.set(routePath, {
                    concretePathSpec: spec,
                    noSsr,
                    sourceFile
                });
                const id = getStaticComponentId(routePath, 'page');
                registerStaticComponent(id, page.component, sourceFile);
            } else {
                dynamicPageEntryByRoutePath.set(routePath, {
                    routePathSpec: spec,
                    component: page.component,
                    noSsr,
                    sourceFile,
                    getEtag,
                    searchCodec
                });
            }
            sliceIdsByRoutePath.set(routePath, slices);
        };
        const registerStaticPageWithoutSlugs = ()=>{
            const routePath = pathnameToRoutePath(getGrouplessPath(page.path));
            staticPageEntryByRoutePath.set(routePath, {
                concretePathSpec: routePathSpec,
                noSsr,
                sourceFile
            });
            const id = getStaticComponentId(routePath, 'page');
            if (routePath !== pageRoutePath) {
                groupedRoutePathByRoutePath.set(routePath, pageRoutePath);
            }
            registerStaticComponent(id, page.component, sourceFile);
            sliceIdsByRoutePath.set(routePath, slices);
        };
        const registerStaticPageWithSlugs = (staticPathsInput)=>{
            forEachConcreteStaticPath(routePathSpec, staticPathsInput, ({ concretePath, pathItems, mapping })=>{
                const routePath = pathnameToRoutePath(getGrouplessPath(concretePath));
                const concretePathSpec = pathItems.map((name)=>({
                        type: 'literal',
                        name
                    }));
                staticPageEntryByRoutePath.set(routePath, {
                    concretePathSpec,
                    pathPatternSpec: routePathSpec,
                    noSsr,
                    sourceFile
                });
                const concreteRoutePath = pathnameToRoutePath(concretePath);
                if (routePath !== concreteRoutePath) {
                    groupedRoutePathByRoutePath.set(routePath, concreteRoutePath);
                }
                const id = getStaticComponentId(routePath, 'page');
                const WrappedComponent = (props)=>/*#__PURE__*/ createElement(page.component, {
                        ...props,
                        ...mapping
                    });
                registerStaticComponent(id, WrappedComponent, sourceFile);
                sliceIdsByRoutePath.set(routePath, slices);
            });
        };
        const registerDynamicPageWithoutWildcard = ()=>{
            const routePath = pathnameToRoutePath(getGrouplessPath(page.path));
            if (routePath !== pageRoutePath) {
                groupedRoutePathByRoutePath.set(routePath, pageRoutePath);
            }
            dynamicPageEntryByRoutePath.set(routePath, {
                routePathSpec,
                component: page.component,
                noSsr,
                sourceFile,
                getEtag,
                searchCodec
            });
            sliceIdsByRoutePath.set(routePath, slices);
        };
        const registerDynamicPageWithWildcard = ()=>{
            const routePath = pathnameToRoutePath(getGrouplessPath(page.path));
            if (routePath !== pageRoutePath) {
                groupedRoutePathByRoutePath.set(routePath, pageRoutePath);
            }
            wildcardPageEntryByRoutePath.set(routePath, {
                routePathSpec,
                component: page.component,
                noSsr,
                sourceFile,
                getEtag,
                searchCodec
            });
            sliceIdsByRoutePath.set(routePath, slices);
        };
        if (page.exactPath) {
            registerPageWithExactPath();
        } else if (page.render === 'static' && numSlugs === 0) {
            registerStaticPageWithoutSlugs();
        } else if (page.render === 'static' && numSlugs > 0 && 'staticPaths' in page) {
            registerStaticPageWithSlugs(page.staticPaths);
        } else if (page.render === 'dynamic' && numWildcards === 0) {
            registerDynamicPageWithoutWildcard();
        } else if (page.render === 'dynamic' && numWildcards === 1) {
            registerDynamicPageWithWildcard();
        } else {
            throw new Error('Invalid page configuration ' + JSON.stringify(page));
        }
        return page;
    };
    const createLayout = (layout)=>{
        if (configured) {
            throw new Error('createLayout no longer available');
        }
        if (!layout.component) {
            return;
        }
        const routePath = pathnameToRoutePath(layout.path);
        const sourceFile = layout.unstable_sourceFile;
        const getEtag = layout.unstable_getEtag;
        if (layout.render === 'static') {
            const id = getStaticComponentId(routePath, 'layout');
            registerStaticComponent(id, layout.component, sourceFile);
        } else if (layout.render === 'dynamic') {
            if (dynamicLayoutEntryByRoutePath.has(routePath)) {
                throw new Error(`Duplicated dynamic path: ${layout.path}`);
            }
            const routePathSpec = parsePathWithSlug(routePath);
            dynamicLayoutEntryByRoutePath.set(routePath, {
                routePathSpec,
                component: layout.component,
                sourceFile,
                getEtag
            });
        } else {
            throw new Error('Invalid layout configuration');
        }
    };
    const createApi = (options)=>{
        if (configured) {
            throw new Error('createApi no longer available');
        }
        if (options.render === 'static') {
            if (!options.handler) {
                return;
            }
        } else {
            if (!options.handlers || !Object.values(options.handlers).some(Boolean)) {
                return;
            }
        }
        const routePath = pathnameToRoutePath(options.path);
        const sourceFile = options.unstable_sourceFile;
        if (pagePathExists(routePath)) {
            throw new Error(`Duplicated api path: ${options.path}`);
        }
        const routePathSpec = parsePathWithSlug(routePath);
        if (options.render === 'static') {
            const { numSlugs } = countSlugsAndWildcards(routePathSpec);
            if (numSlugs > 0 && options.staticPaths) {
                forEachConcreteStaticPath(routePathSpec, options.staticPaths, ({ concretePath, pathItems, mapping })=>{
                    const concreteRoutePath = pathnameToRoutePath(concretePath);
                    if (pagePathExists(concreteRoutePath)) {
                        throw new Error(`Duplicated api path: ${concretePath}`);
                    }
                    apiEntryByRoutePath.set(concreteRoutePath, {
                        render: 'static',
                        routePathSpec: pathItems.map((name)=>({
                                type: 'literal',
                                name
                            })),
                        handlers: {
                            GET: options.handler
                        },
                        staticParams: mapping,
                        sourceFile
                    });
                });
            } else {
                apiEntryByRoutePath.set(routePath, {
                    render: 'static',
                    routePathSpec,
                    handlers: {
                        GET: options.handler
                    },
                    sourceFile
                });
            }
        } else {
            apiEntryByRoutePath.set(routePath, {
                render: 'dynamic',
                routePathSpec,
                handlers: options.handlers,
                sourceFile
            });
        }
    };
    const createRoot = (root)=>{
        if (configured) {
            throw new Error('createRoot no longer available');
        }
        if (!root.component) {
            return;
        }
        if (rootItem) {
            throw new Error(`Duplicated root component`);
        }
        if (root.render === 'static' || root.render === 'dynamic') {
            rootItem = root;
        } else {
            throw new Error('Invalid root configuration');
        }
    };
    const createSlice = (slice)=>{
        if (configured) {
            throw new Error('createSlice no longer available');
        }
        if (!slice.component) {
            return;
        }
        const slicePathSpec = parsePathWithSlug(slice.id);
        const { numSlugs } = countSlugsAndWildcards(slicePathSpec);
        const sourceFile = slice.unstable_sourceFile;
        const getEtag = slice.unstable_getEtag;
        if (slice.render === 'static' && numSlugs > 0) {
            if (!('staticPaths' in slice) || !slice.staticPaths) {
                throw new Error(`Static slice with slug requires staticPaths: ${slice.id}`);
            }
            forEachConcreteStaticPath(slicePathSpec, slice.staticPaths, ({ concretePath, mapping })=>{
                const concreteId = concretePath.replace(/^\//, '');
                if (sliceEntryById.has(concreteId)) {
                    throw new Error(`Duplicated slice id: ${concreteId}`);
                }
                const WrappedComponent = (props)=>/*#__PURE__*/ createElement(slice.component, {
                        ...props,
                        ...mapping
                    });
                sliceEntryById.set(concreteId, {
                    component: WrappedComponent,
                    isStatic: true,
                    sourceFile
                });
            });
            return;
        }
        if (sliceEntryById.has(slice.id)) {
            throw new Error(`Duplicated slice id: ${slice.id}`);
        }
        sliceEntryById.set(slice.id, {
            component: slice.component,
            isStatic: slice.render === 'static',
            sourceFile,
            getEtag
        });
    };
    const interceptors = [];
    const createInterceptor = (interceptor)=>{
        if (configured) {
            throw new Error('createInterceptor no longer available');
        }
        interceptors.push(interceptor);
    };
    let ready;
    const configure = async ()=>{
        if (!configured && !ready) {
            ready = fn({
                createPage,
                createLayout,
                createRoot,
                createApi,
                createSlice,
                createInterceptor
            });
            await ready;
            configured = true;
        }
        await ready;
    };
    const getLayouts = (routePathSpec)=>{
        const pathSegments = routePathSpec.reduce((acc, _segment, index)=>{
            acc.push(pathSpecAsString(routePathSpec.slice(0, index + 1)));
            return acc;
        }, [
            '/'
        ]);
        return pathSegments.filter((segment)=>dynamicLayoutEntryByRoutePath.has(segment) || getStaticLayout(segment));
    };
    const definedRouter = unstable_defineRouter({
        getConfigs: async ()=>{
            await configure();
            const collectLayoutMatches = (spec, routePath)=>getLayouts(spec).map((layoutPath)=>({
                        layoutPath,
                        layoutIdPath: routePath ? getLayoutIdPath(layoutPath, routePath) : layoutPath
                    }));
            // Memoize so component and getEtag get the same props reference.
            const buildElementSpec = (component, buildProps, isStatic, sourceFile, getEtag)=>{
                const propsCache = new WeakMap();
                const toProps = (option)=>{
                    if (!propsCache.has(option)) {
                        propsCache.set(option, buildProps(option));
                    }
                    return propsCache.get(option);
                };
                return {
                    isStatic,
                    renderer: (option)=>/*#__PURE__*/ createElement(component, toProps(option), /*#__PURE__*/ _jsx(Children, {})),
                    ...sourceFile ? {
                        sourceFile
                    } : {},
                    ...getEtag ? {
                        getEtagFromOption: (option)=>getEtag(toProps(option))
                    } : {}
                };
            };
            const buildLayoutElement = (layoutPath)=>{
                const dynamicEntry = dynamicLayoutEntryByRoutePath.get(layoutPath);
                const staticEntry = dynamicEntry ? undefined : staticComponentById.get(getStaticComponentId(layoutPath, 'layout'));
                const layout = dynamicEntry?.component ?? staticEntry?.component;
                if (!layout) {
                    throw new Error('Invalid layout ' + layoutPath);
                }
                const sourceFile = dynamicEntry?.sourceFile ?? staticEntry?.sourceFile;
                const getLayoutPropsMapping = createLayoutPropsMapper(layoutPath);
                return buildElementSpec(layout, (option)=>getLayoutPropsMapping(option.routePath), !dynamicEntry, sourceFile, dynamicEntry?.getEtag);
            };
            const buildLayoutElements = (matches)=>Object.fromEntries(matches.map(({ layoutPath, layoutIdPath })=>[
                        getLayoutSlotId(layoutIdPath),
                        buildLayoutElement(layoutPath)
                    ]));
            const buildPageElement = (component, getPropsMapping, isStatic, sourceFile, getEtag, searchCodec)=>buildElementSpec(component, (option)=>({
                        ...getPropsMapping(option.routePath),
                        ...option.query ? {
                            query: option.query
                        } : {},
                        ...searchCodec ? {
                            search: parseSearchOrThrow(searchCodec, option.query ?? '')
                        } : {},
                        path: option.routePath
                    }), isStatic, sourceFile, getEtag);
            const rootIsStatic = !rootItem || rootItem.render === 'static';
            const rootSourceFile = rootItem?.unstable_sourceFile;
            const rootGetEtag = rootItem?.unstable_getEtag;
            const buildRootElement = ()=>({
                    isStatic: rootIsStatic,
                    renderer: renderRoot,
                    ...rootSourceFile ? {
                        sourceFile: rootSourceFile
                    } : {},
                    ...rootGetEtag ? {
                        getEtagFromOption: ()=>rootGetEtag()
                    } : {}
                });
            const buildRouteConfigBase = (routePath, pathSpec, layouts, elements, noSsr)=>({
                    type: 'route',
                    path: pathSpec.filter((part)=>!part.name?.startsWith('(')),
                    isStatic: rootIsStatic && isAllElementsStatic(elements) && isAllSlicesStatic(routePath),
                    rootElement: buildRootElement(),
                    routeElement: {
                        isStatic: true,
                        renderer: buildRouteElement(layouts, routePath)
                    },
                    elements,
                    noSsr,
                    slices: sliceIdsByRoutePath.get(routePath)
                });
            const buildStaticRouteConfigs = ()=>Array.from(staticPageEntryByRoutePath, ([routePath, { concretePathSpec, pathPatternSpec, noSsr }])=>{
                    const groupedRoutePath = groupedRoutePathByRoutePath.get(routePath) ?? routePath;
                    const layouts = collectLayoutMatches(pathPatternSpec ?? concretePathSpec, groupedRoutePath);
                    const pageEntry = staticComponentById.get(getStaticComponentId(routePath, 'page'));
                    const getPropsMapping = createPathPropsMapper(routePath);
                    const elements = buildLayoutElements(layouts);
                    elements[getPageSlotId(routePath)] = buildPageElement(pageEntry.component, getPropsMapping, true, pageEntry.sourceFile);
                    return {
                        ...buildRouteConfigBase(routePath, concretePathSpec, layouts, elements, noSsr),
                        ...pathPatternSpec && {
                            pathPattern: pathPatternSpec
                        }
                    };
                });
            const buildDynamicLikeRouteConfig = (routePath, { routePathSpec, component, noSsr, sourceFile, getEtag, searchCodec })=>{
                const layouts = collectLayoutMatches(routePathSpec);
                const getPropsMapping = createPathPropsMapper(routePath);
                const elements = buildLayoutElements(layouts);
                elements[getPageSlotId(routePath)] = buildPageElement(component, getPropsMapping, false, sourceFile, getEtag, searchCodec);
                return {
                    ...buildRouteConfigBase(routePath, routePathSpec, layouts, elements, noSsr),
                    ...searchCodec ? {
                        searchCodec
                    } : {}
                };
            };
            const buildDynamicRouteConfigs = ()=>Array.from(dynamicPageEntryByRoutePath, ([routePath, entry])=>buildDynamicLikeRouteConfig(routePath, entry));
            const buildWildcardRouteConfigs = ()=>Array.from(wildcardPageEntryByRoutePath, ([routePath, entry])=>buildDynamicLikeRouteConfig(routePath, entry));
            const buildApiConfigs = ()=>Array.from(apiEntryByRoutePath.values(), ({ routePathSpec, render, handlers, staticParams, sourceFile })=>({
                        type: 'api',
                        path: routePathSpec,
                        isStatic: render === 'static',
                        handler: async (req, apiContext)=>{
                            const path = new URL(req.url).pathname;
                            const method = req.method;
                            const handler = handlers[method] ?? handlers.all;
                            if (!handler) {
                                throw new Error('API method not found: ' + method + 'for path: ' + path);
                            }
                            return handler(req, staticParams ? {
                                params: staticParams
                            } : apiContext);
                        },
                        ...sourceFile ? {
                            sourceFile
                        } : {}
                    }));
            const buildSliceConfigs = ()=>Array.from(sliceEntryById, ([id, { isStatic, sourceFile, getEtag }])=>{
                    const slicePathSpec = parsePathWithSlug(id);
                    const hasSlug = slicePathSpec.some((s)=>s.type !== 'literal');
                    return {
                        type: 'slice',
                        id,
                        ...hasSlug ? {
                            pathSpec: slicePathSpec
                        } : {},
                        isStatic,
                        renderer: async (params)=>{
                            const slice = sliceEntryById.get(id);
                            if (!slice) {
                                throw new Error('Slice not found: ' + id);
                            }
                            return /*#__PURE__*/ createElement(slice.component, params, /*#__PURE__*/ _jsx(Children, {}));
                        },
                        ...sourceFile ? {
                            sourceFile
                        } : {},
                        ...getEtag ? {
                            getEtagFromParams: getEtag
                        } : {}
                    };
                });
            const routeConfigs = [
                ...buildStaticRouteConfigs(),
                ...buildDynamicRouteConfigs(),
                ...buildWildcardRouteConfigs()
            ];
            const pathConfigs = [
                ...routeConfigs,
                ...buildApiConfigs()
            ]// Sort routes by priority: "standard routes" -> api routes -> api wildcard routes -> standard wildcard routes
            .sort((configA, configB)=>routePriorityComparator(configA, configB));
            return [
                ...pathConfigs,
                ...buildSliceConfigs()
            ];
        },
        ...options?.unstable_skipBuild && {
            unstable_skipBuild: options.unstable_skipBuild
        },
        unstable_interceptors: interceptors
    });
    return definedRouter;
};
function expandStaticRoutePath(routePathSpec, staticSegments) {
    const mapping = {};
    let slugIndex = 0;
    const pathItems = [];
    routePathSpec.forEach((spec)=>{
        switch(spec.type){
            case 'literal':
                pathItems.push(spec.name);
                break;
            case 'wildcard':
                mapping[spec.name] = staticSegments.slice(slugIndex);
                staticSegments.slice(slugIndex++).forEach((slug)=>{
                    pathItems.push(slug);
                });
                break;
            case 'group':
                {
                    const slug = staticSegments[slugIndex++];
                    const prefix = spec.prefix ?? '';
                    const suffix = spec.suffix ?? '';
                    pathItems.push(`${prefix}${slug}${suffix}`);
                    mapping[spec.name] = slug;
                    break;
                }
        }
    });
    const concretePath = '/' + pathItems.join('/');
    return {
        concretePath,
        pathItems,
        mapping
    };
}
